{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NCBI BLAST\n",
    "\n",
    "## Online BLAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "record = SeqIO.read('./resources/HBB-human.fasta', 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on module Bio.Blast.NCBIWWW in Bio.Blast:\n",
      "\n",
      "NAME\n",
      "    Bio.Blast.NCBIWWW - Code to invoke the NCBI BLAST server over the internet.\n",
      "\n",
      "DESCRIPTION\n",
      "    This module provides code to work with the WWW version of BLAST\n",
      "    provided by the NCBI. https://blast.ncbi.nlm.nih.gov/\n",
      "\n",
      "FUNCTIONS\n",
      "    qblast(program, database, sequence, url_base='https://blast.ncbi.nlm.nih.gov/Blast.cgi', auto_format=None, composition_based_statistics=None, db_genetic_code=None, endpoints=None, entrez_query='(none)', expect=10.0, filter=None, gapcosts=None, genetic_code=None, hitlist_size=50, i_thresh=None, layout=None, lcase_mask=None, matrix_name=None, nucl_penalty=None, nucl_reward=None, other_advanced=None, perc_ident=None, phi_pattern=None, query_file=None, query_believe_defline=None, query_from=None, query_to=None, searchsp_eff=None, service=None, threshold=None, ungapped_alignment=None, word_size=None, short_query=None, alignments=500, alignment_view=None, descriptions=500, entrez_links_new_window=None, expect_low=None, expect_high=None, format_entrez_query=None, format_object=None, format_type='XML', ncbi_gi=None, results_file=None, show_overview=None, megablast=None, template_type=None, template_length=None)\n",
      "        BLAST search using NCBI's QBLAST server or a cloud service provider.\n",
      "        \n",
      "        Supports all parameters of the old qblast API for Put and Get.\n",
      "        \n",
      "        Please note that NCBI uses the new Common URL API for BLAST searches\n",
      "        on the internet (http://ncbi.github.io/blast-cloud/dev/api.html). Thus,\n",
      "        some of the parameters used by this function are not (or are no longer)\n",
      "        officially supported by NCBI. Although they are still functioning, this\n",
      "        may change in the future.\n",
      "        \n",
      "        The Common URL API (http://ncbi.github.io/blast-cloud/dev/api.html) allows\n",
      "        doing BLAST searches on cloud servers. To use this feature, please set\n",
      "        ``url_base='http://host.my.cloud.service.provider.com/cgi-bin/blast.cgi'``\n",
      "        and ``format_object='Alignment'``. For more details, please see\n",
      "        https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=CloudBlast\n",
      "        \n",
      "        Some useful parameters:\n",
      "        \n",
      "         - program        blastn, blastp, blastx, tblastn, or tblastx (lower case)\n",
      "         - database       Which database to search against (e.g. \"nr\").\n",
      "         - sequence       The sequence to search.\n",
      "         - ncbi_gi        TRUE/FALSE whether to give 'gi' identifier.\n",
      "         - descriptions   Number of descriptions to show.  Def 500.\n",
      "         - alignments     Number of alignments to show.  Def 500.\n",
      "         - expect         An expect value cutoff.  Def 10.0.\n",
      "         - matrix_name    Specify an alt. matrix (PAM30, PAM70, BLOSUM80, BLOSUM45).\n",
      "         - filter         \"none\" turns off filtering.  Default no filtering\n",
      "         - format_type    \"HTML\", \"Text\", \"ASN.1\", or \"XML\".  Def. \"XML\".\n",
      "         - entrez_query   Entrez query to limit Blast search\n",
      "         - hitlist_size   Number of hits to return. Default 50\n",
      "         - megablast      TRUE/FALSE whether to use MEga BLAST algorithm (blastn only)\n",
      "         - short_query    TRUE/FALSE whether to adjust the search parameters for a\n",
      "                          short query sequence. Note that this will override\n",
      "                          manually set parameters like word size and e value. Turns\n",
      "                          off when sequence length is > 30 residues. Default: None.\n",
      "         - service        plain, psi, phi, rpsblast, megablast (lower case)\n",
      "        \n",
      "        This function does no checking of the validity of the parameters\n",
      "        and passes the values to the server as is.  More help is available at:\n",
      "        https://ncbi.github.io/blast-cloud/dev/api.html\n",
      "\n",
      "DATA\n",
      "    NCBI_BLAST_URL = 'https://blast.ncbi.nlm.nih.gov/Blast.cgi'\n",
      "\n",
      "FILE\n",
      "    c:\\users\\sÃ©bastien\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\bio\\blast\\ncbiwww.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "help(NCBIWWW)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_cmd = NCBIWWW.qblast(\n",
    "    program='blastn', # DNA sequences\n",
    "    database='nt',\n",
    "    sequence=record.seq,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('blast_results.xml', 'w+') as w:\n",
    "    w.write(blast_cmd.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing BLAST output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "record = NCBIXML.read(open('./blast_results.xml'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Descriptions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***\n",
      "Alignment Index: 0\n",
      "Title: gi|1401724401|ref|NM_000518.5| Homo sapiens hemoglobin subunit beta (HBB), mRNA\n",
      "Score: 1256.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 1\n",
      "Title: gi|694943029|ref|XM_508242.4| PREDICTED: Pan troglodytes hemoglobin subunit beta (HBB), mRNA\n",
      "Score: 1251.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 2\n",
      "Title: gi|1849004923|ref|XM_003819029.3| PREDICTED: Pan paniscus hemoglobin subunit beta (LOC100976465), mRNA\n",
      "Score: 1246.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 3\n",
      "Title: gi|13937928|gb|BC007075.1| Homo sapiens hemoglobin, beta, mRNA (cDNA clone MGC:14540 IMAGE:4292125), complete cds\n",
      "Score: 1239.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 4\n",
      "Title: gi|29436|emb|V00497.1| Human messenger RNA for beta-globin\n",
      "Score: 1237.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 5\n",
      "Title: gi|1099100069|ref|XM_019036164.1| PREDICTED: Gorilla gorilla gorilla hemoglobin subunit beta (LOC101126932), mRNA\n",
      "Score: 1231.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 6\n",
      "Title: gi|40886940|gb|AY509193.1| Homo sapiens hemoglobin beta mRNA, complete cds\n",
      "Score: 1231.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 7\n",
      "Title: gi|1351393134|ref|XM_002822127.4| PREDICTED: Pongo abelii hemoglobin subunit beta (HBB), mRNA\n",
      "Score: 1199.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 8\n",
      "Title: gi|1743165722|ref|XM_004090649.3| PREDICTED: Nomascus leucogenys hemoglobin subunit beta (HBB), mRNA\n",
      "Score: 1186.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 9\n",
      "Title: gi|1800003084|ref|XM_032166808.1| PREDICTED: Hylobates moloch hemoglobin subunit beta (HBB), mRNA\n",
      "Score: 1181.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 10\n",
      "Title: gi|187940240|gb|EU694432.1| Homo sapiens hemoglobin beta chain variant (HBB) mRNA, HBB-Dothan allele, complete cds\n",
      "Score: 1158.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 11\n",
      "Title: gi|6003533|gb|AF181989.1|AF181989 Homo sapiens hemoglobin beta subunit variant (HBB) mRNA, complete cds\n",
      "Score: 1153.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 12\n",
      "Title: gi|13549111|gb|AF349114.1|AF349114 Homo sapiens beta globin chain variant (HBB) mRNA, complete cds\n",
      "Score: 1147.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 13\n",
      "Title: gi|795328120|ref|XM_011963601.1| PREDICTED: Colobus angolensis palliatus hemoglobin, beta (HBB), mRNA\n",
      "Score: 1141.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 14\n",
      "Title: gi|1825958351|ref|XM_033207068.1| PREDICTED: Trachypithecus francoisi hemoglobin subunit beta (LOC117081249), mRNA\n",
      "Score: 1139.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 15\n",
      "Title: gi|1751218711|ref|XM_010363344.2| PREDICTED: Rhinopithecus roxellana hemoglobin subunit beta (LOC104662339), mRNA\n",
      "Score: 1134.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 16\n",
      "Title: gi|1059130395|ref|XM_017861621.1| PREDICTED: Rhinopithecus bieti hemoglobin subunit beta (LOC108521664), mRNA\n",
      "Score: 1134.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 17\n",
      "Title: gi|795109730|ref|XM_011975167.1| PREDICTED: Mandrillus leucophaeus hemoglobin subunit beta (LOC105535916), transcript variant X2, mRNA\n",
      "Score: 1131.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 18\n",
      "Title: gi|1297635475|ref|XM_023209613.1| PREDICTED: Piliocolobus tephrosceles hemoglobin subunit beta (HBB), mRNA\n",
      "Score: 1121.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 19\n",
      "Title: gi|1050115306|ref|NM_001329918.1| Chlorocebus sabaeus hemoglobin subunit beta (HBB), mRNA\n",
      "Score: 1114.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 20\n",
      "Title: gi|4378803|gb|AF117710.1|AF117710 Homo sapiens hemoglobin beta chain (HBB) mRNA, complete cds\n",
      "Score: 1114.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 21\n",
      "Title: gi|183944|gb|M25113.1|HUMHEMOB Human sickle beta-hemoglobin mRNA\n",
      "Score: 1114.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 22\n",
      "Title: gi|817284032|ref|XM_012456809.1| PREDICTED: Aotus nancymaae hemoglobin subunit beta (LOC105720341), mRNA\n",
      "Score: 1106.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 23\n",
      "Title: gi|795109736|ref|XM_011975168.1| PREDICTED: Mandrillus leucophaeus hemoglobin subunit beta (LOC105535916), transcript variant X3, mRNA\n",
      "Score: 1106.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 24\n",
      "Title: gi|1411109004|ref|XM_025358026.1| PREDICTED: Theropithecus gelada hemoglobin subunit beta (LOC112607021), transcript variant X2, mRNA\n",
      "Score: 1101.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 25\n",
      "Title: gi|795544769|ref|XM_011718821.1| PREDICTED: Macaca nemestrina hemoglobin subunit beta (LOC105468624), mRNA\n",
      "Score: 1099.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 26\n",
      "Title: gi|1804407134|ref|XM_032240524.1| PREDICTED: Sapajus apella hemoglobin subunit beta (HBB), mRNA\n",
      "Score: 1086.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 27\n",
      "Title: gi|725557265|ref|XM_003923369.2| PREDICTED: Saimiri boliviensis boliviensis hemoglobin subunit beta (LOC101036159), mRNA\n",
      "Score: 1076.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 28\n",
      "Title: gi|1044422541|ref|XM_017507286.1| PREDICTED: Cebus imitator hemoglobin subunit beta (HBB), mRNA\n",
      "Score: 1071.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 29\n",
      "Title: gi|90077965|dbj|AB171600.1| Macaca fascicularis brain cDNA clone: QflA-16301, similar to human hemoglobin, beta (HBB), mRNA, RefSeq: NM_000518.4\n",
      "Score: 1068.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 30\n",
      "Title: gi|29445|emb|V00500.1| Human messenger RNA for beta-globin\n",
      "Score: 1061.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 31\n",
      "Title: gi|675697671|ref|XM_002754891.3| PREDICTED: Callithrix jacchus hemoglobin subunit beta (LOC100410611), mRNA\n",
      "Score: 1060.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 32\n",
      "Title: gi|23268448|gb|AY136510.1| Homo sapiens hemoglobin beta chain variant Hb S-Wake (HBB) mRNA, complete cds\n",
      "Score: 1053.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 33\n",
      "Title: gi|164697558|dbj|AK311825.1| Homo sapiens cDNA, FLJ92086, Homo sapiens hemoglobin, beta (HBB), mRNA\n",
      "Score: 988.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 34\n",
      "Title: gi|6003531|gb|AF181832.1|AF181832 Homo sapiens hemoglobin beta subunit variant (HBB) mRNA, partial cds\n",
      "Score: 971.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 35\n",
      "Title: gi|640790084|ref|XM_008052701.1| PREDICTED: Carlito syrichta hemoglobin subunit beta (LOC103254684), mRNA\n",
      "Score: 900.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 36\n",
      "Title: gi|1101972663|tpe|LT548096.1| TPA_inf: Homo sapiens GLNA1 gene for globin A1\n",
      "Score: 888.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 37\n",
      "Title: gi|123992966|gb|DQ893159.2| Synthetic construct clone IMAGE:100005789; FLH194450.01X; RZPDo839B0680D hemoglobin, beta (HBB) gene, encodes complete protein\n",
      "Score: 888.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 38\n",
      "Title: gi|123999886|gb|DQ896453.2| Synthetic construct Homo sapiens clone IMAGE:100010913; FLH194446.01L; RZPDo839B0670D hemoglobin, beta (HBB) gene, encodes complete protein >gi|157928557|gb|EU176774.1| Synthetic construct Homo sapiens clone IMAGE:100011599; FLH263683.01L; RZPDo839F03256D hemoglobin, beta (HBB) gene, encodes complete protein\n",
      "Score: 883.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 39\n",
      "Title: gi|49168543|emb|CR536530.1| Homo sapiens full open reading frame cDNA clone RZPDo834D0222D for gene HBB, hemoglobin, beta; complete cds, incl. stopcodon >gi|1101972687|tpe|LT548108.1| TPA_inf: Pan troglodytes GLNA1 gene for globin A1\n",
      "Score: 883.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 40\n",
      "Title: gi|117645525|emb|AM393351.1| Synthetic construct Homo sapiens clone IMAGE:100002185 for hypothetical protein (HBB gene) >gi|117646599|emb|AM392537.1| Synthetic construct Homo sapiens clone IMAGE:100002184 for hypothetical protein (HBB gene)\n",
      "Score: 880.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 41\n",
      "Title: gi|1147681442|emb|LT736709.1| Human ORFeome Gateway entry vector pENTR223-HBB, complete sequence\n",
      "Score: 878.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 42\n",
      "Title: gi|649119173|gb|KJ896959.1| Synthetic construct Homo sapiens clone ccsbBroadEn_06353 HBB gene, encodes complete protein\n",
      "Score: 878.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 43\n",
      "Title: gi|61361759|gb|AY890157.1| Synthetic construct Homo sapiens clone FLH028849.01X hemoglobin beta (HBB) mRNA, complete cds\n",
      "Score: 878.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 44\n",
      "Title: gi|49456780|emb|CR541913.1| Homo sapiens full open reading frame cDNA clone RZPDo834E0633D for gene HBB, hemoglobin, beta; complete cds, without stopcodon\n",
      "Score: 877.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 45\n",
      "Title: gi|60833496|gb|AY894014.1| Synthetic construct Homo sapiens clone FLH130860.01L hemoglobin beta (HBB) mRNA, partial cds\n",
      "Score: 877.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 46\n",
      "Title: gi|795109724|ref|XM_011975165.1| PREDICTED: Mandrillus leucophaeus hemoglobin subunit beta (LOC105535916), transcript variant X1, mRNA\n",
      "Score: 876.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 47\n",
      "Title: gi|60653726|gb|AY892640.1| Synthetic construct Homo sapiens clone FLH028845.01L hemoglobin beta (HBB) mRNA, partial cds\n",
      "Score: 874.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 48\n",
      "Title: gi|823670799|gb|KR710229.1| Synthetic construct Homo sapiens clone CCSBHm_00010626 HBB (HBB) mRNA, encodes complete protein\n",
      "Score: 873.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 49\n",
      "Title: gi|823670797|gb|KR710228.1| Synthetic construct Homo sapiens clone CCSBHm_00010601 HBB (HBB) mRNA, encodes complete protein\n",
      "Score: 873.0\n",
      "E_value: 0.0\n"
     ]
    }
   ],
   "source": [
    "for ind, align in enumerate(record.descriptions):\n",
    "    print('***')\n",
    "    print('Alignment Index:', ind)\n",
    "    print('Title:', align.title)\n",
    "    print('Score:', align.score)\n",
    "    print('E_value:', align.e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***\n",
      "Alignment Index: 0\n",
      "1. Title: gi|1401724401|ref|NM_000518.5| Homo sapiens hemoglobin subunit beta (HBB), mRNA\n",
      "2. Length: 628\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D2896C40>]\n",
      "3.1. Score: 1256.0\n",
      "3.2. Bits: 1133.8\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 628\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 1\n",
      "1. Title: gi|694943029|ref|XM_508242.4| PREDICTED: Pan troglodytes hemoglobin subunit beta (HBB), mRNA\n",
      "2. Length: 748\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB0A0>]\n",
      "3.1. Score: 1251.0\n",
      "3.2. Bits: 1129.29\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 627\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 121\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||| |||||||||||||||||||||||||||||||||||||||||\n",
      "ATGGTGCACCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 2\n",
      "1. Title: gi|1849004923|ref|XM_003819029.3| PREDICTED: Pan paniscus hemoglobin subunit beta (LOC100976465), mRNA\n",
      "2. Length: 646\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D40DB130>]\n",
      "3.1. Score: 1246.0\n",
      "3.2. Bits: 1124.79\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 626\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 19\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||| |||||||||||||||||||||||||||||||||||||||||\n",
      "ATGGTGCACCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 3\n",
      "1. Title: gi|13937928|gb|BC007075.1| Homo sapiens hemoglobin, beta, mRNA (cDNA clone MGC:14540 IMAGE:4292125), complete cds\n",
      "2. Length: 658\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB160>]\n",
      "3.1. Score: 1239.0\n",
      "3.2. Bits: 1118.47\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 624\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 2\n",
      "3.10. Alignment Length: 627\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||| |||||||||||||||||||||||||||||||||||||||||\n",
      "ATGGTGCACCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 4\n",
      "1. Title: gi|29436|emb|V00497.1| Human messenger RNA for beta-globin\n",
      "2. Length: 626\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB1F0>]\n",
      "3.1. Score: 1237.0\n",
      "3.2. Bits: 1116.67\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 623\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 626\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||| ||||||||||||||||||||||| |||||||||||||||||\n",
      "ATGGTGCACCTGACTCCTGAGGAGAAGTCTGCGGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 5\n",
      "1. Title: gi|1099100069|ref|XM_019036164.1| PREDICTED: Gorilla gorilla gorilla hemoglobin subunit beta (LOC101126932), mRNA\n",
      "2. Length: 753\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB280>]\n",
      "3.1. Score: 1231.0\n",
      "3.2. Bits: 1111.26\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 623\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 126\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||| |||||||||||||||||||||||||||||||||||||||||\n",
      "ATGGTGCACCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 6\n",
      "1. Title: gi|40886940|gb|AY509193.1| Homo sapiens hemoglobin beta mRNA, complete cds\n",
      "2. Length: 647\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB310>]\n",
      "3.1. Score: 1231.0\n",
      "3.2. Bits: 1111.26\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 623\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||| |||||||||||||||||||||||||||||||||||||||||\n",
      "ATGGTGCACCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 7\n",
      "1. Title: gi|1351393134|ref|XM_002822127.4| PREDICTED: Pongo abelii hemoglobin subunit beta (HBB), mRNA\n",
      "2. Length: 627\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB3A0>]\n",
      "3.1. Score: 1199.0\n",
      "3.2. Bits: 1082.41\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 616\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 627\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||| ||||||||||||||||||||||| |||||||||||||||||\n",
      "ATGGTGCACCTGACTCCTGAGGAGAAGTCTGCTGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 8\n",
      "1. Title: gi|1743165722|ref|XM_004090649.3| PREDICTED: Nomascus leucogenys hemoglobin subunit beta (HBB), mRNA\n",
      "2. Length: 753\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB430>]\n",
      "3.1. Score: 1186.0\n",
      "3.2. Bits: 1070.68\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 614\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 126\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||| |||||||||||||||||||| ||||||||||||||||||||\n",
      "ATGGTGCACCTGACTCCTGAGGAGAAGTCCGCCGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 9\n",
      "1. Title: gi|1800003084|ref|XM_032166808.1| PREDICTED: Hylobates moloch hemoglobin subunit beta (HBB), mRNA\n",
      "2. Length: 753\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB4C0>]\n",
      "3.1. Score: 1181.0\n",
      "3.2. Bits: 1066.18\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 613\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 126\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "||||| || ||||||||||||||||||||||| |||||||||||||||||\n",
      "ATGGTACACCTGACTCCTGAGGAGAAGTCTGCTGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 10\n",
      "1. Title: gi|187940240|gb|EU694432.1| Homo sapiens hemoglobin beta chain variant (HBB) mRNA, HBB-Dothan allele, complete cds\n",
      "2. Length: 587\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB550>]\n",
      "3.1. Score: 1158.0\n",
      "3.2. Bits: 1045.44\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 586\n",
      "3.5. Gaps: 3\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 590\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||| |||||||||||||||||||||||||||||||||||||||||\n",
      "ATGGTGCACCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 11\n",
      "1. Title: gi|6003533|gb|AF181989.1|AF181989 Homo sapiens hemoglobin beta subunit variant (HBB) mRNA, complete cds\n",
      "2. Length: 579\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB5E0>]\n",
      "3.1. Score: 1153.0\n",
      "3.2. Bits: 1040.93\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 578\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 18\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 579\n",
      "TGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "TGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATG\n",
      "***\n",
      "Alignment Index: 12\n",
      "1. Title: gi|13549111|gb|AF349114.1|AF349114 Homo sapiens beta globin chain variant (HBB) mRNA, complete cds\n",
      "2. Length: 581\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB670>]\n",
      "3.1. Score: 1147.0\n",
      "3.2. Bits: 1035.52\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 578\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 16\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 581\n",
      "CCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGA\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "CCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGA\n",
      "***\n",
      "Alignment Index: 13\n",
      "1. Title: gi|795328120|ref|XM_011963601.1| PREDICTED: Colobus angolensis palliatus hemoglobin, beta (HBB), mRNA\n",
      "2. Length: 753\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB700>]\n",
      "3.1. Score: 1141.0\n",
      "3.2. Bits: 1030.11\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 605\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 126\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||||||||||||||| |||||| |||||||||| |||||||||||\n",
      "ATGGTGCATCTGACTCCTGATGAGAAGGCTGCCGTTACCGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 14\n",
      "1. Title: gi|1825958351|ref|XM_033207068.1| PREDICTED: Trachypithecus francoisi hemoglobin subunit beta (LOC117081249), mRNA\n",
      "2. Length: 749\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB790>]\n",
      "3.1. Score: 1139.0\n",
      "3.2. Bits: 1028.3\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 606\n",
      "3.5. Gaps: 1\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 121\n",
      "3.10. Alignment Length: 629\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "||||||||||||||||||||||||||| |||||||||| |||||||||||\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGGCTGCCGTTACCGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 15\n",
      "1. Title: gi|1751218711|ref|XM_010363344.2| PREDICTED: Rhinopithecus roxellana hemoglobin subunit beta (LOC104662339), mRNA\n",
      "2. Length: 749\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB820>]\n",
      "3.1. Score: 1134.0\n",
      "3.2. Bits: 1023.8\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 605\n",
      "3.5. Gaps: 1\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 121\n",
      "3.10. Alignment Length: 629\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||||||||||||||| |||||| ||||||||||||||||||||||\n",
      "ATGGTGCATCTGACTCCTGATGAGAAGGCTGCCGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 16\n",
      "1. Title: gi|1059130395|ref|XM_017861621.1| PREDICTED: Rhinopithecus bieti hemoglobin subunit beta (LOC108521664), mRNA\n",
      "2. Length: 749\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB8B0>]\n",
      "3.1. Score: 1134.0\n",
      "3.2. Bits: 1023.8\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 605\n",
      "3.5. Gaps: 1\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 121\n",
      "3.10. Alignment Length: 629\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||||||||||||||| |||||| ||||||||||||||||||||||\n",
      "ATGGTGCATCTGACTCCTGATGAGAAGGCTGCCGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 17\n",
      "1. Title: gi|795109730|ref|XM_011975167.1| PREDICTED: Mandrillus leucophaeus hemoglobin subunit beta (LOC105535916), transcript variant X2, mRNA\n",
      "2. Length: 753\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB940>]\n",
      "3.1. Score: 1131.0\n",
      "3.2. Bits: 1021.09\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 603\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 126\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "||||||||||||||||||||||||||| ||||||||||  ||||||||||\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGACTGCCGTTACCACCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 18\n",
      "1. Title: gi|1297635475|ref|XM_023209613.1| PREDICTED: Piliocolobus tephrosceles hemoglobin subunit beta (HBB), mRNA\n",
      "2. Length: 748\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28AB9D0>]\n",
      "3.1. Score: 1121.0\n",
      "3.2. Bits: 1012.07\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 601\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 121\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||||||||||||||| |||||| ||||||||||   |||||||||\n",
      "ATGGTGCATCTGACTCCTGATGAGAAGGCTGCCGTTACCAACCTGTGGGG\n",
      "***\n",
      "Alignment Index: 19\n",
      "1. Title: gi|1050115306|ref|NM_001329918.1| Chlorocebus sabaeus hemoglobin subunit beta (HBB), mRNA\n",
      "2. Length: 616\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28ABA60>]\n",
      "3.1. Score: 1114.0\n",
      "3.2. Bits: 1005.76\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 595\n",
      "3.5. Gaps: 4\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 7\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 620\n",
      "CATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGT\n",
      "||||||||||||||||||||| ||||||||||  ||||||||||||||||\n",
      "CATCTGACTCCTGAGGAGAAGACTGCCGTTACCACCCTGTGGGGCAAGGT\n",
      "***\n",
      "Alignment Index: 20\n",
      "1. Title: gi|4378803|gb|AF117710.1|AF117710 Homo sapiens hemoglobin beta chain (HBB) mRNA, complete cds\n",
      "2. Length: 562\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28ABAF0>]\n",
      "3.1. Score: 1114.0\n",
      "3.2. Bits: 1005.76\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 560\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 21\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 562\n",
      "GGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "GGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAG\n",
      "***\n",
      "Alignment Index: 21\n",
      "1. Title: gi|183944|gb|M25113.1|HUMHEMOB Human sickle beta-hemoglobin mRNA\n",
      "2. Length: 576\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28ABB80>]\n",
      "3.1. Score: 1114.0\n",
      "3.2. Bits: 1005.76\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 567\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 51\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 576\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "***\n",
      "Alignment Index: 22\n",
      "1. Title: gi|817284032|ref|XM_012456809.1| PREDICTED: Aotus nancymaae hemoglobin subunit beta (LOC105720341), mRNA\n",
      "2. Length: 748\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28ABC10>]\n",
      "3.1. Score: 1106.0\n",
      "3.2. Bits: 998.549\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 598\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 121\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||||||||||  ||| |||||  |||||||||| |||||||||||\n",
      "ATGGTGCATCTGACTGGTGAAGAGAAAGCTGCCGTTACCGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 23\n",
      "1. Title: gi|795109736|ref|XM_011975168.1| PREDICTED: Mandrillus leucophaeus hemoglobin subunit beta (LOC105535916), transcript variant X3, mRNA\n",
      "2. Length: 622\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28ABCA0>]\n",
      "3.1. Score: 1106.0\n",
      "3.2. Bits: 998.549\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 589\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 16\n",
      "3.9. Subject Start: 10\n",
      "3.10. Alignment Length: 613\n",
      "CCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGA\n",
      "|||||||||||| ||||||||||  |||||||||||||||||||||||||\n",
      "CCTGAGGAGAAGACTGCCGTTACCACCCTGTGGGGCAAGGTGAACGTGGA\n",
      "***\n",
      "Alignment Index: 24\n",
      "1. Title: gi|1411109004|ref|XM_025358026.1| PREDICTED: Theropithecus gelada hemoglobin subunit beta (LOC112607021), transcript variant X2, mRNA\n",
      "2. Length: 753\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28ABD30>]\n",
      "3.1. Score: 1101.0\n",
      "3.2. Bits: 994.041\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 597\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 126\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||||||||||||||||||||||  |||||||||  ||||||||||\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGAATGCCGTTACCACCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 25\n",
      "1. Title: gi|795544769|ref|XM_011718821.1| PREDICTED: Macaca nemestrina hemoglobin subunit beta (LOC105468624), mRNA\n",
      "2. Length: 749\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28ABDC0>]\n",
      "3.1. Score: 1099.0\n",
      "3.2. Bits: 992.237\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 598\n",
      "3.5. Gaps: 1\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 121\n",
      "3.10. Alignment Length: 629\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||||||||||||||||||||||  |||||| ||  ||||||||||\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGAATGCCGTCACCACCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 26\n",
      "1. Title: gi|1804407134|ref|XM_032240524.1| PREDICTED: Sapajus apella hemoglobin subunit beta (HBB), mRNA\n",
      "2. Length: 678\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28ABE80>]\n",
      "3.1. Score: 1086.0\n",
      "3.2. Bits: 980.516\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 594\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 51\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "||||||||||||||| |||| ||||| |||||||||||  ||||||||||\n",
      "ATGGTGCATCTGACTGCTGAAGAGAAATCTGCCGTTACCACCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 27\n",
      "1. Title: gi|725557265|ref|XM_003923369.2| PREDICTED: Saimiri boliviensis boliviensis hemoglobin subunit beta (LOC101036159), mRNA\n",
      "2. Length: 753\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28ABF10>]\n",
      "3.1. Score: 1076.0\n",
      "3.2. Bits: 971.499\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 592\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 126\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||||||||||  ||| |||||  |||||||||| |||||||||||\n",
      "ATGGTGCATCTGACTGGTGATGAGAAAGCTGCCGTTACCGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 28\n",
      "1. Title: gi|1044422541|ref|XM_017507286.1| PREDICTED: Cebus imitator hemoglobin subunit beta (HBB), mRNA\n",
      "2. Length: 678\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28ABFA0>]\n",
      "3.1. Score: 1071.0\n",
      "3.2. Bits: 966.99\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 591\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 51\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "||||||||||||||| |||| ||||| |||||||||||  ||||||||||\n",
      "ATGGTGCATCTGACTGCTGAAGAGAAATCTGCCGTTACCACCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 29\n",
      "1. Title: gi|90077965|dbj|AB171600.1| Macaca fascicularis brain cDNA clone: QflA-16301, similar to human hemoglobin, beta (HBB), mRNA, RefSeq: NM_000518.4\n",
      "2. Length: 604\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5070>]\n",
      "3.1. Score: 1068.0\n",
      "3.2. Bits: 964.285\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 576\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 604\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||||||||||||||||||||||  |||||| ||  ||||||||||\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGAATGCCGTCACCACCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 30\n",
      "1. Title: gi|29445|emb|V00500.1| Human messenger RNA for beta-globin\n",
      "2. Length: 549\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5100>]\n",
      "3.1. Score: 1061.0\n",
      "3.2. Bits: 957.973\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 540\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 51\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 549\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "***\n",
      "Alignment Index: 31\n",
      "1. Title: gi|675697671|ref|XM_002754891.3| PREDICTED: Callithrix jacchus hemoglobin subunit beta (LOC100410611), mRNA\n",
      "2. Length: 744\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5190>]\n",
      "3.1. Score: 1060.0\n",
      "3.2. Bits: 957.072\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 589\n",
      "3.5. Gaps: 4\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 121\n",
      "3.10. Alignment Length: 628\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "|||||||||||||||  ||| || || |||||||||||||||||||||||\n",
      "ATGGTGCATCTGACTGGTGAAGAAAAATCTGCCGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 32\n",
      "1. Title: gi|23268448|gb|AY136510.1| Homo sapiens hemoglobin beta chain variant Hb S-Wake (HBB) mRNA, complete cds\n",
      "2. Length: 534\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5220>]\n",
      "3.1. Score: 1053.0\n",
      "3.2. Bits: 950.76\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 531\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 51\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 534\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "***\n",
      "Alignment Index: 33\n",
      "1. Title: gi|164697558|dbj|AK311825.1| Homo sapiens cDNA, FLJ92086, Homo sapiens hemoglobin, beta (HBB), mRNA\n",
      "2. Length: 494\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B52B0>]\n",
      "3.1. Score: 988.0\n",
      "3.2. Bits: 892.15\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 494\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 1\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 494\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG\n",
      "***\n",
      "Alignment Index: 34\n",
      "1. Title: gi|6003531|gb|AF181832.1|AF181832 Homo sapiens hemoglobin beta subunit variant (HBB) mRNA, partial cds\n",
      "2. Length: 488\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5340>]\n",
      "3.1. Score: 971.0\n",
      "3.2. Bits: 876.822\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 487\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 108\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 488\n",
      "CCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "CCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTG\n",
      "***\n",
      "Alignment Index: 35\n",
      "1. Title: gi|640790084|ref|XM_008052701.1| PREDICTED: Carlito syrichta hemoglobin subunit beta (LOC103254684), mRNA\n",
      "2. Length: 750\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B53D0>]\n",
      "3.1. Score: 900.0\n",
      "3.2. Bits: 812.802\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 555\n",
      "3.5. Gaps: 2\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 5\n",
      "3.9. Subject Start: 129\n",
      "3.10. Alignment Length: 624\n",
      "TGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAG\n",
      "||||||||||| |||| |||||| | ||||| ||||||||||||||||||\n",
      "TGCATCTGACTGCTGAAGAGAAGGCCGCCGTCACTGCCCTGTGGGGCAAG\n",
      "***\n",
      "Alignment Index: 36\n",
      "1. Title: gi|1101972663|tpe|LT548096.1| TPA_inf: Homo sapiens GLNA1 gene for globin A1\n",
      "2. Length: 444\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5460>]\n",
      "3.1. Score: 888.0\n",
      "3.2. Bits: 801.982\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 444\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 51\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 444\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "***\n",
      "Alignment Index: 37\n",
      "1. Title: gi|123992966|gb|DQ893159.2| Synthetic construct clone IMAGE:100005789; FLH194450.01X; RZPDo839B0680D hemoglobin, beta (HBB) gene, encodes complete protein\n",
      "2. Length: 484\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B54F0>]\n",
      "3.1. Score: 888.0\n",
      "3.2. Bits: 801.982\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 456\n",
      "3.5. Gaps: 2\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 47\n",
      "3.9. Subject Start: 19\n",
      "3.10. Alignment Length: 463\n",
      "GGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "GGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTG\n",
      "***\n",
      "Alignment Index: 38\n",
      "1. Title: gi|123999886|gb|DQ896453.2| Synthetic construct Homo sapiens clone IMAGE:100010913; FLH194446.01L; RZPDo839B0670D hemoglobin, beta (HBB) gene, encodes complete protein >gi|157928557|gb|EU176774.1| Synthetic construct Homo sapiens clone IMAGE:100011599; FLH263683.01L; RZPDo839F03256D hemoglobin, beta (HBB) gene, encodes complete protein\n",
      "2. Length: 484\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5580>]\n",
      "3.1. Score: 883.0\n",
      "3.2. Bits: 797.474\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 455\n",
      "3.5. Gaps: 2\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 47\n",
      "3.9. Subject Start: 19\n",
      "3.10. Alignment Length: 463\n",
      "GGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "GGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTG\n",
      "***\n",
      "Alignment Index: 39\n",
      "1. Title: gi|49168543|emb|CR536530.1| Homo sapiens full open reading frame cDNA clone RZPDo834D0222D for gene HBB, hemoglobin, beta; complete cds, incl. stopcodon >gi|1101972687|tpe|LT548108.1| TPA_inf: Pan troglodytes GLNA1 gene for globin A1\n",
      "2. Length: 444\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5610>]\n",
      "3.1. Score: 883.0\n",
      "3.2. Bits: 797.474\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 443\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 51\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 444\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "***\n",
      "Alignment Index: 40\n",
      "1. Title: gi|117645525|emb|AM393351.1| Synthetic construct Homo sapiens clone IMAGE:100002185 for hypothetical protein (HBB gene) >gi|117646599|emb|AM392537.1| Synthetic construct Homo sapiens clone IMAGE:100002184 for hypothetical protein (HBB gene)\n",
      "2. Length: 483\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B56A0>]\n",
      "3.1. Score: 880.0\n",
      "3.2. Bits: 794.769\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 443\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 47\n",
      "3.9. Subject Start: 17\n",
      "3.10. Alignment Length: 445\n",
      "GGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "GGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTG\n",
      "***\n",
      "Alignment Index: 41\n",
      "1. Title: gi|1147681442|emb|LT736709.1| Human ORFeome Gateway entry vector pENTR223-HBB, complete sequence\n",
      "2. Length: 3231\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5730>]\n",
      "3.1. Score: 878.0\n",
      "3.2. Bits: 792.965\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 442\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 50\n",
      "3.9. Subject Start: 487\n",
      "3.10. Alignment Length: 444\n",
      "GCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "GCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTG\n",
      "***\n",
      "Alignment Index: 42\n",
      "1. Title: gi|649119173|gb|KJ896959.1| Synthetic construct Homo sapiens clone ccsbBroadEn_06353 HBB gene, encodes complete protein\n",
      "2. Length: 573\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B57C0>]\n",
      "3.1. Score: 878.0\n",
      "3.2. Bits: 792.965\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 442\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 50\n",
      "3.9. Subject Start: 65\n",
      "3.10. Alignment Length: 444\n",
      "GCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "GCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTG\n",
      "***\n",
      "Alignment Index: 43\n",
      "1. Title: gi|61361759|gb|AY890157.1| Synthetic construct Homo sapiens clone FLH028849.01X hemoglobin beta (HBB) mRNA, complete cds\n",
      "2. Length: 444\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5850>]\n",
      "3.1. Score: 878.0\n",
      "3.2. Bits: 792.965\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 442\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 51\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 444\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "***\n",
      "Alignment Index: 44\n",
      "1. Title: gi|49456780|emb|CR541913.1| Homo sapiens full open reading frame cDNA clone RZPDo834E0633D for gene HBB, hemoglobin, beta; complete cds, without stopcodon\n",
      "2. Length: 441\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B58E0>]\n",
      "3.1. Score: 877.0\n",
      "3.2. Bits: 792.064\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 440\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 51\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 441\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "***\n",
      "Alignment Index: 45\n",
      "1. Title: gi|60833496|gb|AY894014.1| Synthetic construct Homo sapiens clone FLH130860.01L hemoglobin beta (HBB) mRNA, partial cds\n",
      "2. Length: 444\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5970>]\n",
      "3.1. Score: 877.0\n",
      "3.2. Bits: 792.064\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 440\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 51\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 441\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "***\n",
      "Alignment Index: 46\n",
      "1. Title: gi|795109724|ref|XM_011975165.1| PREDICTED: Mandrillus leucophaeus hemoglobin subunit beta (LOC105535916), transcript variant X1, mRNA\n",
      "2. Length: 756\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5A00>, <Bio.Blast.Record.HSP object at 0x00000179D28B59A0>]\n",
      "3.1. Score: 876.0\n",
      "3.2. Bits: 791.162\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 468\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 141\n",
      "3.9. Subject Start: 269\n",
      "3.10. Alignment Length: 488\n",
      "GGATCTGTCCACTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTC\n",
      "|||||||||| |||||||||||||||||||||||||||||||||||||||\n",
      "GGATCTGTCCTCTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTC\n",
      "3.1. Score: 259.0\n",
      "3.2. Bits: 234.823\n",
      "3.3. E_value: 1.243e-56\n",
      "3.4. Identities: 137\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 5\n",
      "3.9. Subject Start: 3\n",
      "3.10. Alignment Length: 142\n",
      "TGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAG\n",
      "||||||||||||||||||||||| ||||||||||  ||||||||||||||\n",
      "TGCATCTGACTCCTGAGGAGAAGACTGCCGTTACCACCCTGTGGGGCAAG\n",
      "***\n",
      "Alignment Index: 47\n",
      "1. Title: gi|60653726|gb|AY892640.1| Synthetic construct Homo sapiens clone FLH028845.01L hemoglobin beta (HBB) mRNA, partial cds\n",
      "2. Length: 444\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5AC0>]\n",
      "3.1. Score: 874.0\n",
      "3.2. Bits: 789.358\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 440\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 51\n",
      "3.9. Subject Start: 1\n",
      "3.10. Alignment Length: 442\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "CAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGG\n",
      "***\n",
      "Alignment Index: 48\n",
      "1. Title: gi|823670799|gb|KR710229.1| Synthetic construct Homo sapiens clone CCSBHm_00010626 HBB (HBB) mRNA, encodes complete protein\n",
      "2. Length: 573\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5B50>]\n",
      "3.1. Score: 873.0\n",
      "3.2. Bits: 788.457\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 441\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 50\n",
      "3.9. Subject Start: 65\n",
      "3.10. Alignment Length: 444\n",
      "GCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "GCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTG\n",
      "***\n",
      "Alignment Index: 49\n",
      "1. Title: gi|823670797|gb|KR710228.1| Synthetic construct Homo sapiens clone CCSBHm_00010601 HBB (HBB) mRNA, encodes complete protein\n",
      "2. Length: 573\n",
      "3. HSP: [<Bio.Blast.Record.HSP object at 0x00000179D28B5BE0>]\n",
      "3.1. Score: 873.0\n",
      "3.2. Bits: 788.457\n",
      "3.3. E_value: 0.0\n",
      "3.4. Identities: 441\n",
      "3.5. Gaps: 0\n",
      "3.6. Strand: ('Plus', 'Plus')\n",
      "3.7. Frame: (1, 1)\n",
      "3.8. Query Start: 50\n",
      "3.9. Subject Start: 65\n",
      "3.10. Alignment Length: 444\n",
      "GCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "GCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTG\n"
     ]
    }
   ],
   "source": [
    "for ind, align in enumerate(record.alignments):\n",
    "    print('***')\n",
    "    print('Alignment Index:', ind)\n",
    "    print('1. Title:', align.title)\n",
    "    print('2. Length:', align.length)\n",
    "    print('3. HSP:', align.hsps)\n",
    "\n",
    "    for hsp in align.hsps:\n",
    "        print(\"3.1. Score:\", hsp.score)\n",
    "        print(\"3.2. Bits:\", hsp.bits) # degree of similarity\n",
    "        print(\"3.3. E_value:\", hsp.expect)\n",
    "        print(\"3.4. Identities:\", hsp.identities)\n",
    "        print(\"3.5. Gaps:\", hsp.gaps)\n",
    "        print(\"3.6. Strand:\", hsp.strand)\n",
    "        print(\"3.7. Frame:\", hsp.frame)\n",
    "        print(\"3.8. Query Start:\", hsp.query_start)\n",
    "        print(\"3.9. Subject Start:\", hsp.sbjct_start)\n",
    "        print(\"3.10. Alignment Length:\", hsp.align_length)\n",
    "\n",
    "        print(hsp.query[50:100]) # extract a specific piece of information\n",
    "        print(hsp.match[50:100])\n",
    "        print(hsp.sbjct[50:100])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Local BLAST\n",
    "\n",
    "Install [Blast+](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download) on the computer before use."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a local BLAST database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class NcbimakeblastdbCommandline in module Bio.Blast.Applications:\n",
      "\n",
      "class NcbimakeblastdbCommandline(Bio.Application.AbstractCommandline)\n",
      " |  NcbimakeblastdbCommandline(cmd='makeblastdb', **kwargs)\n",
      " |  \n",
      " |  Wrapper for the NCBI BLAST+ program makeblastdb.\n",
      " |  \n",
      " |  This is a wrapper for the NCBI BLAST+ makeblastdb application\n",
      " |  to create BLAST databases. By default, this creates a blast database\n",
      " |  with the same name as the input file. The default output location\n",
      " |  is the same directory as the input.\n",
      " |  \n",
      " |  >>> from Bio.Blast.Applications import NcbimakeblastdbCommandline\n",
      " |  >>> cline = NcbimakeblastdbCommandline(dbtype=\"prot\",\n",
      " |  ...                                    input_file=\"NC_005816.faa\")\n",
      " |  >>> cline\n",
      " |  NcbimakeblastdbCommandline(cmd='makeblastdb', dbtype='prot', input_file='NC_005816.faa')\n",
      " |  >>> print(cline)\n",
      " |  makeblastdb -dbtype prot -in NC_005816.faa\n",
      " |  \n",
      " |  You would typically run the command line with cline() or via the Python\n",
      " |  subprocess module, as described in the Biopython tutorial.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      NcbimakeblastdbCommandline\n",
      " |      Bio.Application.AbstractCommandline\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, cmd='makeblastdb', **kwargs)\n",
      " |      Initialize the class.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __call__(self, stdin=None, stdout=True, stderr=True, cwd=None, env=None)\n",
      " |      Execute command, wait for it to finish, return (stdout, stderr).\n",
      " |      \n",
      " |      Runs the command line tool and waits for it to finish. If it returns\n",
      " |      a non-zero error level, an exception is raised. Otherwise two strings\n",
      " |      are returned containing stdout and stderr.\n",
      " |      \n",
      " |      The optional stdin argument should be a string of data which will be\n",
      " |      passed to the tool as standard input.\n",
      " |      \n",
      " |      The optional stdout and stderr argument may be filenames (string),\n",
      " |      but otherwise are treated as a booleans, and control if the output\n",
      " |      should be captured as strings (True, default), or ignored by sending\n",
      " |      it to /dev/null to avoid wasting memory (False). If sent to a file\n",
      " |      or ignored, then empty string(s) are returned.\n",
      " |      \n",
      " |      The optional cwd argument is a string giving the working directory\n",
      " |      to run the command from. See Python's subprocess module documentation\n",
      " |      for more details.\n",
      " |      \n",
      " |      The optional env argument is a dictionary setting the environment\n",
      " |      variables to be used in the new process. By default the current\n",
      " |      process' environment variables are used. See Python's subprocess\n",
      " |      module documentation for more details.\n",
      " |      \n",
      " |      Default example usage::\n",
      " |      \n",
      " |          from Bio.Emboss.Applications import WaterCommandline\n",
      " |          water_cmd = WaterCommandline(gapopen=10, gapextend=0.5,\n",
      " |                                       stdout=True, auto=True,\n",
      " |                                       asequence=\"a.fasta\", bsequence=\"b.fasta\")\n",
      " |          print(\"About to run: %s\" % water_cmd)\n",
      " |          std_output, err_output = water_cmd()\n",
      " |      \n",
      " |      This functionality is similar to subprocess.check_output(). In general\n",
      " |      if you require more control over running the command, use subprocess\n",
      " |      directly.\n",
      " |      \n",
      " |      When the program called returns a non-zero error level, a custom\n",
      " |      ApplicationError exception is raised. This includes any stdout and\n",
      " |      stderr strings captured as attributes of the exception object, since\n",
      " |      they may be useful for diagnosing what went wrong.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a representation of the command line object for debugging.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> cline\n",
      " |      WaterCommandline(cmd='water', outfile='temp_water.txt', asequence='asis:ACCCGGGCGCGGT', bsequence='asis:ACCCGAGCGCGGT', gapopen=10, gapextend=0.5)\n",
      " |  \n",
      " |  __setattr__(self, name, value)\n",
      " |      Set attribute name to value (PRIVATE).\n",
      " |      \n",
      " |      This code implements a workaround for a user interface issue.\n",
      " |      Without this __setattr__ attribute-based assignment of parameters\n",
      " |      will silently accept invalid parameters, leading to known instances\n",
      " |      of the user assuming that parameters for the application are set,\n",
      " |      when they are not.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5, stdout=True)\n",
      " |      >>> cline.asequence = \"a.fasta\"\n",
      " |      >>> cline.bsequence = \"b.fasta\"\n",
      " |      >>> cline.csequence = \"c.fasta\"\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      ValueError: Option name csequence was not found.\n",
      " |      >>> print(cline)\n",
      " |      water -stdout -asequence=a.fasta -bsequence=b.fasta -gapopen=10 -gapextend=0.5\n",
      " |      \n",
      " |      This workaround uses a whitelist of object attributes, and sets the\n",
      " |      object attribute list as normal, for these.  Other attributes are\n",
      " |      assumed to be parameters, and passed to the self.set_parameter method\n",
      " |      for validation and assignment.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Make the commandline string with the currently set options.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> str(cline)\n",
      " |      'water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5'\n",
      " |  \n",
      " |  set_parameter(self, name, value=None)\n",
      " |      Set a commandline option for a program (OBSOLETE).\n",
      " |      \n",
      " |      Every parameter is available via a property and as a named\n",
      " |      keyword when creating the instance. Using either of these is\n",
      " |      preferred to this legacy set_parameter method which is now\n",
      " |      OBSOLETE, and likely to be DEPRECATED and later REMOVED in\n",
      " |      future releases.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  parameters = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast.Applications import *\n",
    "help(NcbimakeblastdbCommandline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "make_db = NcbimakeblastdbCommandline(\n",
    "    cmd='C:/blast-2.12.0+/bin/makeblastdb.exe',\n",
    "    input_file='./resources/HBB.fasta',\n",
    "    dbtype='nucl', # data type\n",
    "    out='HBBdb', # name of the database\n",
    ")\n",
    "\n",
    "std_out, std_err = make_db()\n",
    "# create mulitple files (.ndb, .nhr etc.) with same root name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BLAST an unknown sequence to the sequences from the local database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class NcbiblastnCommandline in module Bio.Blast.Applications:\n",
      "\n",
      "class NcbiblastnCommandline(_NcbiblastMain2SeqCommandline)\n",
      " |  NcbiblastnCommandline(cmd='blastn', **kwargs)\n",
      " |  \n",
      " |  Wrapper for the NCBI BLAST+ program blastn (for nucleotides).\n",
      " |  \n",
      " |  With the release of BLAST+ (BLAST rewritten in C++ instead of C), the NCBI\n",
      " |  replaced the old blastall tool with separate tools for each of the searches.\n",
      " |  This wrapper therefore replaces BlastallCommandline with option -p blastn.\n",
      " |  \n",
      " |  For example, to run a search against the \"nt\" nucleotide database using the\n",
      " |  FASTA nucleotide file \"m_code.fasta\" as the query, with an expectation value\n",
      " |  cut off of 0.001, saving the output to a file in XML format:\n",
      " |  \n",
      " |  >>> from Bio.Blast.Applications import NcbiblastnCommandline\n",
      " |  >>> cline = NcbiblastnCommandline(query=\"m_cold.fasta\", db=\"nt\", strand=\"plus\",\n",
      " |  ...                               evalue=0.001, out=\"m_cold.xml\", outfmt=5)\n",
      " |  >>> cline\n",
      " |  NcbiblastnCommandline(cmd='blastn', out='m_cold.xml', outfmt=5, query='m_cold.fasta', db='nt', evalue=0.001, strand='plus')\n",
      " |  >>> print(cline)\n",
      " |  blastn -out m_cold.xml -outfmt 5 -query m_cold.fasta -db nt -evalue 0.001 -strand plus\n",
      " |  \n",
      " |  You would typically run the command line with cline() or via the Python\n",
      " |  subprocess module, as described in the Biopython tutorial.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      NcbiblastnCommandline\n",
      " |      _NcbiblastMain2SeqCommandline\n",
      " |      _Ncbiblast2SeqCommandline\n",
      " |      _NcbiblastCommandline\n",
      " |      _NcbibaseblastCommandline\n",
      " |      Bio.Application.AbstractCommandline\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, cmd='blastn', **kwargs)\n",
      " |      Initialize the class.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __call__(self, stdin=None, stdout=True, stderr=True, cwd=None, env=None)\n",
      " |      Execute command, wait for it to finish, return (stdout, stderr).\n",
      " |      \n",
      " |      Runs the command line tool and waits for it to finish. If it returns\n",
      " |      a non-zero error level, an exception is raised. Otherwise two strings\n",
      " |      are returned containing stdout and stderr.\n",
      " |      \n",
      " |      The optional stdin argument should be a string of data which will be\n",
      " |      passed to the tool as standard input.\n",
      " |      \n",
      " |      The optional stdout and stderr argument may be filenames (string),\n",
      " |      but otherwise are treated as a booleans, and control if the output\n",
      " |      should be captured as strings (True, default), or ignored by sending\n",
      " |      it to /dev/null to avoid wasting memory (False). If sent to a file\n",
      " |      or ignored, then empty string(s) are returned.\n",
      " |      \n",
      " |      The optional cwd argument is a string giving the working directory\n",
      " |      to run the command from. See Python's subprocess module documentation\n",
      " |      for more details.\n",
      " |      \n",
      " |      The optional env argument is a dictionary setting the environment\n",
      " |      variables to be used in the new process. By default the current\n",
      " |      process' environment variables are used. See Python's subprocess\n",
      " |      module documentation for more details.\n",
      " |      \n",
      " |      Default example usage::\n",
      " |      \n",
      " |          from Bio.Emboss.Applications import WaterCommandline\n",
      " |          water_cmd = WaterCommandline(gapopen=10, gapextend=0.5,\n",
      " |                                       stdout=True, auto=True,\n",
      " |                                       asequence=\"a.fasta\", bsequence=\"b.fasta\")\n",
      " |          print(\"About to run: %s\" % water_cmd)\n",
      " |          std_output, err_output = water_cmd()\n",
      " |      \n",
      " |      This functionality is similar to subprocess.check_output(). In general\n",
      " |      if you require more control over running the command, use subprocess\n",
      " |      directly.\n",
      " |      \n",
      " |      When the program called returns a non-zero error level, a custom\n",
      " |      ApplicationError exception is raised. This includes any stdout and\n",
      " |      stderr strings captured as attributes of the exception object, since\n",
      " |      they may be useful for diagnosing what went wrong.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a representation of the command line object for debugging.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> cline\n",
      " |      WaterCommandline(cmd='water', outfile='temp_water.txt', asequence='asis:ACCCGGGCGCGGT', bsequence='asis:ACCCGAGCGCGGT', gapopen=10, gapextend=0.5)\n",
      " |  \n",
      " |  __setattr__(self, name, value)\n",
      " |      Set attribute name to value (PRIVATE).\n",
      " |      \n",
      " |      This code implements a workaround for a user interface issue.\n",
      " |      Without this __setattr__ attribute-based assignment of parameters\n",
      " |      will silently accept invalid parameters, leading to known instances\n",
      " |      of the user assuming that parameters for the application are set,\n",
      " |      when they are not.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5, stdout=True)\n",
      " |      >>> cline.asequence = \"a.fasta\"\n",
      " |      >>> cline.bsequence = \"b.fasta\"\n",
      " |      >>> cline.csequence = \"c.fasta\"\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      ValueError: Option name csequence was not found.\n",
      " |      >>> print(cline)\n",
      " |      water -stdout -asequence=a.fasta -bsequence=b.fasta -gapopen=10 -gapextend=0.5\n",
      " |      \n",
      " |      This workaround uses a whitelist of object attributes, and sets the\n",
      " |      object attribute list as normal, for these.  Other attributes are\n",
      " |      assumed to be parameters, and passed to the self.set_parameter method\n",
      " |      for validation and assignment.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Make the commandline string with the currently set options.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> str(cline)\n",
      " |      'water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5'\n",
      " |  \n",
      " |  set_parameter(self, name, value=None)\n",
      " |      Set a commandline option for a program (OBSOLETE).\n",
      " |      \n",
      " |      Every parameter is available via a property and as a named\n",
      " |      keyword when creating the instance. Using either of these is\n",
      " |      preferred to this legacy set_parameter method which is now\n",
      " |      OBSOLETE, and likely to be DEPRECATED and later REMOVED in\n",
      " |      future releases.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  parameters = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(NcbiblastnCommandline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "make_blastn = NcbiblastnCommandline(\n",
    "    cmd='C:/blast-2.12.0+/bin/blastn.exe',\n",
    "    out='unknown_HBB.xml', # same name as input file\n",
    "    outfmt=5, # XML format\n",
    "    query='./resources/unknownHBB.fasta', # unknown DNA sequence\n",
    "    db='HBBdb', # local database\n",
    "    evalue=0.001,\n",
    "    strand='plus',\n",
    ")\n",
    "\n",
    "std_out, std_err = make_blastn()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read the alignment output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "record = NCBIXML.read(open('./unknown_HBB.xml'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***\n",
      "Alignment Index: 0\n",
      "Title: gnl|BL_ORD_ID|16 NM_001164018.1 Equus caballus hemoglobin, beta (HBB), mRNA\n",
      "Score: 444.0\n",
      "E_value: 0.0\n",
      "***\n",
      "Alignment Index: 1\n",
      "Title: gnl|BL_ORD_ID|19 NM_001304883.1 Ailuropoda melanoleuca hemoglobin subunit beta-like (HBB), mRNA\n",
      "Score: 291.0\n",
      "E_value: 5.08184e-156\n",
      "***\n",
      "Alignment Index: 2\n",
      "Title: gnl|BL_ORD_ID|23 NM_001304885.1 Ailuropoda melanoleuca hemoglobin subunit beta (HBB), mRNA\n",
      "Score: 282.0\n",
      "E_value: 5.11792e-151\n",
      "***\n",
      "Alignment Index: 3\n",
      "Title: gnl|BL_ORD_ID|2 NM_000518.5 Homo sapiens hemoglobin subunit beta (HBB), mRNA\n",
      "Score: 268.0\n",
      "E_value: 3.10206e-143\n",
      "***\n",
      "Alignment Index: 4\n",
      "Title: gnl|BL_ORD_ID|14 NM_001082260.3 Oryctolagus cuniculus hemoglobin, beta (HBB2), transcript variant 1, mRNA\n",
      "Score: 261.0\n",
      "E_value: 2.41507e-139\n",
      "***\n",
      "Alignment Index: 5\n",
      "Title: gnl|BL_ORD_ID|13 NM_001314043.1 Oryctolagus cuniculus hemoglobin, beta (HBB2), transcript variant 1, mRNA\n",
      "Score: 261.0\n",
      "E_value: 2.41507e-139\n",
      "***\n",
      "Alignment Index: 6\n",
      "Title: gnl|BL_ORD_ID|11 NM_173917.2 Bos taurus hemoglobin, beta (HBB), mRNA\n",
      "Score: 260.0\n",
      "E_value: 8.68612e-139\n",
      "***\n",
      "Alignment Index: 7\n",
      "Title: gnl|BL_ORD_ID|9 NM_001144841.1 Sus scrofa hemoglobin, beta (HBB), mRNA\n",
      "Score: 259.0\n",
      "E_value: 3.12409e-138\n",
      "***\n",
      "Alignment Index: 8\n",
      "Title: gnl|BL_ORD_ID|18 NM_001329918.1 Chlorocebus sabaeus hemoglobin subunit beta (HBB), mRNA\n",
      "Score: 253.0\n",
      "E_value: 6.76246e-135\n",
      "***\n",
      "Alignment Index: 9\n",
      "Title: gnl|BL_ORD_ID|10 NM_001097648.1 Ovis aries hemoglobin, beta (HBB), mRNA\n",
      "Score: 253.0\n",
      "E_value: 6.76246e-135\n",
      "***\n",
      "Alignment Index: 10\n",
      "Title: gnl|BL_ORD_ID|20 NM_001279263.1 Condylura cristata hemoglobin, beta (HBB), mRNA\n",
      "Score: 204.0\n",
      "E_value: 1.17234e-107\n",
      "***\n",
      "Alignment Index: 11\n",
      "Title: gnl|BL_ORD_ID|1 NM_033234.1 Rattus norvegicus hemoglobin subunit beta (Hbb), mRNA\n",
      "Score: 190.0\n",
      "E_value: 7.10578e-100\n",
      "***\n",
      "Alignment Index: 12\n",
      "Title: gnl|BL_ORD_ID|7 NM_001246752.1 Cricetulus griseus beta major globin chain (Hbb), mRNA\n",
      "Score: 181.0\n",
      "E_value: 7.15623e-95\n"
     ]
    }
   ],
   "source": [
    "for ind, align in enumerate(record.descriptions):\n",
    "    print('***')\n",
    "    print('Alignment Index:', ind)\n",
    "    print('Title:', align.title)\n",
    "    print('Score:', align.score)\n",
    "    print('E_value:', align.e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***\n",
      "Title: gnl|BL_ORD_ID|16 NM_001164018.1 Equus caballus hemoglobin, beta (HBB), mRNA\n",
      "Length: 444\n",
      "Score: 444.0\n",
      "Bits: 821.033\n",
      "E_value: 0.0\n",
      "Identities: 444\n",
      "Gaps: 0\n",
      "Alignment Length: 444\n",
      "Percent identity = 100.00\n",
      "CAAGGTGAATGAGGAAGAAGTTGGTGGTGAAGCCCTGGGCAGGCTGCTGG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "CAAGGTGAATGAGGAAGAAGTTGGTGGTGAAGCCCTGGGCAGGCTGCTGG\n"
     ]
    }
   ],
   "source": [
    "align = record.alignments[0] # we look at the top alignment only\n",
    "\n",
    "print('***')\n",
    "print('Title:', align.title)\n",
    "print('Length:', align.length) # same score as length = 100% identity\n",
    "\n",
    "for hsp in align.hsps:\n",
    "    print(\"Score:\", hsp.score)\n",
    "    print(\"Bits:\", hsp.bits) # degree of similarity\n",
    "    print(\"E_value:\", hsp.expect)\n",
    "    print(\"Identities:\", hsp.identities)\n",
    "    print(\"Gaps:\", hsp.gaps)\n",
    "    print(\"Alignment Length:\", hsp.align_length)\n",
    "    # we must compute %identity ourselves\n",
    "    print(f\"Percent identity = {100 * hsp.identities / hsp.align_length:.2f}%\")\n",
    "\n",
    "    print(hsp.query[50:100]) # extract a specific piece of information\n",
    "    print(hsp.match[50:100])\n",
    "    print(hsp.sbjct[50:100])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "b51595f36012c8da2bf2284ebbd8d377d5c2d31fa7484164f2d2bcda7a09b462"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
