{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence input/ouput\n",
    "\n",
    "## Writing and converting sequence files\n",
    "\n",
    "Conversion of files from FASTA (multi-lines with 50 bases per line) to [Fasta-2line](https://codechalleng.es/bites/298/) using [SeqIO](https://biopython.org/wiki/SeqIO)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "from Bio import Entrez, SeqIO\n",
    "Entrez.email = 'sbwiecko@free.fr'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Item count after search = 7\n",
      "Item count after filtering = 1\n"
     ]
    }
   ],
   "source": [
    "handle = Entrez.esearch(\n",
    "    db='nucleotide',\n",
    "    term='HBB[Gene Name] AND RefSeq[Keyword] AND Homo sapiens[Organism]',\n",
    "    idtype='acc',\n",
    ")\n",
    "\n",
    "record = Entrez.read(handle)\n",
    "handle.close()\n",
    "\n",
    "fetch_list = []\n",
    "\n",
    "for ID in record['IdList']:\n",
    "    if 'NM_' in ID:\n",
    "        fetch = Entrez.efetch(\n",
    "            db='nucleotide',\n",
    "            id=ID,\n",
    "            rettype='fasta',\n",
    "            retmode='text',\n",
    "        )\n",
    "\n",
    "        read_fetch = fetch.read() # same as Entrez.read\n",
    "        fetch_list.append(read_fetch)\n",
    "\n",
    "print(f\"Item count after search = {record['Count']}\")\n",
    "print(f\"Item count after filtering = {len(fetch_list)}\")\n",
    "\n",
    "for file in fetch_list:\n",
    "    with open('HBB-human.fasta', 'a+') as saved_file:\n",
    "        saved_file.write(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('HBB-human.fasta2lines', 'w+') as converted_file:\n",
    "    SeqIO.convert(\n",
    "        'HBB-human.fasta', # original file\n",
    "        'fasta',           # format of the original file\n",
    "        converted_file,    # destination file\n",
    "        'fasta-2line',     # format converted\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Conversion from GenBank to fasta-2line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('HBB.fasta2lines', 'w+') as converted_file:\n",
    "    SeqIO.convert(\n",
    "        './resources/HBB.gb',\n",
    "        'gb',\n",
    "        converted_file,\n",
    "        'fasta-2line',\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "table = pd.read_html(\"https://biopython.org/wiki/SeqIO\")\n",
    "#print(table[0].set_index('Format name', drop=True).to_markdown())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File formats\n",
    "\n",
    "| Format name           |   Read | Write       | Index   | Notes                                                                                                                                                                                                                                                                                                                                                                                       |\n",
    "|:----------------------|-------:|:------------|:--------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    "| abi                   |   1.58 | No          | nan     | Reads the ABI \"Sanger\"capillary sequence traces files, including the PHRED quality scores for the base calls. This allows ABI to FASTQ conversion. Note each ABI file contains one and only one sequence (so there is no point in indexing the file).                                                                                                                                      |\n",
    "| abi-trim              |   1.71 | No          | nan     | Same as \"abi\"but with quality trimming with Mott\"s algorithm.                                                                                                                                                                                                                                                                                                                              |\n",
    "| ace                   |   1.47 | No          | 1.52    | Reads the contig sequences from an ACE assembly file. Uses Bio.Sequencing.Ace internally                                                                                                                                                                                                                                                                                                    |\n",
    "| cif-atom              |   1.73 | No          | No      | Uses Bio.PDB.MMCIFParser to determine the (partial) protein sequence as it appears in the structure based on the atomic coordinates.                                                                                                                                                                                                                                                        |\n",
    "| cif-seqres            |   1.73 | No          | No      | Reads a macromolecular Crystallographic Information File (mmCIF) file to determine the complete protein sequence as defined by the _pdbx_poly_seq_scheme records.                                                                                                                                                                                                                           |\n",
    "| clustal               |   1.43 | 1.43        | No      | The alignment format of Clustal X and Clustal W.                                                                                                                                                                                                                                                                                                                                            |\n",
    "| embl                  |   1.43 | 1.54        | 1.52    | The EMBL flat file format. Uses Bio.GenBank internally.                                                                                                                                                                                                                                                                                                                                     |\n",
    "| fasta                 |   1.43 | 1.43        | 1.52    | This refers to the input FASTA file format introduced for Bill Pearson\"s FASTA tool, where each record starts with a \">\"line.                                                                                                                                                                                                                                                              |\n",
    "| fasta-2line           |   1.71 | 1.71        | No      | FASTA format variant with no line wrapping and exactly two lines per record.                                                                                                                                                                                                                                                                                                                |\n",
    "| fastq-sanger or fastq |   1.5  | 1.50        | 1.52    | FASTQ files are a bit like FASTA files but also include sequencing qualities. In Biopython, \"fastq\"(or the alias \"fastq-sanger‚Äù) refers to Sanger style FASTQ files which encode PHRED qualities using an ASCII offset of 33. See also the incompatible \"fastq-solexa\"and \"fastq-illumina\"variants used in early Solexa/Illumina pipelines, Illumina pipeline 1.8 produces Sanger FASTQ. |\n",
    "| fastq-solexa          |   1.5  | 1.50        | 1.52    | In Biopython, \"fastq-solexa\"refers to the original Solexa/Illumina style FASTQ files which encode Solexa qualities using an ASCII offset of 64. See also what we call the \"fastq-illumina\"format.                                                                                                                                                                                         |\n",
    "| fastq-illumina        |   1.51 | 1.51        | 1.52    | In Biopython, \"fastq-illumina\"refers to early Solexa/Illumina style FASTQ files (from pipeline version 1.3 to 1.7) which encode PHRED qualities using an ASCII offset of 64. For good quality reads, PHRED and Solexa scores are approximately equal, so the \"fastq-solexa\"and \"fastq-illumina\"variants are almost equivalent.                                                           |\n",
    "| gck                   |   1.75 | No          | No      | The native format used by Gene Construction Kit.                                                                                                                                                                                                                                                                                                                                            |\n",
    "| genbank or gb         |   1.43 | 1.48 / 1.51 | 1.52    | The GenBank or GenPept flat file format. Uses Bio.GenBank internally for parsing. Biopython 1.48 to 1.50 wrote basic GenBank files with only minimal annotation, while 1.51 onwards will also write the features table.                                                                                                                                                                     |\n",
    "| ig                    |   1.47 | No          | 1.52    | This refers to the IntelliGenetics file format, apparently the same as the MASE alignment format.                                                                                                                                                                                                                                                                                           |\n",
    "| imgt                  |   1.56 | 1.56        | 1.56    | This refers to the IMGT variant of the EMBL plain text file format.                                                                                                                                                                                                                                                                                                                         |\n",
    "| nexus                 |   1.43 | 1.48        | No      | The NEXUS multiple alignment format, also known as PAUP format. Uses Bio.Nexus internally.                                                                                                                                                                                                                                                                                                  |\n",
    "| pdb-seqres            |   1.61 | No          | No      | Reads a Protein Data Bank (PDB) file to determine the complete protein sequence as it appears in the header (no dependency on Bio.PDB and NumPy).                                                                                                                                                                                                                                           |\n",
    "| pdb-atom              |   1.61 | No          | No      | Uses Bio.PDB to determine the (partial) protein sequence as it appears in the structure based on the atom coordinate section of the file (requires NumPy).                                                                                                                                                                                                                                  |\n",
    "| phd                   |   1.46 | 1.52        | 1.52    | PHD files are output from PHRED, used by PHRAP and CONSED for input. Uses Bio.Sequencing.Phd internally.                                                                                                                                                                                                                                                                                    |\n",
    "| phylip                |   1.43 | 1.43        | No      | PHYLIP files. Truncates names at 10 characters.                                                                                                                                                                                                                                                                                                                                             |\n",
    "| pir                   |   1.48 | 1.71        | 1.52    | A \"FASTA like\"format introduced by the National Biomedical Research Foundation (NBRF) for the Protein Information Resource (PIR) database, now part of UniProt.                                                                                                                                                                                                                            |\n",
    "| seqxml                |   1.58 | 1.58        | No      | Simple sequence XML file format.                                                                                                                                                                                                                                                                                                                                                            |\n",
    "| sff                   |   1.54 | 1.54        | 1.54    | Standard Flowgram Format (SFF) binary files produced by Roche 454 and IonTorrent/IonProton sequencing machines.                                                                                                                                                                                                                                                                             |\n",
    "| sff-trim              |   1.54 | No          | 1.54    | Standard Flowgram Format applying the trimming listed in the file.                                                                                                                                                                                                                                                                                                                          |\n",
    "| snapgene              |   1.75 | No          | No      | The native format used by SnapGene.                                                                                                                                                                                                                                                                                                                                                         |\n",
    "| stockholm             |   1.43 | 1.43        | No      | The Stockholm alignment format is also known as PFAM format.                                                                                                                                                                                                                                                                                                                                |\n",
    "| swiss                 |   1.43 | No          | 1.52    | Swiss-Prot aka UniProt format. Uses Bio.SwissProt internally. See also the UniProt XML format.                                                                                                                                                                                                                                                                                              |\n",
    "| tab                   |   1.48 | 1.48        | 1.52    | Simple two column tab separated sequence files, where each line holds a record\"s identifier and sequence. For example, this is used by Aligent\"s eArray software when saving microarray probes in a minimal tab delimited text file.                                                                                                                                                        |\n",
    "| qual                  |   1.5  | 1.50        | 1.52    | Qual files are a bit like FASTA files but instead of the sequence, record space separated integer sequencing values as PHRED quality scores. A matched pair of FASTA and QUAL files are often used as an alternative to a single FASTQ file.                                                                                                                                                |\n",
    "| uniprot-xml           |   1.56 | No          | 1.56    | UniProt XML format, successor to the plain text Swiss-Prot format.                                                                                                                                                                                                                                                                                                                          |\n",
    "| xdna                  |   1.75 | 1.75        | No      | The native format used by Christian Marck\"s DNA Strider and Serial Cloner.                                                                                                                                                                                                                                                                                                                  |\n",
    "| nan                   | nan    | nan         | nan     | nan       "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quiz\n",
    "\n",
    "Convert the attached sample.ab1 file (see the trace using [Chromas](http://technelysium.com.au/wp/chromas/) to fastq file and then answer the following:\n",
    "\n",
    "1. What is the name of the header?\n",
    "2. Does the third line contain any letters or signs?\n",
    "3. Does quality vary between nucleotides?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@M2927_AU-NFP-NFP5\n",
      "CTCCGGCATTAAACAANCAAGATTNTTGTNCTGACGTCTCCCCTTTATATNTATCTCTTCTTTGGACATCAAATATCTTTTGGTTTGATATTAAAAATATATGCTGATGTCTCTGTCATTAGTTCTACTATTTTTCTTTTTCAAATCTTAGACACTGTTGGCAAATCCCAAACAAGTATGAAATCATGCGCAGAATATAATAAAGATGGTAAGAATAATATGGTATTATAACCCAATTTGGTGTCCATACACGATGTCAACTAGCTGGCTGGTACCTAAAGACCATCTAACTTGATGCCTTTACCAAATTTTTAAGGGGCAAACTAGCAAAAGATATGGTTTAACGTTTAACAAAGGCTAATGCCATTATATATATGCAATTATAATATAATACAGGCAATGAACAGCTTTCAATGTNGACACTTCCAGTTACATTGGTTTCAATATATTAAAACACCTTGACAAACTCNCCATTTNTATTATTTGTNCCCTATTCCTAGTNCATAGCTAATTCGGATAGAAACNGTGTTNACATCGNCANGTAAGCNTGACNTCTAATATATTNAAANANTANGATGCCCTTCTNACCNGCTCGCCTNGGCTTGNGCGTAGCTTTNGGTNCNNCATANNCANCTANCGCGNNGCNNGNCCNGNNCGATTGANTGAGNATGGCNGGGGGCNCTGCNCNCCGCTNGCNNGACCCGCNGNNGCCGAGGCCGACCNNCGGCGANACGANNNTGACAGNCGGTGGCGCTCCGCAGGNNNCGNNNGNGGNCGCTNCNGGGCNCGGGNGNNCCNGNNCCNGCGNNCNNACCCCNCAGGGCGCGGGGGAGGCCGGNTGGCCCCNGGNNNNCNNCGNNNNNCAGGNNGNCGNGGCNGGNNGGNGCCCCCCGGCCNCNNTGGGTNGGCNANGNGCNNNTTGGGCCCNNGNGCNTCGNTGCGNGCNGCNCTGCAGGGTGGNCNNTNGNGNGGGNGCGNGGNGGNCNANGGNCCGCGGNNCNGNGCGNCNCTGGGNNNNGGCNNCGAGCCNGGNCCGGNGNGGGGGNCNCNNGCCGCANTCGCCGCNNGCGANGTCCCGGTNNCGCNNNGGNGNGGNGGNNTCNGNCCCNCGGCGGNGCGNCGCCGNGCGNNGGNGCNNGGCNGTCGGCCGGGCCGGNCGNNCGCGCNCGCACGGCNGGGCNGCCNCCCGCGNGNGGCGNGCGGGCCNGCNNGGCGCCCCNNGGNGNNNGCNTCGGCAGGCAGGGGCCCGCNGGGNCGCGCCGNAGNCGCGCGCCGCGCNANNCNCNNAGCCGAGCGACCNNNGNGGCNGCGNNGCNNGNGCCGCCGGNGGNNNNGCGGGCGNNAACNNCCCGCACNGCGGCCGGCAGCGGNCGNTGCCGNGNAGAGCACGNGTCGCGTGCNCGTGCCNGNGCGNCGCTNCCAGNCNGCNCNGCCGCNNGTGCGNCTCCGNC\n",
      "+\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open('sample.fastq', 'w+') as converted_file:\n",
    "    SeqIO.convert(\n",
    "        './resources/sample.abi',\n",
    "        'abi',\n",
    "        converted_file,\n",
    "        'fastq',\n",
    "    )\n",
    "\n",
    "f = open('./resources/sample.fastq', 'r')\n",
    "print(f.read())\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing FASTA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NM_000518.5\n",
      "Name: NM_000518.5\n",
      "Description: NM_000518.5 Homo sapiens hemoglobin subunit beta (HBB), mRNA\n",
      "Number of features: 0\n",
      "Seq('ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGG...CAA')\n",
      "---------------\n",
      "NM_000518.5\n",
      "NM_000518.5\n",
      "NM_000518.5 Homo sapiens hemoglobin subunit beta (HBB), mRNA\n",
      "[]\n",
      "ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTCGGTGCCTTTAGTGATGGCCTGGCTCACCTGGACAACCTCAAGGGCACCTTTGCCACACTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGTGCTGGTCTGTGTGCTGGCCCATCACTTTGGCAAAGAATTCACCCCACCAGTGCAGGCTGCCTATCAGAAAGTGGTGGCTGGTGTGGCTAATGCCCTGGCCCACAAGTATCACTAAGCTCGCTTTCTTGCTGTCCAATTTCTATTAAAGGTTCCTTTGTTCCCTAAGTCCAACTACTAAACTGGGGGATATTATGAAGGGCCTTGAGCATCTGGATTCTGCCTAATAAAAAACATTTATTTTCATTGCAA\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.parse( # parse all the sequences even if only 1 sequence\n",
    "    'HBB-human.fasta',\n",
    "    'fasta',\n",
    ")\n",
    "\n",
    "for element in record:\n",
    "    print(element)\n",
    "    print('-'*15)\n",
    "    print(element.id) # parse has created an object with multiple attributes\n",
    "    print(element.name)\n",
    "    print(element.description)\n",
    "    print(element.features)\n",
    "    print(element.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------------\n",
      "NM_001300984.2\n",
      "3071\n",
      "---------------\n",
      "NM_000275.3\n",
      "3143\n",
      "---------------\n",
      "NM_001182895.3\n",
      "594\n",
      "---------------\n",
      "NM_214094.2\n",
      "3267\n",
      "---------------\n",
      "NM_021879.3\n",
      "3133\n",
      "---------------\n",
      "NM_001271493.1\n",
      "2752\n",
      "---------------\n",
      "NM_001131493.2\n",
      "3181\n",
      "---------------\n",
      "NM_001104792.1\n",
      "3424\n",
      "---------------\n",
      "NM_001320209.1\n",
      "2673\n",
      "---------------\n",
      "NM_001022940.2\n",
      "2486\n"
     ]
    }
   ],
   "source": [
    "# same for FASTA files that contain multiple sequences\n",
    "record = SeqIO.parse(\n",
    "    'OCA2.fasta',\n",
    "    'fasta',\n",
    ")\n",
    "\n",
    "for element in record:\n",
    "    print('-'*15)\n",
    "    print(element.id)\n",
    "    print(len(element.seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quiz\n",
    "\n",
    "What is the length of the sequence in the converted file (fastq) in the third quiz?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of the sequence = 1451\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.parse(\n",
    "    handle='sample.fastq',\n",
    "    format='fastq',\n",
    "\n",
    ")\n",
    "\n",
    "for element in record:\n",
    "    print(f\"Length of the sequence = {len(element.seq)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing GenBank files\n",
    "\n",
    "More information are contained in a GenBank file, therefore the elements in the record will have more attributes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NM_000518.5\n",
      "Name: NM_000518\n",
      "Description: Homo sapiens hemoglobin subunit beta (HBB), mRNA\n",
      "Number of features: 48\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=PRI\n",
      "/date=20-APR-2020\n",
      "/accessions=['NM_000518']\n",
      "/sequence_version=5\n",
      "/keywords=['RefSeq', 'MANE Select']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='[Analysis of beta-globin gene variants in Liuzhou area of Guangxi]', ...), Reference(title='Genetic polymorphisms of HbE/beta thalassemia related to clinical presentation: implications for clinical diversity', ...), Reference(title='Use of >100,000 NHLBI Trans-Omics for Precision Medicine (TOPMed) Consortium whole genome sequences improves imputation quality and detection of rare variant associations in admixed African and Hispanic/Latino populations', ...), Reference(title='Effect of N(Epsilon)-(carboxymethyl)lysine on Laboratory Parameters and Its Association with beta (S) Haplotype in Children with Sickle Cell Anemia', ...), Reference(title='Upstream open reading frames cause widespread reduction of protein expression and are polymorphic among humans', ...), Reference(title='Beta-Thalassemia', ...), Reference(title='Sickle Cell Disease', ...), Reference(title='The structure of hemoglobin Indianapolis [beta112(G14) arginine]. An unstable variant detectable only by isotopic labeling', ...), Reference(title='Human beta-globin messenger RNA. III. Nucleotide sequences derived from complementary DNA', ...), Reference(title=\"Complete 3' noncoding region sequences of rabbit and human beta-globin messenger RNAs\", ...)]\n",
      "/comment=REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "reference sequence was derived from AK311825.1 and BU661647.1.\n",
      "This sequence is a reference standard in the RefSeqGene project.\n",
      "On Jun 13, 2018 this sequence version replaced NM_000518.4.\n",
      "Summary: The alpha (HBA) and beta (HBB) loci determine the\n",
      "structure of the 2 types of polypeptide chains in adult hemoglobin,\n",
      "Hb A. The normal adult hemoglobin tetramer consists of two alpha\n",
      "chains and two beta chains. Mutant beta globin causes sickle cell\n",
      "anemia. Absence of beta chain causes beta-zero-thalassemia. Reduced\n",
      "amounts of detectable beta globin causes beta-plus-thalassemia. The\n",
      "order of the genes in the beta-globin cluster is 5'-epsilon --\n",
      "gamma-G -- gamma-A -- delta -- beta--3'. [provided by RefSeq, Jul\n",
      "2008].\n",
      "Publication Note:  This RefSeq record includes a subset of the\n",
      "publications that are available for this gene. Please see the Gene\n",
      "record to access additional publications.\n",
      "COMPLETENESS: full length.\n",
      "/structured_comment=OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'AU139015.1, SRR5189652.246572.1 [ECO:0000332]')])), ('RefSeq-Attributes', OrderedDict([('MANE Ensembl match', 'ENST00000335295.4/ ENSP00000333994.3'), ('RefSeq Select criteria', 'based on single protein-coding transcript')]))])\n",
      "Seq('ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGG...CAA')\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.parse(\n",
    "    handle = './resources/HBB-human.gb',\n",
    "    format='gb',\n",
    ")\n",
    "\n",
    "for element in record:\n",
    "    print(element)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "molecule_type : mRNA\n",
      "topology : linear\n",
      "data_file_division : PRI\n",
      "date : 20-APR-2020\n",
      "accessions : ['NM_000518']\n",
      "sequence_version : 5\n",
      "keywords : ['RefSeq', 'MANE Select']\n",
      "source : Homo sapiens (human)\n",
      "organism : Homo sapiens\n",
      "taxonomy : ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "references : [Reference(title='[Analysis of beta-globin gene variants in Liuzhou area of Guangxi]', ...), Reference(title='Genetic polymorphisms of HbE/beta thalassemia related to clinical presentation: implications for clinical diversity', ...), Reference(title='Use of >100,000 NHLBI Trans-Omics for Precision Medicine (TOPMed) Consortium whole genome sequences improves imputation quality and detection of rare variant associations in admixed African and Hispanic/Latino populations', ...), Reference(title='Effect of N(Epsilon)-(carboxymethyl)lysine on Laboratory Parameters and Its Association with beta (S) Haplotype in Children with Sickle Cell Anemia', ...), Reference(title='Upstream open reading frames cause widespread reduction of protein expression and are polymorphic among humans', ...), Reference(title='Beta-Thalassemia', ...), Reference(title='Sickle Cell Disease', ...), Reference(title='The structure of hemoglobin Indianapolis [beta112(G14) arginine]. An unstable variant detectable only by isotopic labeling', ...), Reference(title='Human beta-globin messenger RNA. III. Nucleotide sequences derived from complementary DNA', ...), Reference(title=\"Complete 3' noncoding region sequences of rabbit and human beta-globin messenger RNAs\", ...)]\n",
      "comment : REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "reference sequence was derived from AK311825.1 and BU661647.1.\n",
      "This sequence is a reference standard in the RefSeqGene project.\n",
      "On Jun 13, 2018 this sequence version replaced NM_000518.4.\n",
      "Summary: The alpha (HBA) and beta (HBB) loci determine the\n",
      "structure of the 2 types of polypeptide chains in adult hemoglobin,\n",
      "Hb A. The normal adult hemoglobin tetramer consists of two alpha\n",
      "chains and two beta chains. Mutant beta globin causes sickle cell\n",
      "anemia. Absence of beta chain causes beta-zero-thalassemia. Reduced\n",
      "amounts of detectable beta globin causes beta-plus-thalassemia. The\n",
      "order of the genes in the beta-globin cluster is 5'-epsilon --\n",
      "gamma-G -- gamma-A -- delta -- beta--3'. [provided by RefSeq, Jul\n",
      "2008].\n",
      "Publication Note:  This RefSeq record includes a subset of the\n",
      "publications that are available for this gene. Please see the Gene\n",
      "record to access additional publications.\n",
      "COMPLETENESS: full length.\n",
      "structured_comment : OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'AU139015.1, SRR5189652.246572.1 [ECO:0000332]')])), ('RefSeq-Attributes', OrderedDict([('MANE Ensembl match', 'ENST00000335295.4/ ENSP00000333994.3'), ('RefSeq Select criteria', 'based on single protein-coding transcript')]))])\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.parse(\n",
    "    handle = './resources/HBB-human.gb',\n",
    "    format='gb',\n",
    ")\n",
    "\n",
    "for element in record:\n",
    "    for key in element.annotations.keys():\n",
    "        print(key, ':', element.annotations[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "source 0 628\n",
      "gene 0 628\n",
      "exon 0 142\n",
      "misc_feature 29 32\n",
      "CDS 50 494\n",
      "misc_feature 53 56\n",
      "misc_feature 53 56\n",
      "misc_feature 53 56\n",
      "misc_feature 71 77\n",
      "misc_feature 74 77\n",
      "misc_feature 77 80\n",
      "misc_feature 86 89\n",
      "misc_feature 101 104\n",
      "misc_feature 125 131\n",
      "misc_feature 137 143\n",
      "mat_peptide 146 176\n",
      "mat_peptide 146 167\n",
      "misc_feature 155 161\n",
      "misc_feature 161 167\n",
      "misc_feature 182 185\n",
      "misc_feature 185 191\n",
      "misc_feature 200 203\n",
      "misc_feature 206 212\n",
      "misc_feature 218 224\n",
      "misc_feature 227 230\n",
      "misc_feature 227 230\n",
      "misc_feature 248 251\n",
      "misc_feature 263 269\n",
      "misc_feature 272 278\n",
      "misc_feature 296 299\n",
      "misc_feature 296 299\n",
      "misc_feature 302 308\n",
      "misc_feature 311 314\n",
      "misc_feature 326 332\n",
      "misc_feature 329 332\n",
      "misc_feature 335 338\n",
      "misc_feature 362 368\n",
      "misc_feature 380 386\n",
      "misc_feature 407 413\n",
      "misc_feature 410 413\n",
      "misc_feature 416 422\n",
      "misc_feature 434 440\n",
      "misc_feature 470 476\n",
      "misc_feature 482 488\n",
      "misc_feature 482 485\n",
      "misc_feature 482 485\n",
      "exon 142 365\n",
      "exon 365 628\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.parse(\n",
    "    handle = './resources/HBB-human.gb',\n",
    "    format='gb',\n",
    ")\n",
    "\n",
    "for element in record:\n",
    "    for feature in element.features:\n",
    "        print(feature.type, feature.location.start, feature.location.end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CDS [50:494](+)\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTCGGTGCCTTTAGTGATGGCCTGGCTCACCTGGACAACCTCAAGGGCACCTTTGCCACACTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGTGCTGGTCTGTGTGCTGGCCCATCACTTTGGCAAAGAATTCACCCCACCAGTGCAGGCTGCCTATCAGAAAGTGGTGGCTGGTGTGGCTAATGCCCTGGCCCACAAGTATCACTAA\n"
     ]
    }
   ],
   "source": [
    "# filtering a specific feature type\n",
    "record = SeqIO.parse(\n",
    "    handle = './resources/HBB-human.gb',\n",
    "    format='gb',\n",
    ")\n",
    "\n",
    "for element in record:\n",
    "    for feature in element.features:\n",
    "        if feature.type == 'CDS':\n",
    "            print(feature.type, feature.location)\n",
    "            # extract the sequence of the feature knowing the location coordinates:\n",
    "            print(element.seq[feature.location.start:feature.location.end])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "references\n",
      "comment\n",
      "structured_comment\n",
      "molecule_type\n",
      "topology\n",
      "data_file_division\n",
      "date\n",
      "accessions\n",
      "sequence_version\n",
      "keywords\n",
      "source\n",
      "organism\n",
      "taxonomy\n",
      "comment\n",
      "structured_comment\n"
     ]
    }
   ],
   "source": [
    "# same for GenBank file containing many sequences\n",
    "record = SeqIO.parse(\n",
    "    handle = './resources/HBB.gb',\n",
    "    format='gb',\n",
    ")\n",
    "\n",
    "for elements in record:\n",
    "    print('\\n'.join(elements.annotations.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Papio anubis\n",
      "Rattus norvegicus\n",
      "Homo sapiens\n",
      "Esox lucius\n",
      "Esox lucius\n",
      "Esox lucius\n",
      "Esox lucius\n",
      "Cricetulus griseus\n",
      "Xenopus laevis\n",
      "Sus scrofa\n",
      "Ovis aries\n",
      "Bos taurus\n",
      "Danio rerio\n",
      "Oryctolagus cuniculus\n",
      "Oryctolagus cuniculus\n",
      "Macaca fascicularis\n",
      "Equus caballus\n",
      "Salmo salar\n",
      "Chlorocebus sabaeus\n",
      "Ailuropoda melanoleuca\n",
      "Condylura cristata\n",
      "Ictalurus punctatus\n",
      "Macaca mulatta\n",
      "Ailuropoda melanoleuca\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.parse(\n",
    "    handle = './resources/HBB.gb',\n",
    "    format='gb',\n",
    ")\n",
    "\n",
    "for element in record:\n",
    "    print(element.annotations['organism'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exerice on parsing GenBank file\n",
    "\n",
    "You are going to use a file called HBB.gb and this file you will find in the udemybiopython.zip file found in the tenth lecture resources.\n",
    "\n",
    "```python\n",
    "r = SeqIO.parse(\"HBB.gb\", 'gb')\n",
    "for i in r:\n",
    "    if \"references\" in i.annotations.keys():\n",
    "        for j in i.annotations[\"references\"]:\n",
    "            print(j.pubmed_id)\n",
    "```\n",
    "\n",
    "Explain the previous code and what the results appear in front of you.\n",
    "Then write a code to extract all the titles of the research in the file. The code will be close to it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10723742\n",
      "28066926\n",
      "24930900\n",
      "24333691\n",
      "23952145\n",
      "23382103\n",
      "1512262\n",
      "1520632\n",
      "2239966\n",
      "1272328\n",
      "242324\n",
      "32219817\n",
      "32078010\n",
      "31869403\n",
      "31636731\n",
      "19372376\n",
      "20301599\n",
      "20301551\n",
      "429365\n",
      "68958\n",
      "67897\n",
      "20433749\n",
      "20433749\n",
      "20433749\n",
      "20433749\n",
      "1562610\n",
      "12454917\n",
      "2999708\n",
      "6093050\n",
      "6688076\n",
      "23487454\n",
      "17145712\n",
      "15679890\n",
      "14681463\n",
      "10713517\n",
      "9250869\n",
      "7990139\n",
      "7764326\n",
      "2988373\n",
      "565742\n",
      "16962804\n",
      "2494347\n",
      "6161931\n",
      "23991043\n",
      "21269357\n",
      "19393038\n",
      "17718518\n",
      "11369847\n",
      "10449916\n",
      "8411160\n",
      "6322113\n",
      "4561255\n",
      "6048711\n",
      "27189481\n",
      "24611545\n",
      "24123299\n",
      "25603810\n",
      "23594743\n",
      "16709914\n",
      "16379021\n",
      "15169611\n",
      "10207156\n",
      "9002973\n",
      "29100196\n",
      "9692979\n",
      "3365379\n",
      "7350525\n",
      "7357610\n",
      "264241\n",
      "63556\n",
      "61580\n",
      "4530669\n",
      "5789874\n",
      "29100196\n",
      "9692979\n",
      "3365379\n",
      "7350525\n",
      "7357610\n",
      "264241\n",
      "63556\n",
      "61580\n",
      "4530669\n",
      "5789874\n",
      "22002653\n",
      "17194215\n",
      "10723742\n",
      "3110424\n",
      "5499429\n",
      "27377386\n",
      "26107351\n",
      "20383026\n",
      "7409745\n",
      "561852\n",
      "5529282\n",
      "4876811\n",
      "5659637\n",
      "5659617\n",
      "20433749\n",
      "8924215\n",
      "8151712\n",
      "22726727\n",
      "22246260\n",
      "20634964\n",
      "19959874\n"
     ]
    }
   ],
   "source": [
    "# look at ALL the references from ALL the sequences in the GenBank file\n",
    "# and print only the PubMed ID\n",
    "r = SeqIO.parse(\"./resources/HBB.gb\", 'gb')\n",
    "for i in r:\n",
    "    if \"references\" in i.annotations.keys():\n",
    "        for j in i.annotations[\"references\"]:\n",
    "            print(j.pubmed_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Strand symmetry around the beta-globin origin of replication in primates\n",
      "Low affinity hemoglobinopathy (Hb Vigo) due to a new mutation of beta globin gene (c200 A>T; Lys>Ile). A cause of rare anemia misdiagnosis\n",
      "Changes in hematological parameters in alpha-thalassemia individuals co-inherited with erythroid Kruppel-like factor mutations\n",
      "A mitochondrial location for haemoglobins--dynamic distribution in ageing and Parkinson's disease\n",
      "Acute chest syndrome is associated with single nucleotide polymorphism-defined beta globin cluster haplotype in children with sickle cell anaemia\n",
      "Platelet proteome analysis reveals integrin-dependent aggregation defects in patients with myelodysplastic syndromes\n",
      "A third quaternary structure of human hemoglobin A at 1.7-A resolution\n",
      "Beta 141 Leu is not deleted in the unstable haemoglobin Atlanta-Coventry but is replaced by a novel amino acid of mass 129 daltons\n",
      "Rapid detection of the hemoglobin C mutation by allele-specific polymerase chain reaction\n",
      "Hemoglobin Beth Israel. A mutant causing clinically apparent cyanosis\n",
      "Rat haemoglobin heterogeneity. Two structurally distinct alpha chains and functional behaviour of selected components\n",
      "[Analysis of beta-globin gene variants in Liuzhou area of Guangxi]\n",
      "Genetic polymorphisms of HbE/beta thalassemia related to clinical presentation: implications for clinical diversity\n",
      "Use of >100,000 NHLBI Trans-Omics for Precision Medicine (TOPMed) Consortium whole genome sequences improves imputation quality and detection of rare variant associations in admixed African and Hispanic/Latino populations\n",
      "Effect of N(Epsilon)-(carboxymethyl)lysine on Laboratory Parameters and Its Association with beta (S) Haplotype in Children with Sickle Cell Anemia\n",
      "Upstream open reading frames cause widespread reduction of protein expression and are polymorphic among humans\n",
      "Beta-Thalassemia\n",
      "Sickle Cell Disease\n",
      "The structure of hemoglobin Indianapolis [beta112(G14) arginine]. An unstable variant detectable only by isotopic labeling\n",
      "Human beta-globin messenger RNA. III. Nucleotide sequences derived from complementary DNA\n",
      "Complete 3' noncoding region sequences of rabbit and human beta-globin messenger RNAs\n",
      "Salmo salar and Esox lucius full-length cDNA sequences reveal changes in evolutionary pressures on a post-tetraploidization genome\n",
      "Salmo salar and Esox lucius full-length cDNA sequences reveal changes in evolutionary pressures on a post-tetraploidization genome\n",
      "Salmo salar and Esox lucius full-length cDNA sequences reveal changes in evolutionary pressures on a post-tetraploidization genome\n",
      "Salmo salar and Esox lucius full-length cDNA sequences reveal changes in evolutionary pressures on a post-tetraploidization genome\n",
      "Cloning of two adult hamster globin cDNAs (alpha and beta major)\n",
      "Genetic and genomic tools for Xenopus research: The NIH Xenopus initiative\n",
      "Comparative nucleotide sequence analysis of two types of larval beta-globin mRNAs of Xenopus laevis\n",
      "The primary structure of the larval beta 1-globin gene of Xenopus laevis and its flanking regions\n",
      "The nucleotide sequence of the mRNA encoding a tadpole beta-globin polypeptide of Xenopus laevis\n",
      "Hemoglobin subunit beta interacts with the capsid protein and antagonizes the growth of classical swine fever virus\n",
      "PEDE (Pig EST Data Explorer) has been expanded into Pig Expression Data Explorer, including 10 147 porcine full-length cDNA sequences\n",
      "Comparative analysis of protein coding sequences from human, mouse and the domesticated pig\n",
      "PEDE (Pig EST Data Explorer): construction of a database for ESTs derived from porcine full-length cDNA libraries\n",
      "Structure determination of porcine haemoglobin\n",
      "The cytogenetic map of the domestic pig\n",
      "Structure determination of aquomet porcine hemoglobin at 2.8 A resolution\n",
      "An isologous porcine promoter permits high level expression of human hemoglobin in transgenic swine\n",
      "Restriction site polymorphisms in the pig beta-globin gene cluster\n",
      "[Hemoglobins, XXI: sequence analysis of porcine hemoglobin (author's transl)]\n",
      "Barbary sheep (Ammotragus lervia): the structure of the adult beta-globin gene and the functional properties of its hemoglobin\n",
      "A comparison of the beta A-and beta B-globin gene clusters of sheep\n",
      "Hemoglobin switching in sheep. Isolation of the fetal gamma-globin gene and demonstration that the fetal gamma- and adult beta A-globin genes lie within eight kilobase segments of homologous DNA\n",
      "Molten globule of hemoglobin proceeds into aggregates and advanced glycated end products\n",
      "Polymorphisms in the bovine hemoglobin-beta gene provide evidence for gene-flow between wild species of Bos (Bibos) and domestic cattle in Southeast Asia\n",
      "A whole-genome assembly of the domestic cow, Bos taurus\n",
      "Symmetric behavior of hemoglobin alpha- and beta- subunits during acid-induced denaturation observed by electrospray mass spectrometry\n",
      "The X-ray structure determination of bovine carbonmonoxy hemoglobin at 2.1 A resoultion and its relationship to the quaternary structures of other hemoglobin crystal froms\n",
      "Physical mapping of ADCY2, FSHB and HBB to bovine chromosome 15 by FISH of bovine bacterial artificial chromosome clones\n",
      "A novel allosteric mechanism in haemoglobin. Structure of bovine deoxyhaemoglobin, absence of specific chloride-binding sites and origin of the chloride-linked Bohr effect in bovine and human haemoglobin\n",
      "Ruminant globin gene structures suggest an evolutionary role for Alu-type repeats\n",
      "Amino acid sequences in the -chains of adult bovine hemoglobins C-Rhodesia and D-Zambia\n",
      "A comparison of amino acid sequences in the beta-chains of adult bovine hemoglobins A and B\n",
      "Gene evolution and gene expression after whole genome duplication in fish: the PhyloFish database\n",
      "Global protein expression profiling of zebrafish organs based on in vivo incorporation of stable isotopes\n",
      "Zebrafish brain proteomics reveals central proteins involved in neurodegeneration\n",
      "Robustness analysis on interspecies interaction network for iron and glucose competition between Candida albicans and zebrafish during infection\n",
      "The zebrafish reference genome sequence and its relationship to the human genome\n",
      "Hypoxia induces a complex response of globin expression in zebrafish (Danio rerio)\n",
      "Expressed sequence tag analysis of zebrafish eye tissues for NEIBank\n",
      "Genomic and phylogenic comparisons of the alpha-globin and beta-globin intergenic sequences between zebra fish (Danio rerio) and six closely related Cyprinindae species\n",
      "A genetic linkage map for zebrafish: comparative analysis and localization of genes and expressed sequences\n",
      "Characterization of adult alpha- and beta-globin genes in the zebrafish\n",
      "Comparative analysis of the heme iron electronic structure and stereochemistry in tetrameric rabbit hemoglobin and monomeric soybean leghemoglobin a using Mossbauer spectroscopy with a high velocity resolution\n",
      "Rabbit beta-globin is extended beyond its UGA stop codon by multiple suppressions and translational reading gaps\n",
      "Immunopurification of the suppressor tRNA dependent rabbit beta-globin readthrough protein\n",
      "A UGA termination suppression tRNATrp active in rabbit reticulocytes\n",
      "Secondary structure of mouse and rabbit alpha- and beta-globin mRNAs: differential accessibility of alpha and beta initiator AUG codons towards nucleases\n",
      "Comparison of cloned rabbit and mouse beta-globin genes showing strong evolutionary divergence of two homologous pairs of introns\n",
      "Sequence analysis of the 3' non-coding regions of rabbit alpha- and beta-globin messenger RNAs\n",
      "Nucleotide sequence from the coding region of rabbit beta-globin messenger RNA\n",
      "Genetic variation in the primary structure of the beta chain of rabbit hemoglobin\n",
      "Haemoglobins, XVII. The primary structure of the beta-chain of rabbit haemoglobin\n",
      "Comparative analysis of the heme iron electronic structure and stereochemistry in tetrameric rabbit hemoglobin and monomeric soybean leghemoglobin a using Mossbauer spectroscopy with a high velocity resolution\n",
      "Rabbit beta-globin is extended beyond its UGA stop codon by multiple suppressions and translational reading gaps\n",
      "Immunopurification of the suppressor tRNA dependent rabbit beta-globin readthrough protein\n",
      "A UGA termination suppression tRNATrp active in rabbit reticulocytes\n",
      "Secondary structure of mouse and rabbit alpha- and beta-globin mRNAs: differential accessibility of alpha and beta initiator AUG codons towards nucleases\n",
      "Comparison of cloned rabbit and mouse beta-globin genes showing strong evolutionary divergence of two homologous pairs of introns\n",
      "Sequence analysis of the 3' non-coding regions of rabbit alpha- and beta-globin messenger RNAs\n",
      "Nucleotide sequence from the coding region of rabbit beta-globin messenger RNA\n",
      "Genetic variation in the primary structure of the beta chain of rabbit hemoglobin\n",
      "Haemoglobins, XVII. The primary structure of the beta-chain of rabbit haemoglobin\n",
      "Genome sequencing and comparison of two nonhuman primate animal models, the cynomolgus and Chinese rhesus macaques\n",
      "Rate of evolution in brain-expressed genes in humans and other primates\n",
      "Strand symmetry around the beta-globin origin of replication in primates\n",
      "Nucleotide sequence of the beta-globin genes in gorilla and macaque: the origin of nucleotide polymorphisms in human\n",
      "Structural studies on the major and minor haemoglobin of the monkey Macaca irus\n",
      "Visualizing the Bohr effect in hemoglobin: neutron structure of equine cyanomethemoglobin in the R state and comparison with human deoxyhemoglobin in the T state\n",
      "Annotation of the Protein Coding Regions of the Equine Genome\n",
      "Preliminary neutron and X-ray crystallographic studies of equine cyanomethemoglobin\n",
      "[Hemoglobins, XXXIII. Note on the Sequence of the hemoglobins of the horse (author's transl)]\n",
      "The structure of horse methaemoglobin at 2-0 A resolution\n",
      "Amide groups of some tryptic peptides from the beta-chain of horse hemoglobin\n",
      "Amino acid sequences of some tryptic peptides from the beta-chain of horse hemoglobin\n",
      "Three-dimensional Fourier synthesis of horse oxyhaemoglobin at 2.8 A resolution: the atomic model\n",
      "Three-dimensional Fourier synthesis of horse oxyhaemoglobin at 2.8 A resolution: (1) x-ray analysis\n",
      "Salmo salar and Esox lucius full-length cDNA sequences reveal changes in evolutionary pressures on a post-tetraploidization genome\n",
      "Structural organization and sequence analysis of the globin locus in Atlantic salmon\n",
      "Tail-to-tail orientation of the Atlantic salmon alpha- and beta-globin genes\n",
      "Expression sequence tag library derived from peripheral blood mononuclear cells of the chlorocebus sabaeus\n",
      "Origin and mechanism of thermal insensitivity in mole hemoglobins: a test of the 'additional' chloride binding site hypothesis\n",
      "Identification and characterization of full-length cDNAs in channel catfish (Ictalurus punctatus) and blue catfish (Ictalurus furcatus)\n",
      "Global genomic analysis reveals rapid control of a robust innate response in SIV-infected sooty mangabeys\n"
     ]
    }
   ],
   "source": [
    "# print the titles instead of the pubmed_id\n",
    "r = SeqIO.parse(\"./resources/HBB.gb\", 'gb')\n",
    "for i in r:\n",
    "    if \"references\" in i.annotations.keys():\n",
    "        for j in i.annotations[\"references\"]:\n",
    "            print(j.title)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating FASTA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from GenBank to FASTA\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "####\n",
    "# no more alphabet in Biopython!\n",
    "# see https://biopython.org/wiki/Alphabet\n",
    "#from Bio.Alphabet import DNAAlphabet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NM_001168847.1 Papio anubis hemoglobin, beta (HBB), mRNA\n",
      "ATGGTGCATCTGACTCCTGAGGAGAAGAATGCCGTTACCGCCCTGTGGGGCAAAGTGAAC\n",
      "GTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGTTGCTGGTGGTCTACCCTTGGACCCAG\n",
      "AGGTTCTTTGATTCCTTTGGGGATCTGTCCTCTCCTGCTGCTGTTATGGGCAACCCTAAG\n",
      "GTGAAGGCTCATGGCAAGAAAGTGCTTGGTGCCTTTAGTGATGGCCTGAATCACCTGGAC\n",
      "AACCTCAAGGGCACCTTTGCCCAGCTCAGTGAGCTGCACTGTGACAAGCTGCATGTGGAT\n",
      "CCTGAGAACTTCAAGCTCCTGGGCAACGTGCTGGTGTGTGTGCTGGCCCATCACTTTGGC\n",
      "AAAGAATTCACCCCGCAAGTGCAGGCTGCCTATCAGAAAGTGGTGGCTGGTGTGGCTAAT\n",
      "GCCCTGGCCCACAAGTACCACTAA\n",
      "\n",
      ">NM_033234.1 Rattus norvegicus hemoglobin subunit beta (Hbb), mRNA\n",
      "TGCTTCTGACATAGTTGTGTTGACTCACAAACTCAGAAACAGACACCATGGTGCACCTGA\n",
      "CTGATGCTGAGAAGGCTGCTGTTAATGGCCTGTGGGGAAAGGTGAACCCTGATGATGTTG\n",
      "GTGGCGAGGCCCTGGGCAGGCTGCTGGTTGTCTACCCTTGGACCCAGAGGTACTTTGATA\n",
      "GCTTTGGGGACCTGTCCTCTGCCTCTGCTATCATGGGTAACCCTAAGGTGAAGGCCCATG\n",
      "GCAAGAAGGTGATAAACGCCTTCAATGATGGCCTGAAACACTTGGACAACCTCAAGGGCA\n",
      "CCTTTGCTCATCTGAGTGAACTCCACTGTGACAAGCTGCATGTGGATCCTGAGAACTTCA\n",
      "GGCTCCTGGGCAATATGATTGTGATTGTGTTGGGCCACCACCTGGGCAAGGAATTCACCC\n",
      "CCTGTGCACAGGCTGCCTTCCAGAAGGTGGTGGCTGGAGTGGCCAGTGCCCTGGCTCACA\n",
      "AGTACCACTAAACCTCTTTTCCTGCTCTTGTCTTTGTGCAATGGTCAATTGTTCCCAAGA\n",
      "GAGCATCTGTCAGTTGTTGTCAAAATGACAAAGACCTTTGAAAATCTGTCCTACTAATAA\n",
      "AAGGCATTTACTTTCACTGC\n",
      "\n",
      ">NM_000518.5 Homo sapiens hemoglobin subunit beta (HBB), mRNA\n",
      "ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGGTGCATC\n",
      "TGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAG\n",
      "TTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTG\n",
      "AGTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTC\n",
      "ATGGCAAGAAAGTGCTCGGTGCCTTTAGTGATGGCCTGGCTCACCTGGACAACCTCAAGG\n",
      "GCACCTTTGCCACACTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGATCCTGAGAACT\n",
      "TCAGGCTCCTGGGCAACGTGCTGGTCTGTGTGCTGGCCCATCACTTTGGCAAAGAATTCA\n",
      "CCCCACCAGTGCAGGCTGCCTATCAGAAAGTGGTGGCTGGTGTGGCTAATGCCCTGGCCC\n",
      "ACAAGTATCACTAAGCTCGCTTTCTTGCTGTCCAATTTCTATTAAAGGTTCCTTTGTTCC\n",
      "CTAAGTCCAACTACTAAACTGGGGGATATTATGAAGGGCCTTGAGCATCTGGATTCTGCC\n",
      "TAATAAAAAACATTTATTTTCATTGCAA\n",
      "\n",
      ">NM_001304110.1 Esox lucius hemoglobin subunit beta (LOC105010112), mRNA\n",
      "GGGTATTCCTATAGAAGAACTATTGTGCGAGTCAAAACCCCTAAATCGCAAATATGGTTC\n",
      "AGTGGACAGCAGCTGAGCGCAAAGCTATAGCAAGCGTTTGGGGAAGTATCAGCGCTGATG\n",
      "AGATTGGACCCCAATCCGTGGCCAGACTTTTGATCGTGTTTCCATGGACTCGGAGGTATT\n",
      "TCAGCAGCTTCGGTAACCTGGCCGACTCAGCTGCCATCTTGGGTAACCCCAAAGTGGCTA\n",
      "ACCATGGGAAGACGGTGATGAAGGCACTTGACAAGGCTGTGCAGAACCTGGATAACATCA\n",
      "AGAAAACCTACACCGCACTCAGTGTGACCCATTCCGAGAAACTGCATGTGGATCCTGACA\n",
      "ACTTCAAGCTCTTGTCCGAGTGCATCACCGTGTGCATTGCCGCAAAGCTGGGCCCCACCG\n",
      "TTTTCGATGCTTATACCCATGAAGCCTTCTACAAGTTCATGTGTGTGGTTGTGTCCGCTC\n",
      "TCAGCAAACAGTACCACTAGAGCATTATTATTTTACAGAGATGTGCAACAGCTCCAGAAT\n",
      "TAAAGGCTGTTGTCTTTTAATGTCAATTAAAACCCATTAAAAAGCATTACAATAC\n",
      "\n",
      ">NM_001303935.1 Esox lucius hemoglobin subunit beta (LOC105013372), mRNA\n",
      "GGGCAGGGATCCCTTGGAACATTTTACAAGTACCAACGGAAATCAACAAACGTCAACATG\n",
      "GTCGACTGGACAACAGCTGAGCGCAGTGCCATCCTTGGCCTTTGGGGAAAGCTCAACGCT\n",
      "GATGAGATCGGAACCCAGGCCTTTATCAGATGTCTGGTCGTGTCTCATTGGACTCAGAGG\n",
      "TACTTCAGCTCCTTCGGTAACCTGTCCACACCCGCTGCCATCGCAGGTAACCCCAACGTG\n",
      "GCTAAGCATGGGAGGACCGTGATGCATGGACTGGACAGAGCTATCCAGAACCTGGATGAC\n",
      "GTCAAGAACACCTACGCTCCACTGAGTGTGATGCACTCCGAGAAACTGCACGTGGATCCC\n",
      "GACAACTTTAGGCCCCTCGCGGAATGCATCACCGTGTGCGTCGCCGCCAAGCTGGGTCCT\n",
      "TCCGTTTTCGATGCTGACACTCATGAAGCCTTCCAGAAGTTCCTGCGTGTGGTTGTTTCG\n",
      "GCTCTGGGCAGACAGTACCACTAGAGCACAACTCAAAAACATCAATGTGGAATACAGATG\n",
      "ACGCTCCTACTTCAGTCTTTTAGACTGCAAGCCGTGCCATATCAAGAAAATATTAAAATA\n",
      "ATAAAATGATTTAAAAGC\n",
      "\n",
      ">NM_001303868.1 Esox lucius hemoglobin, beta (LOC105028448), mRNA\n",
      "AGTGATCCTTTGGAACATTTTGCTAGTGCCAACGGAAATCAACAAACAAACGTCATCATG\n",
      "GTCGACTGGACAGATGCTGAGCGCAGTGCCATCCTTGGTCTTTGGGGAAAGTTGAGCGCG\n",
      "GACGAGATCGGACCCCAGGCCTTTGCCAGATGTCTGATCGTGTTTCCTTGGACTCAGAGG\n",
      "TACTTCAGCTCCTTCGGTAACCTGTCCACACCCGCTGCCATCATGGGTAACCCCAACGTG\n",
      "GCTAAGCATGGGAGGACCGTGATGCACGGTCTGGACAAAGCTATTCAGAACCTGGATGAC\n",
      "ATCAAGAACACCTACGCTGCTCTCAGTGTGAAGCACTCCGAGAAACTGCACGTGGATCCT\n",
      "GACAATTTCAGGGTCCTCGCCGAATGCATCACCGTGTGCGTGGCCGCAAAACTCGGGCCC\n",
      "GCGGTTTTCAATGCTGATACTCAGGAAGCCTTCCAGAAGTTCCTGTCTGTCGTTGTCTCG\n",
      "GCTCTCGGCAGACAGTACCACTGAGCAGCATTCCACAACATCAATATGGAAGAGAGATGA\n",
      "TACTCAAACGTCAGTCTATTCGGCTGAAAGCGGTGCCATATCTACAAATATTGAAATAAT\n",
      "AAATAGATTTTAAAATTAATGCTTGTCAACTATTGTATGTTTTACGGTAAACGCTCTGTT\n",
      "GCAACATTAAAATGTTTTGCATATGGCTTTTGAAAACCCACACCATCAAGTAATACTTCC\n",
      "GTAGTTTTGATAGAGATATCGTGGAATGTTTTGAGATTCCAATAATGTTAATATACTTTT\n",
      "ATTTATTCTGAGTTTCCAAAAGCGCGTGAAATTCGTGCACAAAATATATTTAAGTTTACC\n",
      "ATATTAAACCTAATCATCCATTTACTTTTAGCGTTGAAGCATCATCCAGACCGTTTCTTT\n",
      "GGTATGCGCATACAACTCAGCTAAATAAATCTCTGAAACTGAACTTTTGCCAAAATAACT\n",
      "GTCTGGATGATGAGGGAACACTAAAAGCGAAAAATTCAACCGTTCCTATCATTAGGTTTG\n",
      "TTTGGAACTAATAAACCGCGATTACTTTGAGTAGGGGATAGGTGTAACATGGGTTAGTTG\n",
      "TAACACATCCAATTTCTCTAATTTAGGAATAAATTACGAATAATTTGATTCACCCTGCAC\n",
      "ATGCTCAATTTAATCATTTTTTCACATGTGAAGAAGTAGCAATGGCAGACACAACCATTC\n",
      "TTCAAAGGAAACGAATTTTGAAAGTTTTTCTTTTACCTCATTAACTTTTTGTGATAGTAT\n",
      "TGCCACCTATGTAATAAAATTCATCAAAC\n",
      "\n",
      ">NM_001303858.1 Esox lucius hemoglobin subunit beta (LOC105028446), mRNA\n",
      "GGGGGGGCACAGGAATCCTTTGGAACATTTTGGTATTACTAACGGAAATCGCTGAAAAAA\n",
      "CGTCAACATGGTCGTCTGGACAACGGCTGAACGCAATGCCATCGTTGGCCTTTGGGGAAA\n",
      "GATCGGTGTGGATGAGATCGGACACCAGGCCTTTGCCAGATGTCTGATCGTGTTTCCTTG\n",
      "GACTCAGAGGTACTTCAGCTCATTCGGTAACCTGTCCACACCCGCTGCCATCATGGGTAA\n",
      "CCCCAACGTGGCTAAGCATGGGAGGACCGTGATACACGGTCTGGACAAAGCTATTCAGAA\n",
      "CCTGGATGACATCAAGAACACTTACGCTGCTCTCAGTGTGAAGCACTCCGAGAAACTGCA\n",
      "CGTGGATCCCGACAATTTCAGGCTCCTCGCGGAATGCATCACCGTGTGCGTGGCTGCCAA\n",
      "GCTCGGGCCTGAAGTTTTCAATGCTGACGCTCATGAAGCCTTTCATAAGTTCCTGGCTGT\n",
      "TGTTGTTTCGGCTCTCGGCAGACAATACCACTAGAGTGGCACTCGACAGCAATATGGAGG\n",
      "GAGGAAACACCAAGTCTAGTCATGTAGGTTGAAGCTGGTCATATCTCCGAATATTGAAAT\n",
      "AATAAAACCGATATAAAATA\n",
      "\n",
      ">NM_001246752.1 Cricetulus griseus beta major globin chain (Hbb), mRNA\n",
      "TTTGCTTCTGACATTAGTTGTGTTGAGTCACAAGCCCAGAAACAAGACACCATGGTGCAC\n",
      "CTGACTGATGCTGAGAAGGCCCTTGTCACTGGCCTGTGGGGAAAGGTGAACGCCGATGCA\n",
      "GTTGGCGCTGAGGCCCTGGGCAGGTTGCTGGTTGTCTACCCTTGGACCCAGAGGTTCTTT\n",
      "GAACACTTTGGAGACCTGTCTCTGCCAGTTGCTGTCATGAATAACCCCCAGGTGAAGGCC\n",
      "CATGGCAAGAAGGTGATCCACTCCTTCGCTGATGGCCTGAAACACCTGGACAACCTGAAG\n",
      "GGCGCCTTTTCCAGCCTGAGTGAGCTCCACTGTGACAAGCTGCACGTGGATCCTGAGAAC\n",
      "TTCAAGCTCCTGGGCAATATGATCATCATTGTGCTGATCCACGACCTGGGCAAGGACTTC\n",
      "ACTCCCAGTGCACAGTCTGCCTTTCATAAGGTGGTGGCTGGTGTGGCCAATGCCCTGGCT\n",
      "CACAAGTACCACTAAACTCCCTTTCCTGCCCTTATTTCTGTGACTCAAAAGTCACTTGTT\n",
      "CCCTAGAGAACAACTGTCAACCTGTGAGAAAAATGAGAAGAGCACTTTAGGGACATCTGC\n",
      "\n",
      ">NM_001086273.2 Xenopus laevis hemoglobin subunit delta S homeolog (hbd.S), mRNA\n",
      "ACAGCAGCCTCTTACTTCTGCCATTTGAAGTCTTAGCAGCTACTCCCATCTACAGCCACC\n",
      "ATGGTGCACCTGTCAGCTGATGAGAAATCCGCCATTAACGCCGTATGGTCAAAGGTCAAC\n",
      "ATTGAAAATGATGGCCATGATGCCCTGACCAGGCTGCTGGTTGTGTTTCCCTGGACCCAG\n",
      "AGGTATTTCAGCAGCTTTGGAAACCTATCCAATGTGGCTGCCATCTCTGGAAATGCCAAG\n",
      "GTTCGTGCCCATGGCAAGAAGGTTTTGTCTGCCGTCGATGAATCCATCCACCATTTAGAT\n",
      "GATATCAAGAACTTCCTCTCTGTACTGAGCACGAAGCACGCTGAGGAACTCCACGTGGAC\n",
      "CCTGAAAACTTCAAGCGTCTTGCTGATGTGCTGGTGATTGTCTTGGCTGGCAAACTGGGA\n",
      "GCTGCCTTCACTCCTCAAGTCCAGGCTGCTTGGGAGAAATTCAGCGCTGGGCTGGTGGCT\n",
      "GCTCTTAGCCATGGCTACTTCTAAAGAATTTTTTTCCATGCAAAGCCCAACATCTGCTCG\n",
      "CCTGCAACAAAGACAACTCTCACAATAAGTGCATTGCTTGTTGTCAATAAAGATTTCTTG\n",
      "TCAAACTGCCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      "\n",
      ">NM_001144841.1 Sus scrofa hemoglobin, beta (HBB), mRNA\n",
      "ACATTTGCTTCTGACACAACCGTGTTCACTAGCAACTGCACAAACAGACAACATGGTGCA\n",
      "TCTGTCTGCTGAGGAGAAGGAGGCCGTCCTCGGCCTGTGGGGCAAAGTGAATGTGGACGA\n",
      "AGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTTGTCTACCCCTGGACTCAGAGGTTCTT\n",
      "CGAGTCCTTTGGGGACCTGTCCAATGCCGATGCCGTCATGGGCAATCCCAAGGTGAAGGC\n",
      "CCACGGCAAGAAGGTGCTCCAGTCCTTCAGTGACGGCCTGAAACATCTCGACAACCTCAA\n",
      "GGGCACCTTTGCTAAGCTGAGCGAGCTGCACTGTGACCAGCTGCACGTGGATCCTGAGAA\n",
      "CTTCAGGCTCCTGGGCAACGTGATAGTGGTTGTTCTGGCTCGCCGCCTTGGCCATGACTT\n",
      "CAACCCGAATGTGCAGGCTGCTTTTCAGAAGGTGGTGGCTGGTGTTGCTAATGCCCTGGC\n",
      "CCACAAGTACCACTAA\n",
      "\n",
      ">NM_001097648.1 Ovis aries hemoglobin, beta (HBB), mRNA\n",
      "ATGCTGACTGCTGAGGAGAAGGCTGCCGTCACCGGCTTCTGGGGCAAGGTGAAAGTGGAT\n",
      "GAAGTTGGTGCTGAAGCCCTGGGCAGGCTGCTGGTTGTCTACCCCTGGACTCAGAGGTTC\n",
      "TTTGAGCACTTTGGGGACTTGTCCTCTGCTGATGCTGTTATGAACAACGCTAAGGTGAAG\n",
      "GCCCATGGCAAGAAGGTGCTAGACTCCTTTAGTAACGGCGTGCAGCATCTTGACGACCTC\n",
      "AAGGGCACCTTTGCTCAGCTGAGTGAGCTGCACTGTGATAAGCTGCACGTGGATCCTGAG\n",
      "AACTTCAGGCTCCTGGGCAACGTGCTGGTGGTTGTGCTGGCTCGCCACCATGGCAGTGAA\n",
      "TTCACCCCGGTGCTGCAGGCTGAGTTTCAGAAGGTGGTGGCTGGTGTTGCCAATGCCCTG\n",
      "GCCCACAGATATCACTAA\n",
      "\n",
      ">NM_173917.2 Bos taurus hemoglobin, beta (HBB), mRNA\n",
      "ACACTTGCTTCTGACACAACCGTGTTCACTAGCAACTACACAAACAGACACCATGCTGAC\n",
      "TGCTGAGGAGAAGGCTGCCGTCACCGCCTTTTGGGGCAAGGTGAAAGTGGATGAAGTTGG\n",
      "TGGTGAGGCCCTGGGCAGGCTGCTGGTTGTCTACCCCTGGACTCAGAGGTTCTTTGAGTC\n",
      "CTTTGGGGACTTGTCCACTGCTGATGCTGTTATGAACAACCCTAAGGTGAAGGCCCATGG\n",
      "CAAGAAGGTGCTAGATTCCTTTAGTAATGGCATGAAGCATCTCGATGACCTCAAGGGCAC\n",
      "CTTTGCTGCGCTGAGTGAGCTGCACTGTGATAAGCTGCATGTGGATCCTGAGAACTTCAA\n",
      "GCTCCTGGGCAACGTGCTAGTGGTTGTGCTGGCTCGCAATTTTGGCAAGGAATTCACCCC\n",
      "GGTGCTGCAGGCTGACTTTCAGAAGGTGGTGGCTGGTGTGGCCAATGCCCTGGCCCACAG\n",
      "ATATCATTAAGCTCCCTTTCCTGCTTTCCAGGAAAGGTTTTTTCATCCTCAGAGCCCAAA\n",
      "GATTGAATATGGAAAAATTATGAAGTGTTTTGAGCATCTGGCCTCTGCCTAATAAAGACA\n",
      "TTTATTTTCATTGCAAAAAAAAAAAAAAAAAAA\n",
      "\n",
      ">NM_131020.3 Danio rerio hemoglobin, beta adult 1 (hbba1), mRNA\n",
      "CGTCATCTCAGCAAGATTATCCAATAAGATTTTGCATCTTTTATTGGGCGGAGACATCAC\n",
      "ACATCAATAAAATGCACCCCGGTTTCTGAATTTCTGCTAGTGATTCCTACTGGACTTTAA\n",
      "GCAAGCACCTAAGGACATCCTAAAGCGTAAACATGGTTGAGTGGACAGATGCCGAGCGCA\n",
      "CAGCCATCCTTGGCCTGTGGGGAAAGCTCAATATCGATGAAATCGGACCTCAGGCCCTAT\n",
      "CCAGATGTCTGATCGTGTATCCCTGGACTCAGAGATATTTCGCCACATTCGGCAACCTGT\n",
      "CAAGCCCCGCTGCGATCATGGGTAACCCCAAAGTGGCAGCTCATGGGAGGACTGTGATGG\n",
      "GAGGTCTTGAGAGAGCCATCAAGAACATGGACAACGTCAAGAACACCTATGCCGCCCTCA\n",
      "GTGTGATGCACTCTGAGAAACTGCATGTGGATCCCGACAACTTCAGGCTTCTCGCTGATT\n",
      "GCATCACCGTTTGCGCTGCCATGAAGTTCGGCCAAGCTGGTTTCAATGCTGATGTCCAGG\n",
      "AGGCCTGGCAGAAGTTTCTGGCTGTGGTCGTTTCTGCTCTGTGCAGACAGTACCACTAGA\n",
      "GTCTCATCGCCAATGAACGTCAGGTGTTTTGGAGAAGGTCACACCGCGCGGAGATGTTCA\n",
      "AACAAAGATCTAACTCTTTTGCAACAAGATCAATAAAATAATGTTAAAAGCATCTC\n",
      "\n",
      ">NM_001314043.1 Oryctolagus cuniculus hemoglobin, beta (HBB2), transcript variant 1, mRNA\n",
      "ACTGCTGGTATGGGTCTGGGAGATACATAGAAGGAAGGCTGAGTCTGTCAGACTCCTAAG\n",
      "CCATTGCCATAACTGCCAAGGACAGGGGTGCTGTCATCACCCAGACCTCACCCTGCAGAG\n",
      "CCACACCCTGGTGTTGGCCAATCTACACACGGGGTAGGGATTACATAGTTCAGGACTTGG\n",
      "GCATAAAAGGCAGAGCAGGGCAGCTGCTGCTTACACTTGCTTTTGACACAACTGTGTTTA\n",
      "CTTGCAATCCCCCAAAACAGACAGAATGGTGCATCTGTCCAGTGAGGAGAAGTCTGCGGT\n",
      "CACTGCCCTGTGGGGCAAGGTGAATGTGGAAGAAGTTGGTGGTGAGGCCCTGGGCAGGCT\n",
      "GCTGGTTGTCTACCCATGGACCCAGAGGTTCTTCGAGTCCTTTGGGGACCTGTCCTCTGC\n",
      "ACATGCTGTTATGAGCAATCCTAAGGTGAAGGCTCATGGCAAGAAGGTGCTGGCTGCCTT\n",
      "CAGTGAGGGTCTGAATCACCTGGACAACCTCAAAGGCACCTTTGCTAAGCTGAGTGAACT\n",
      "GCACTGTGACAAGCTGCACGTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGTGCTGGT\n",
      "TGTTGTGCTGTCTCATCATTTTGGCAAAGAATTCACTCCTCAGGTGCAGGCTGCCTATCA\n",
      "GAAGGTGGTGGCTGGTGTGGCCAATGCCCTGGCTCACAAATACCACTGAGATCTTTTTCC\n",
      "CTCTGCCAAAAATTATGGGGACATCATGAAGCCCCTTGAGCATCTGACTTCTGGCTAATA\n",
      "AAGGAAATTTATTTTCATTGCAAAAAAAAAAAAAAAAAAAAA\n",
      "\n",
      ">NM_001082260.3 Oryctolagus cuniculus hemoglobin, beta (HBB2), transcript variant 1, mRNA\n",
      "ACTGCTGGTATGGGTCTGGGAGATACATAGAAGGAAGGCTGAGTCTGTCAGACTCCTAAG\n",
      "CCATTGCCATAACTGCCAAGGACAGGGGTGCTGTCATCACCCAGACCTCACCCTGCAGAG\n",
      "CCACACCCTGGTGTTGGCCAATCTACACACGGGGTAGGGATTACATAGTTCAGGACTTGG\n",
      "GCATAAAAGGCAGAGCAGGGCAGCTGCTGCTTACACTTGCTTTTGACACAACTGTGTTTA\n",
      "CTTGCAATCCCCCAAAACAGACAGAATGGTGCATCTGTCCAGTGAGGAGAAGTCTGCGGT\n",
      "CACTGCCCTGTGGGGCAAGGTGAATGTGGAAGAAGTTGGTGGTGAGGCCCTGGGCAGGCT\n",
      "GCTGGTTGTCTACCCATGGACCCAGAGGTTCTTCGAGTCCTTTGGGGACCTGTCCTCTGC\n",
      "ACATGCTGTTATGAGCAATCCTAAGGTGAAGGCTCATGGCAAGAAGGTGCTGGCTGCCTT\n",
      "CAGTGAGGGTCTGAATCACCTGGACAACCTCAAAGGCACCTTTGCTAAGCTGAGTGAACT\n",
      "GCACTGTGACAAGCTGCACGTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGTGCTGGT\n",
      "TGTTGTGCTGTCTCATCATTTTGGCAAAGAATTCACTCCTCAGGTGCAGGCTGCCTATCA\n",
      "GAAGGTGGTGGCTGGTGTGGCCAATGCCCTGGCTCACAAATACCACTGAGATCTTTTTCC\n",
      "CTCTGCCAAAAATTATGGGGACATCATGAAGCCCCTTGAGCATCTGACTTCTGGCTAATA\n",
      "AAGGAAATTTATTTTCATTGCAAAAAAAAAAAAAAAAAAAAA\n",
      "\n",
      ">NM_001283367.1 Macaca fascicularis hemoglobin, beta (HBB), mRNA\n",
      "ACACTTGCTTCTGACACAACTGTGTTCACGAGCAACCTCAAACAGACACCATGGTGCATC\n",
      "TGACTCCTGAGGAGAAGAATGCCGTCACCACCCTGTGGGGCAAGGTGAACGTGGATGAAG\n",
      "TTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTG\n",
      "AGTCCTTTGGGGATCTGTCCTCTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTC\n",
      "ATGGCAAGAAAGTGCTTGGTGCCTTTAGTGATGGCCTGAATCACCTGGACAACCTCAAGG\n",
      "GTACCTTTGCCCAGCTCAGTGAGCTGCACTGTGACAAGCTGCATGTGGATCCTGAGAACT\n",
      "TCAAGCTCCTGGGCAACGTGCTGGTGTGTGTGCTGGCCCATCACTTTGGCAAAGAATTCA\n",
      "CCCCGCAAGTGCAGGCTGCCTATCAGAAAGTGGTGGCTGGTGTGGCTAATGCCCTGGCCC\n",
      "ACAAGTACCACTAAGCTCACTTTCTTGCTGTCCAATTTCTACCAAAGGTTCCTTTGTTCC\n",
      "CAAAGTCCAACTACTGAACTGGGGGATATTATGAAGGGCCTTGAGGATCTGGATTCTGCC\n",
      "TAAT\n",
      "\n",
      ">NM_001164018.1 Equus caballus hemoglobin, beta (HBB), mRNA\n",
      "ATGGTGCAACTGAGTGGTGAAGAGAAGGCAGCTGTCTTGGCCCTGTGGGACAAGGTGAAT\n",
      "GAGGAAGAAGTTGGTGGTGAAGCCCTGGGCAGGCTGCTGGTTGTCTACCCATGGACTCAG\n",
      "AGGTTCTTTGACTCCTTTGGGGATCTGTCCAATCCTGGTGCTGTGATGGGCAACCCCAAG\n",
      "GTGAAGGCCCACGGCAAGAAAGTGCTACACTCCTTTGGTGAGGGCGTGCATCATCTTGAC\n",
      "AACCTCAAGGGCACCTTTGCTGCGCTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGAT\n",
      "CCTGAGAACTTCAGGCTCCTGGGCAACGTGCTGGTTGTTGTGCTGGCTCGCCACTTTGGC\n",
      "AAGGATTTCACCCCAGAGTTGCAGGCTTCCTATCAAAAGGTGGTGGCTGGTGTGGCCAAT\n",
      "GCACTGGCCCACAAATACCACTGA\n",
      "\n",
      ">NM_001123666.1 Salmo salar beta globin (LOC100136576), mRNA\n",
      "TTTGCAAGTACTAAAGGAAATCAACAAACAAACGTCAACATGGTCGACTGGACAGATGCT\n",
      "GAGCGCAGCGCCATCGTAGGCCTGTGGGGAAAGATCAGCGTGGATGAGATCGGACCCCAG\n",
      "GCCCTGGCCAGACTTCTGATCGTGTCTCCATGGACTCAGAGGCACTTTAGCACCTTCGGC\n",
      "AACCTGTCCACACCCGCTGCCATCATGGGTAACCCCGCCGTGGCCAAGCACGGAAAGACC\n",
      "GTGATGCACGGACTGGACAGAGCTGTGCAGAACCTGGATGACATCAAAAACGCCTATACT\n",
      "GCCCTGAGTGTGATGCACTCCGAGAAACTGCACGTGGATCCCGACAACTTCAGGCTCCTC\n",
      "GCCGACTGCATCACCGTGTGCGTGGCAGCCAAGCTCGGTCCCACCGTTTTCAGTGCTGAT\n",
      "ATTCAGGAAGCCTTCCAGAAGTTCCTGGCTGTCGTTGTGTCCGCTCTTGGCAGACAGTAC\n",
      "CACTAGAGCATCACTCGACAGCATCAATATGGAAGAGAGATGACACTCCAATCCAGTCTG\n",
      "TTAGGCTGGAAGCTGTGCCATAGCTACACATTTGAAAAAAATTAAATCTATTTAAAAGCC\n",
      "GTTTATCTTTTTGAAGTGTTGTGTTTTATGGAAAACGTTCTTCTGCAATTGTACTGATGG\n",
      "CACATGGTTTCTACACTCTAAAAATTAGGGAATGTGAATGAAAAAAACTGTTTTGATGAT\n",
      "GGTTTTCATAAACATACTTTGTCCATG\n",
      "\n",
      ">NM_001329918.1 Chlorocebus sabaeus hemoglobin subunit beta (HBB), mRNA\n",
      "GCTTCTGACACAACTGTGTTCACTAGCAACCTCAGACACCATGGTGCATCTGACTCCTGA\n",
      "GGAGAAGACTGCCGTTACCACCCTGTGGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGA\n",
      "GGCCCTGGGCAGGCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGG\n",
      "GGATCTGTCCTCTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTCATGGCAAGAA\n",
      "AGTGCTTGGTGCCTTTAGTGATGGCCTGGCTCACCTGGACAACCTCAAGGGCACCTTTGC\n",
      "CCAGCTGAGTGAGCTGCACTGTGACAAGCTGCATGTGGATCCTGAGAACTTCAAGCTCCT\n",
      "GGGCAACGTGCTGGTGTGTGTGCTGGCCCATCACTTTGGCAAAGAATTCACCCCGCAAGT\n",
      "GCAGGCTGCCTATCAGAAAGTGGTGGCTGGTGTGGCTAATGCCCTGGCCCACAAGTACCA\n",
      "CTAAGTTCACTTTCTTGCTGTCCAATTTCTACCAAAGGTTCCTTTGTTCCCAAAGTCCAA\n",
      "CTACTGAACTGGGGGATATTATGAAGGGCCTTGAGGATCTGGATTCTGCCTAATAAAAAA\n",
      "CATTTATTTTCATTGC\n",
      "\n",
      ">NM_001304883.1 Ailuropoda melanoleuca hemoglobin subunit beta-like (HBB), mRNA\n",
      "ATACTTGCTTCTGACACAACTGTGTTCACTAGCAACCACAAAGAGACACCATGGTGCATC\n",
      "TGACTGGTGAGGAGAAGGCTGCCGTCACCGGCCTGTGGAGCAAGGTGAACGTGGATGAAG\n",
      "TTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTTGTCTACCCCTGGACTCAGAGGTTCTTTG\n",
      "ACTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGAACAACCCCAAGGTCAAGGCCC\n",
      "ATGGCAAGAAGGTGCTGAACTCCTTTAGTGAAGGCCTGAAGAATCTGGACAACCTCAAGG\n",
      "GTACCTTTGCTAAGCTGAGCGAGCTGCACTGTGACAAGCTGCACGTGGATCCCGAGAACT\n",
      "TCAAGCTCCTGAGCAACGTGCTGGTGTGTGTGCTGGCTCACCACTTTGGCAAGGAGTTCA\n",
      "CCCCTCAGGTGCAGGCTGCCTATCAGAAGGTGGTGGCTGGTGTGGCCAATGCCCTGGCCC\n",
      "ACAAGTACCACTGAGTTCCCTTCCTGCTTTCCAGGAAAAGTCCCCTGGTCCCTAGAACCC\n",
      "AACACCTGAACTTGGGGAAATTGTGAAGTGCTTTGAGCACCTGGCTCCTGCCTAATAAAG\n",
      "AACACTTCTTTTCATTGGAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      "\n",
      ">NM_001279263.1 Condylura cristata hemoglobin, beta (HBB), mRNA\n",
      "GAAAGACCAACTGCTGCTCATACTTGCTTCTGAAGCTACCGTGTTCACGAGCAACTACAC\n",
      "AAACAGACACCATGGTGCATCTGTCTGGTGAGGAAAAGGGCCTCGTCACCTCCATGTGGG\n",
      "GCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTTGTCTACC\n",
      "CCTGGACCCAGAGGTTCTTTGACAGCTTTGGTGACCTGTCCTCTGCCTCTGCCATTATGG\n",
      "GCAATGCCAAGGTGAAGGCCCATGGCAAGAAGGTGGCCCACTCCATCACTGATGGCATCA\n",
      "AAAACCTGGACAACCTCAAAGGCACCTATGCCAAGCTGAGTGAGCTGCACTGTGACAAGC\n",
      "TGCACGTGGACCCTGAGAACTTCAGGCTGCTGGGCAACGTGCTGGTGTGTGTGCTGGCTC\n",
      "GCAACCTTGGCAAGGAATTCACCCCCCAGGCTCAGGCTGCCTTCCAGAAGGTGGTGCTTG\n",
      "GTGTGGCCACTGCCCTGGCTCACAAGTACCATTGAGATCTTAAACTCTTCCTGGAGGCCA\n",
      "TTGGAACACACTATTTCCCTATGTTCTGTCCTCTCAGCTTGGGGAAATAATGCCCGCCTT\n",
      "CAGGGCATGTTTTCTGCCTAATAAAG\n",
      "\n",
      ">NM_001201019.1 Ictalurus punctatus hemoglobin subunit beta (hbb), mRNA\n",
      "CCCACGCGTCCGATAGAAACCCCAAAAAGCAGCAACATGGTTGTTTGGACAGACTTCGAG\n",
      "CGCGCTACCATCCAGGACGTCTTTTCCAAGATCGACTACGAGTCAGTTGGTCACCAGGCA\n",
      "CTGTCAAGGTGTCTCGTCGTTTACCCATGGACCCAGAGGTACTTCGGTAGCTTTGGAAAC\n",
      "TTGTACAACGCCGCTGCCATCATTGGAAACCCCAAGGTTGCTGCCCACGGCTTGGTAGTA\n",
      "GTACGTGGTCTGGAGAAAGCAGCGAAGAACATGGACAACATTAAGGCCATTTACGCCGAC\n",
      "CTGAGTGTGCTGCACTCTGAGAAACTGCACGTGGATCCCGATAACTTCAAGCTGCTGGCT\n",
      "GACTGCATCACCATCGTCATTGCCTCTGTCCTGGGTGCCAGCTTCACAGCCGAAGTACAG\n",
      "GCAGCTCTGCAGAAGTTCCTGGCTGTTGTTGTCTCTGCACTCGGAAAGCAGTACCAATAA\n",
      "ATGCTGAAGTGATTTGACGGCCAGAGAGGACCATCAAACGTTTTCAGTTCTAAAAACGAC\n",
      "CATGTACTACTTGTGAAAGCAAGTATTTAAATAAAATTACCAATACACAAAAAAAAAAAA\n",
      "AAAAAAA\n",
      "\n",
      ">NM_001164428.1 Macaca mulatta hemoglobin subunit beta (HBB), mRNA\n",
      "ACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGACTGCCGTTACCACCCTGTGGGGC\n",
      "AAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCT\n",
      "TGGACCCAGAGGTTCTTTGATTCCTTTGGGGATCTGTCCTCTCCTGATGCTGTTATGGGC\n",
      "AACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTTGGTGCCTTTAGTGATGGCCTGAAT\n",
      "CACCTGGACAACCTCAAGGGCACCTTTGCCCAGCTCAGTGAGCTGCACTGTGACAAGCTG\n",
      "CATGTGGATCCTGAGAACTTCAAGCTCCTGGGCAACGTGCTGGTGTGTGTGCTGGCCCAT\n",
      "CACTTTGGCAAAGAATTCACCCCGCAAGTGCAGGCTGCCTATCAGAAAGTGGTGGCTGGT\n",
      "GTGGCTAATGCCCTGGCCCACAAGTACCACTAAGCTCACTTTCTTGCTGTC\n",
      "\n",
      ">NM_001304885.1 Ailuropoda melanoleuca hemoglobin subunit beta (HBB), mRNA\n",
      "ATACTTGCTTCTGACACAACTGTGTTCACTAGCAACCACAAAGAGACACCATGGTGCATC\n",
      "TGACTGGTGAGGAGAAGGCTGCCGTCACCGGCCTGTGGAGCAAGGTGAACGTGGATGAAG\n",
      "TTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTTGTCTACCCCTGGACTCAGAGGTTCTTTG\n",
      "ACTCCTTTGGGGACCTGTCCACTCCTGATGCTGTTATGAACAACCCCAAGGTCAAGGCCC\n",
      "ATGGCAAGAAGGTGCTGAACTCCTTTAGTGAAGGCCTGAAGAATCTGGACAACCTCAAGG\n",
      "GCACCTTTGTTAAGCTGAGCGAGCTGCACTGTGACAAGCTGCACGTGGATCCCGAGAACT\n",
      "TCAAGCTCCTGGGCAACGTGCTGGTGTGTGTGCTGGCTCACCACTTCGGCAAAGAGTTCA\n",
      "CCCCTCAGGTGCAGGCTGCCTATCAGAAGGTGGTGGCTGGTGTGGCCAACGCCCTGGCCC\n",
      "ACAAGTACCACTGAGCTCCTGGCCTGTTTCCTGGTGATGCCTGGAAGACCCTGTTCCCCT\n",
      "AAATTCTATCTTCTGAACTGGGGGAAATAATGTCCACCATCAAGGGTATGGTTTCTGTCT\n",
      "AATAAAGAACCTTCAGCTCAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.parse(\n",
    "    handle = './resources/HBB.gb',\n",
    "    format='gb',\n",
    ")\n",
    "\n",
    "for elements in record:\n",
    "    new_record = SeqRecord(\n",
    "        seq=elements.seq, # or manually Seq(str(any_sequence)))\n",
    "        id=elements.id, # or type in manually, e.g. \"NM_001168847\"\n",
    "        name=elements.name, # or manually, e.g. \"HBB\"\n",
    "        description=elements.description, # or manually, e.g. \"Papio anubis hemoglobin, beta (HBB), mRNA.\"\n",
    "    )\n",
    "\n",
    "    print(new_record.format('fasta'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating GenBank Files\n",
    "\n",
    "Same as for FASTA files except for the format, and more information can be provided. When it's not possible, we need to build the features.\n",
    "\n",
    "Some sequence file formats require the molecule type when writing a file, which previously was recorded with a Bio.Alphabet object as the .alphabet attribute of the Seq object. This is now recorded as a molecule type string in the SeqRecord object annotation dictionary instead. See more details about the values for `mol_type` [here](https://www.insdc.org/files/feature_table.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       NM_000518.5              628 bp    mRNA             UNK 01-JAN-1980\n",
      "DEFINITION  NM_000518.5 Homo sapiens hemoglobin subunit beta (HBB), mRNA.\n",
      "ACCESSION   NM_000518\n",
      "VERSION     NM_000518.5\n",
      "KEYWORDS    .\n",
      "SOURCE      .\n",
      "  ORGANISM  .\n",
      "            .\n",
      "FEATURES             Location/Qualifiers\n",
      "ORIGIN\n",
      "        1 acatttgctt ctgacacaac tgtgttcact agcaacctca aacagacacc atggtgcatc\n",
      "       61 tgactcctga ggagaagtct gccgttactg ccctgtgggg caaggtgaac gtggatgaag\n",
      "      121 ttggtggtga ggccctgggc aggctgctgg tggtctaccc ttggacccag aggttctttg\n",
      "      181 agtcctttgg ggatctgtcc actcctgatg ctgttatggg caaccctaag gtgaaggctc\n",
      "      241 atggcaagaa agtgctcggt gcctttagtg atggcctggc tcacctggac aacctcaagg\n",
      "      301 gcacctttgc cacactgagt gagctgcact gtgacaagct gcacgtggat cctgagaact\n",
      "      361 tcaggctcct gggcaacgtg ctggtctgtg tgctggccca tcactttggc aaagaattca\n",
      "      421 ccccaccagt gcaggctgcc tatcagaaag tggtggctgg tgtggctaat gccctggccc\n",
      "      481 acaagtatca ctaagctcgc tttcttgctg tccaatttct attaaaggtt cctttgttcc\n",
      "      541 ctaagtccaa ctactaaact gggggatatt atgaagggcc ttgagcatct ggattctgcc\n",
      "      601 taataaaaaa catttatttt cattgcaa\n",
      "//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.parse(\n",
    "    handle = 'HBB-human.fasta',\n",
    "    format='fasta',\n",
    ")\n",
    "\n",
    "for elements in record:\n",
    "    new_record = SeqRecord(\n",
    "        seq=elements.seq, # already of Seq type\n",
    "        id=elements.id,\n",
    "        name=elements.name,\n",
    "        description=elements.description,\n",
    "        annotations={\"molecule_type\": \"mRNA\"},\n",
    "    )\n",
    "\n",
    "    print(new_record.format('genbank'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       NM_000518.5              628 bp    mRNA             UNK 01-JAN-1980\n",
      "DEFINITION  NM_000518.5 Homo sapiens hemoglobin subunit beta (HBB), mRNA.\n",
      "ACCESSION   NM_000518\n",
      "VERSION     NM_000518.5\n",
      "KEYWORDS    .\n",
      "SOURCE      Homo sapiens (human)\n",
      "  ORGANISM  Homo sapiens\n",
      "            .\n",
      "REFERENCE   1\n",
      "  AUTHORS   Sebastien WIECKOWSKI\n",
      "  JOURNAL   Nature\n",
      "FEATURES             Location/Qualifiers\n",
      "     gene            1..628\n",
      "     CDS             51..494\n",
      "                     /product=\"hemoglobin subunit beta\"\n",
      "                     /protein_id=\"NP_000509.1\"\n",
      "                     /translation=\"MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFES\n",
      "                     FGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENF\n",
      "                     RLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKYH*\"\n",
      "ORIGIN\n",
      "        1 acatttgctt ctgacacaac tgtgttcact agcaacctca aacagacacc atggtgcatc\n",
      "       61 tgactcctga ggagaagtct gccgttactg ccctgtgggg caaggtgaac gtggatgaag\n",
      "      121 ttggtggtga ggccctgggc aggctgctgg tggtctaccc ttggacccag aggttctttg\n",
      "      181 agtcctttgg ggatctgtcc actcctgatg ctgttatggg caaccctaag gtgaaggctc\n",
      "      241 atggcaagaa agtgctcggt gcctttagtg atggcctggc tcacctggac aacctcaagg\n",
      "      301 gcacctttgc cacactgagt gagctgcact gtgacaagct gcacgtggat cctgagaact\n",
      "      361 tcaggctcct gggcaacgtg ctggtctgtg tgctggccca tcactttggc aaagaattca\n",
      "      421 ccccaccagt gcaggctgcc tatcagaaag tggtggctgg tgtggctaat gccctggccc\n",
      "      481 acaagtatca ctaagctcgc tttcttgctg tccaatttct attaaaggtt cctttgttcc\n",
      "      541 ctaagtccaa ctactaaact gggggatatt atgaagggcc ttgagcatct ggattctgcc\n",
      "      601 taataaaaaa catttatttt cattgcaa\n",
      "//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqFeature import SeqFeature, FeatureLocation, Reference\n",
    "\n",
    "record = SeqIO.parse(\n",
    "    handle = 'HBB-human.fasta',\n",
    "    format='fasta',\n",
    ")\n",
    "\n",
    "for elements in record:\n",
    "    # we first build each individual feature\n",
    "    f1 = SeqFeature(\n",
    "        location=FeatureLocation(0, 628, strand=1), # positive strand\n",
    "        type='gene',\n",
    "    )\n",
    "\n",
    "    f2 = SeqFeature(\n",
    "        location=FeatureLocation(50, 494, strand=1),\n",
    "        type='CDS',\n",
    "        qualifiers={\n",
    "            'product': \"hemoglobin subunit beta\",\n",
    "            'protein_id': \"NP_000509.1\",\n",
    "            'translation': elements.seq[50:494].translate(),\n",
    "        }\n",
    "    )\n",
    "\n",
    "    # reference is another class of SeqFeature\n",
    "    r1=Reference()\n",
    "    r1.authors = \"Sebastien WIECKOWSKI\"\n",
    "    r1.journal = 'Nature'\n",
    "\n",
    "    # we then add the features into the SeqRecord\n",
    "    new_record = SeqRecord(\n",
    "            seq=elements.seq, # already of Seq type\n",
    "            id=elements.id,\n",
    "            name=elements.name,\n",
    "            description=elements.description,\n",
    "            # also adding more annotations e.g. organism, source, etc.\n",
    "            annotations={\n",
    "                \"molecule_type\": \"mRNA\",\n",
    "                'organism': 'Homo sapiens',\n",
    "                'source': 'Homo sapiens (human)',\n",
    "                'references': [r1], # list of references\n",
    "                },\n",
    "            features=[f1, f2], # list of Features\n",
    "        )\n",
    "\n",
    "    print(new_record.format('genbank'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NM_000518.5\n",
      "Name: NM_000518.5\n",
      "Description: NM_000518.5 Homo sapiens hemoglobin subunit beta (HBB), mRNA\n",
      "Number of features: 2\n",
      "/molecule_type=mRNA\n",
      "/organism=Homo sapiens\n",
      "/source=Homo sapiens (human)\n",
      "/references=[Reference(title='', ...)]\n",
      "Seq('ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGG...CAA')\n"
     ]
    }
   ],
   "source": [
    "print(new_record)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Saving into a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = 'new_HBB-human.gb'\n",
    "with open(file, 'w') as output_handle:\n",
    "    SeqIO.write(\n",
    "        sequences=new_record,\n",
    "        handle=output_handle,\n",
    "        format='genbank',\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = 'new_HBB-human.fa'\n",
    "with open(file, 'w') as output_handle:\n",
    "    SeqIO.write(\n",
    "        sequences=new_record,\n",
    "        handle=output_handle,\n",
    "        format='fasta',\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "b51595f36012c8da2bf2284ebbd8d377d5c2d31fa7484164f2d2bcda7a09b462"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit (system)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
