{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence alignment\n",
    "\n",
    "## Pairwise sequence alignment\n",
    "\n",
    "Pairwise sequence alignment is one form of sequence alignment technique, where we compare only two sequences. This process involves finding the optimal alignment between the two sequences, scoring based on their similarity (how similar they are) or distance (how different they are), and then assessing the significance of this score.\n",
    "\n",
    "The basis of sequence alignment lies with the scoring process, where the two sequences are given a score on how similar (or different) they are to each other. The pairwise sequence aligning algorithms require a scoring matrix to keep track of the scores assigned. The scoring matrix assigns a positive score for a match, and a penalty for a mismatch. Three basic aspects are considered when assigning scores. They are:\n",
    "\n",
    "1. Match value — Value assigned for matching characters\n",
    "2. Mismatch value — Value assigned for mismatching characters\n",
    "3. Gap penalty — Value assigned for spaces"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Global alignment\n",
    "\n",
    "This method finds the best alignment over the entire lengths of the 2 sequences. What is the maximum similarity between sequence X and Y?\n",
    "\n",
    "One of the algorithms that uses dynamic programming to obtain global alignment is the Needleman-Wunsch algorithm. This algorithm was published by Needleman and Wunsch in 1970 for alignment of two protein sequences and it was the first application of dynamic programming to biological sequence analysis. The Needleman-Wunsch algorithm finds the best-scoring global alignment between two sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTCGGTGCCTTTAGTGATGGCCTGGCTCACCTGGACAACCTCAAGGGCACCTTTGCCACACTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGTGCTGGTCTGTGTGCTGGCCCATCACTTTGGCAAAGAATTCACCCCACCAGTGCAGGCTGCCTATCAGAAAGTGGTGGCTGGTGTGGCTAATGCCCTGGCCCACAAGTATCACTAAGCTCGCTTTCTTGCTGTCCAATTTCTATTAAAGGTTCCTTTGTTCCCTAAGTCCAACTACTAAACTGGGGGATATTATGAAGGGCCTTGAGCATCTGGATTCTGCCTAATAAAAAACATTTATTTTCATTGCAA\n",
      "TGCTTCTGACATAGTTGTGTTGACTCACAAACTCAGAAACAGACACCATGGTGCACCTGACTGATGCTGAGAAGGCTGCTGTTAATGGCCTGTGGGGAAAGGTGAACCCTGATGATGTTGGTGGCGAGGCCCTGGGCAGGCTGCTGGTTGTCTACCCTTGGACCCAGAGGTACTTTGATAGCTTTGGGGACCTGTCCTCTGCCTCTGCTATCATGGGTAACCCTAAGGTGAAGGCCCATGGCAAGAAGGTGATAAACGCCTTCAATGATGGCCTGAAACACTTGGACAACCTCAAGGGCACCTTTGCTCATCTGAGTGAACTCCACTGTGACAAGCTGCATGTGGATCCTGAGAACTTCAGGCTCCTGGGCAATATGATTGTGATTGTGTTGGGCCACCACCTGGGCAAGGAATTCACCCCCTGTGCACAGGCTGCCTTCCAGAAGGTGGTGGCTGGAGTGGCCAGTGCCCTGGCTCACAAGTACCACTAAACCTCTTTTCCTGCTCTTGTCTTTGTGCAATGGTCAATTGTTCCCAAGAGAGCATCTGTCAGTTGTTGTCAAAATGACAAAGACCTTTGAAAATCTGTCCTACTAATAAAAGGCATTTACTTTCACTGC\n"
     ]
    }
   ],
   "source": [
    "record1 = SeqIO.parse('./resources/HBB-human.fasta', 'fasta')\n",
    "record2 = SeqIO.parse('./resources/HBB-rat.fasta', 'fasta')\n",
    "\n",
    "for element in record1:\n",
    "    hbb_hum = str(element.seq)\n",
    "\n",
    "for element in record2:\n",
    "    hbb_rat = str(element.seq)\n",
    "\n",
    "print(hbb_hum)\n",
    "print(hbb_rat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on module Bio.pairwise2 in Bio:\n",
      "\n",
      "NAME\n",
      "    Bio.pairwise2 - Pairwise sequence alignment using a dynamic programming algorithm.\n",
      "\n",
      "DESCRIPTION\n",
      "    This provides functions to get global and local alignments between two\n",
      "    sequences. A global alignment finds the best concordance between all\n",
      "    characters in two sequences. A local alignment finds just the\n",
      "    subsequences that align the best. Local alignments must have a positive\n",
      "    score to be reported and they will not be extended for 'zero counting'\n",
      "    matches. This means a local alignment will always start and end with\n",
      "    a positive counting match.\n",
      "    \n",
      "    When doing alignments, you can specify the match score and gap\n",
      "    penalties.  The match score indicates the compatibility between an\n",
      "    alignment of two characters in the sequences. Highly compatible\n",
      "    characters should be given positive scores, and incompatible ones\n",
      "    should be given negative scores or 0.  The gap penalties should be\n",
      "    negative.\n",
      "    \n",
      "    The names of the alignment functions in this module follow the\n",
      "    convention\n",
      "    <alignment type>XX\n",
      "    where <alignment type> is either \"global\" or \"local\" and XX is a 2\n",
      "    character code indicating the parameters it takes.  The first\n",
      "    character indicates the parameters for matches (and mismatches), and\n",
      "    the second indicates the parameters for gap penalties.\n",
      "    \n",
      "    The match parameters are::\n",
      "    \n",
      "        CODE  DESCRIPTION & OPTIONAL KEYWORDS\n",
      "        x     No parameters. Identical characters have score of 1, otherwise 0.\n",
      "        m     A match score is the score of identical chars, otherwise mismatch\n",
      "              score. Keywords ``match``, ``mismatch``.\n",
      "        d     A dictionary returns the score of any pair of characters.\n",
      "              Keyword ``match_dict``.\n",
      "        c     A callback function returns scores. Keyword ``match_fn``.\n",
      "    \n",
      "    The gap penalty parameters are::\n",
      "    \n",
      "        CODE  DESCRIPTION & OPTIONAL KEYWORDS\n",
      "        x     No gap penalties.\n",
      "        s     Same open and extend gap penalties for both sequences.\n",
      "              Keywords ``open``, ``extend``.\n",
      "        d     The sequences have different open and extend gap penalties.\n",
      "              Keywords ``openA``, ``extendA``, ``openB``, ``extendB``.\n",
      "        c     A callback function returns the gap penalties.\n",
      "              Keywords ``gap_A_fn``, ``gap_B_fn``.\n",
      "    \n",
      "    All the different alignment functions are contained in an object\n",
      "    ``align``. For example:\n",
      "    \n",
      "        >>> from Bio import pairwise2\n",
      "        >>> alignments = pairwise2.align.globalxx(\"ACCGT\", \"ACG\")\n",
      "    \n",
      "    For better readability, the required arguments can be used with optional keywords:\n",
      "    \n",
      "        >>> alignments = pairwise2.align.globalxx(sequenceA=\"ACCGT\", sequenceB=\"ACG\")\n",
      "    \n",
      "    The result is a list of the alignments between the two strings. Each alignment\n",
      "    is a named tuple consisting of the two aligned sequences, the score and the\n",
      "    start and end positions of the alignment:\n",
      "    \n",
      "       >>> print(alignments)\n",
      "       [Alignment(seqA='ACCGT', seqB='A-CG-', score=3.0, start=0, end=5), ...\n",
      "    \n",
      "    You can access each element of an alignment by index or name:\n",
      "    \n",
      "       >>> alignments[0][2]\n",
      "       3.0\n",
      "       >>> alignments[0].score\n",
      "       3.0\n",
      "    \n",
      "    For a nice printout of an alignment, use the ``format_alignment`` method of\n",
      "    the module:\n",
      "    \n",
      "        >>> from Bio.pairwise2 import format_alignment\n",
      "        >>> print(format_alignment(*alignments[0]))\n",
      "        ACCGT\n",
      "        | || \n",
      "        A-CG-\n",
      "          Score=3\n",
      "        <BLANKLINE>\n",
      "    \n",
      "    All alignment functions have the following arguments:\n",
      "    \n",
      "    - Two sequences: strings, Biopython sequence objects or lists.\n",
      "      Lists are useful for supplying sequences which contain residues that are\n",
      "      encoded by more than one letter.\n",
      "    \n",
      "    - ``penalize_extend_when_opening``: boolean (default: False).\n",
      "      Whether to count an extension penalty when opening a gap. If false, a gap of\n",
      "      1 is only penalized an \"open\" penalty, otherwise it is penalized\n",
      "      \"open+extend\".\n",
      "    \n",
      "    - ``penalize_end_gaps``: boolean.\n",
      "      Whether to count the gaps at the ends of an alignment. By default, they are\n",
      "      counted for global alignments but not for local ones. Setting\n",
      "      ``penalize_end_gaps`` to (boolean, boolean) allows you to specify for the\n",
      "      two sequences separately whether gaps at the end of the alignment should be\n",
      "      counted.\n",
      "    \n",
      "    - ``gap_char``: string (default: ``'-'``).\n",
      "      Which character to use as a gap character in the alignment returned. If your\n",
      "      input sequences are lists, you must change this to ``['-']``.\n",
      "    \n",
      "    - ``force_generic``: boolean (default: False).\n",
      "      Always use the generic, non-cached, dynamic programming function (slow!).\n",
      "      For debugging.\n",
      "    \n",
      "    - ``score_only``: boolean (default: False).\n",
      "      Only get the best score, don't recover any alignments. The return value of\n",
      "      the function is the score. Faster and uses less memory.\n",
      "    \n",
      "    - ``one_alignment_only``: boolean (default: False).\n",
      "      Only recover one alignment.\n",
      "    \n",
      "    The other parameters of the alignment function depend on the function called.\n",
      "    Some examples:\n",
      "    \n",
      "    - Find the best global alignment between the two sequences. Identical\n",
      "      characters are given 1 point. No points are deducted for mismatches or gaps.\n",
      "    \n",
      "        >>> for a in pairwise2.align.globalxx(\"ACCGT\", \"ACG\"):\n",
      "        ...     print(format_alignment(*a))\n",
      "        ACCGT\n",
      "        | || \n",
      "        A-CG-\n",
      "          Score=3\n",
      "        <BLANKLINE>\n",
      "        ACCGT\n",
      "        || | \n",
      "        AC-G-\n",
      "          Score=3\n",
      "        <BLANKLINE>\n",
      "    \n",
      "    - Same thing as before, but with a local alignment. Note that\n",
      "      ``format_alignment`` will only show the aligned parts of the sequences,\n",
      "      together with the starting positions.\n",
      "    \n",
      "        >>> for a in pairwise2.align.localxx(\"ACCGT\", \"ACG\"):\n",
      "        ...     print(format_alignment(*a))\n",
      "        1 ACCG\n",
      "          | ||\n",
      "        1 A-CG\n",
      "          Score=3\n",
      "        <BLANKLINE>\n",
      "        1 ACCG\n",
      "          || |\n",
      "        1 AC-G\n",
      "          Score=3\n",
      "        <BLANKLINE>\n",
      "    \n",
      "      To restore the 'historic' behaviour of ``format_alignemt``, i.e., showing\n",
      "      also the un-aligned parts of both sequences, use the new keyword parameter\n",
      "      ``full_sequences``:\n",
      "    \n",
      "        >>> for a in pairwise2.align.localxx(\"ACCGT\", \"ACG\"):\n",
      "        ...     print(format_alignment(*a, full_sequences=True))\n",
      "        ACCGT\n",
      "        | || \n",
      "        A-CG-\n",
      "          Score=3\n",
      "        <BLANKLINE>\n",
      "        ACCGT\n",
      "        || | \n",
      "        AC-G-\n",
      "          Score=3\n",
      "        <BLANKLINE>\n",
      "    \n",
      "    \n",
      "    - Do a global alignment. Identical characters are given 2 points, 1 point is\n",
      "      deducted for each non-identical character. Don't penalize gaps.\n",
      "    \n",
      "        >>> for a in pairwise2.align.globalmx(\"ACCGT\", \"ACG\", 2, -1):\n",
      "        ...     print(format_alignment(*a))\n",
      "        ACCGT\n",
      "        | || \n",
      "        A-CG-\n",
      "          Score=6\n",
      "        <BLANKLINE>\n",
      "        ACCGT\n",
      "        || | \n",
      "        AC-G-\n",
      "          Score=6\n",
      "        <BLANKLINE>\n",
      "    \n",
      "    - Same as above, except now 0.5 points are deducted when opening a gap, and\n",
      "      0.1 points are deducted when extending it.\n",
      "    \n",
      "        >>> for a in pairwise2.align.globalms(\"ACCGT\", \"ACG\", 2, -1, -.5, -.1):\n",
      "        ...     print(format_alignment(*a))\n",
      "        ACCGT\n",
      "        | || \n",
      "        A-CG-\n",
      "          Score=5\n",
      "        <BLANKLINE>\n",
      "        ACCGT\n",
      "        || | \n",
      "        AC-G-\n",
      "          Score=5\n",
      "        <BLANKLINE>\n",
      "    \n",
      "    - Note that you can use keywords to increase the readability, e.g.:\n",
      "    \n",
      "        >>> a = pairwise2.align.globalms(\"ACGT\", \"ACG\", match=2, mismatch=-1, open=-.5,\n",
      "        ...                              extend=-.1)\n",
      "    \n",
      "    - Depending on the penalties, a gap in one sequence may be followed by a gap in\n",
      "      the other sequence.If you don't like this behaviour, increase the gap-open\n",
      "      penalty:\n",
      "    \n",
      "        >>> for a in pairwise2.align.globalms(\"A\", \"T\", 5, -4, -1, -.1):\n",
      "        ...     print(format_alignment(*a))\n",
      "        A-\n",
      "        <BLANKLINE>\n",
      "        -T\n",
      "          Score=-2\n",
      "        <BLANKLINE>\n",
      "        >>> for a in pairwise2.align.globalms(\"A\", \"T\", 5, -4, -3, -.1):\n",
      "        ...     print(format_alignment(*a))\n",
      "        A\n",
      "        .\n",
      "        T\n",
      "          Score=-4\n",
      "        <BLANKLINE>\n",
      "    \n",
      "    - The alignment function can also use known matrices already included in\n",
      "      Biopython (in ``Bio.Align.substitution_matrices``):\n",
      "    \n",
      "        >>> from Bio.Align import substitution_matrices\n",
      "        >>> matrix = substitution_matrices.load(\"BLOSUM62\")\n",
      "        >>> for a in pairwise2.align.globaldx(\"KEVLA\", \"EVL\", matrix):\n",
      "        ...     print(format_alignment(*a))\n",
      "        KEVLA\n",
      "         ||| \n",
      "        -EVL-\n",
      "          Score=13\n",
      "        <BLANKLINE>\n",
      "    \n",
      "    - With the parameter ``c`` you can define your own match- and gap functions.\n",
      "      E.g. to define an affine logarithmic gap function and using it:\n",
      "    \n",
      "        >>> from math import log\n",
      "        >>> def gap_function(x, y):  # x is gap position in seq, y is gap length\n",
      "        ...     if y == 0:  # No gap\n",
      "        ...         return 0\n",
      "        ...     elif y == 1:  # Gap open penalty\n",
      "        ...         return -2\n",
      "        ...     return - (2 + y/4.0 + log(y)/2.0)\n",
      "        ...\n",
      "        >>> alignment = pairwise2.align.globalmc(\"ACCCCCGT\", \"ACG\", 5, -4,\n",
      "        ...                                      gap_function, gap_function)\n",
      "    \n",
      "      You can define different gap functions for each sequence.\n",
      "      Self-defined match functions must take the two residues to be compared and\n",
      "      return a score.\n",
      "    \n",
      "    To see a description of the parameters for a function, please look at\n",
      "    the docstring for the function via the help function, e.g.\n",
      "    type ``help(pairwise2.align.localds)`` at the Python prompt.\n",
      "\n",
      "CLASSES\n",
      "    builtins.object\n",
      "        affine_penalty\n",
      "        dictionary_match\n",
      "        identity_match\n",
      "    \n",
      "    class affine_penalty(builtins.object)\n",
      "     |  affine_penalty(open, extend, penalize_extend_when_opening=0)\n",
      "     |  \n",
      "     |  Create a gap function for use in an alignment.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, index, length)\n",
      "     |      Call a gap function instance already created.\n",
      "     |  \n",
      "     |  __init__(self, open, extend, penalize_extend_when_opening=0)\n",
      "     |      Initialize the class.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class dictionary_match(builtins.object)\n",
      "     |  dictionary_match(score_dict, symmetric=1)\n",
      "     |  \n",
      "     |  Create a match function for use in an alignment.\n",
      "     |  \n",
      "     |  Attributes:\n",
      "     |   - score_dict     - A dictionary where the keys are tuples (residue 1,\n",
      "     |     residue 2) and the values are the match scores between those residues.\n",
      "     |   - symmetric      - A flag that indicates whether the scores are symmetric.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, charA, charB)\n",
      "     |      Call a dictionary match instance already created.\n",
      "     |  \n",
      "     |  __init__(self, score_dict, symmetric=1)\n",
      "     |      Initialize the class.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class identity_match(builtins.object)\n",
      "     |  identity_match(match=1, mismatch=0)\n",
      "     |  \n",
      "     |  Create a match function for use in an alignment.\n",
      "     |  \n",
      "     |  match and mismatch are the scores to give when two residues are equal\n",
      "     |  or unequal.  By default, match is 1 and mismatch is 0.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, charA, charB)\n",
      "     |      Call a match function instance already created.\n",
      "     |  \n",
      "     |  __init__(self, match=1, mismatch=0)\n",
      "     |      Initialize the class.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "\n",
      "FUNCTIONS\n",
      "    calc_affine_penalty(length, open, extend, penalize_extend_when_opening)\n",
      "        Calculate a penality score for the gap function.\n",
      "    \n",
      "    format_alignment(align1, align2, score, begin, end, full_sequences=False)\n",
      "        Format the alignment prettily into a string.\n",
      "        \n",
      "        IMPORTANT: Gap symbol must be \"-\" (or ['-'] for lists)!\n",
      "        \n",
      "        Since Biopython 1.71 identical matches are shown with a pipe\n",
      "        character, mismatches as a dot, and gaps as a space.\n",
      "        \n",
      "        Prior releases just used the pipe character to indicate the\n",
      "        aligned region (matches, mismatches and gaps).\n",
      "        \n",
      "        Also, in local alignments, if the alignment does not include\n",
      "        the whole sequences, now only the aligned part is shown,\n",
      "        together with the start positions of the aligned subsequences.\n",
      "        The start positions are 1-based; so start position n is the\n",
      "        n-th base/amino acid in the *un-aligned* sequence.\n",
      "        \n",
      "        NOTE: This is different to the alignment's begin/end values,\n",
      "        which give the Python indices (0-based) of the bases/amino acids\n",
      "        in the *aligned* sequences.\n",
      "        \n",
      "        If you want to restore the 'historic' behaviour, that means\n",
      "        displaying the whole sequences (including the non-aligned parts),\n",
      "        use ``full_sequences=True``. In this case, the non-aligned leading\n",
      "        and trailing parts are also indicated by spaces in the match-line.\n",
      "    \n",
      "    print_matrix(matrix)\n",
      "        Print out a matrix for debugging purposes.\n",
      "    \n",
      "    rint(...)\n",
      "\n",
      "DATA\n",
      "    MAX_ALIGNMENTS = 1000\n",
      "    align = <Bio.pairwise2.align object>\n",
      "\n",
      "FILE\n",
      "    c:\\users\\sébastien\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\bio\\pairwise2.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(pairwise2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on align in module Bio.pairwise2 object:\n",
      "\n",
      "class align(builtins.object)\n",
      " |  Provide functions that do alignments.\n",
      " |  \n",
      " |  Alignment functions are called as:\n",
      " |  \n",
      " |    pairwise2.align.globalXX\n",
      " |  \n",
      " |  or\n",
      " |  \n",
      " |    pairwise2.align.localXX\n",
      " |  \n",
      " |  Where XX is a 2 character code indicating the match/mismatch parameters\n",
      " |  (first character, either x, m, d or c) and the gap penalty parameters\n",
      " |  (second character, either x, s, d, or c).\n",
      " |  \n",
      " |  For a detailed description read the main module's docstring (e.g.,\n",
      " |  type ``help(pairwise2)``).\n",
      " |  To see a description of the parameters for a function, please\n",
      " |  look at the docstring for the function, e.g. type\n",
      " |  ``help(pairwise2.align.localds)`` at the Python prompt.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __getattr__(self, attr)\n",
      " |      Call alignment_function() to check and decode the attributes.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  alignment_function = <class 'Bio.pairwise2.align.alignment_function'>\n",
      " |      Callable class which impersonates an alignment function.\n",
      " |      \n",
      " |      The constructor takes the name of the function.  This class\n",
      " |      will decode the name of the function to figure out how to\n",
      " |      interpret the parameters.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(pairwise2.align)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on alignment_function in module Bio.pairwise2 object:\n",
      "\n",
      "globalxs = class alignment_function(builtins.object)\n",
      " |  globalxs(name)\n",
      " |  \n",
      " |  globalxs(sequenceA, sequenceB, open, extend) -> alignments\n",
      " |  \n",
      " |  The following parameters can also be used with optional\n",
      " |  keywords of the same name.\n",
      " |  \n",
      " |  \n",
      " |  sequenceA and sequenceB must be of the same type, either\n",
      " |  strings, lists or Biopython sequence objects.\n",
      " |  \n",
      " |  \n",
      " |  open and extend are the gap penalties when a gap is\n",
      " |  opened and extended.  They should be negative.\n",
      " |  \n",
      " |  alignments is a list of named tuples (seqA, seqB, score,\n",
      " |  begin, end). seqA and seqB are strings showing the alignment\n",
      " |  between the sequences.  score is the score of the alignment.\n",
      " |  begin and end are indexes of seqA and seqB that indicate\n",
      " |  where the alignment occurs.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __call__(self, *args, **keywds)\n",
      " |      Call the alignment instance already created.\n",
      " |  \n",
      " |  __init__(self, name)\n",
      " |      Check to make sure the name of the function is reasonable.\n",
      " |  \n",
      " |  decode(self, *args, **keywds)\n",
      " |      Decode the arguments for the _align function.\n",
      " |      \n",
      " |      keywds will get passed to it, so translate the arguments\n",
      " |      to this function into forms appropriate for _align.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  match2args = {'c': (['match_fn'], 'match_fn is a callback function tha...\n",
      " |  \n",
      " |  penalty2args = {'c': (['gap_A_fn', 'gap_B_fn'], 'gap_A_fn and gap_B_fn...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(pairwise2.align.globalxs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parameters\n",
    "\n",
    "The [pairwise2 module](http://biopython.org/DIST/docs/api/Bio.pairwise2-module.html) provides alignment functions to get global and local alignments between two sequences. Calling the function `<alignment type>XX` is either global or local and `XX` is a 2 character code indicating the parameters it takes. The first character indicates the parameters for **matches** (and mismatches), and the second indicates the parameters for **gap penalties**.\n",
    "\n",
    "##### Match parameters\n",
    "\n",
    "* x - No parameters. Identical characters have score of 1, else 0.\n",
    "* m - A match score is the score of identical chars, else mismatch score.\n",
    "* d - A dictionary returns the score of any pair of characters.\n",
    "* c - A callback function returns scores.\n",
    "\n",
    "##### Gap penalty parameters\n",
    "\n",
    "* x - No gap penalties.\n",
    "* s - Same open and extend gap penalties for both sequences.\n",
    "* d - The sequences have different open and extend gap penalties.\n",
    "* c - A callback function returns the gap penalties."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment(seqA='ACATTTGCTTCTGACACA-AC-T-GTGTTC-ACT-AGCAAC-CTCA-AA-CAGACACCATGGTGCATC-TGACTCCTGA-G--GAGAAGT-CTGCC-GTTAC-TGC-CCTGTGGGGCAA-GGTGAACG--TGGATGAA-GTTGGTGGT-GAGGCCCTGGGCAGGCTGCTGGTG-GTCTACCCTTGGACCCAGAGGTT-CTTTGAGTC--CTTTGGGGATC-TGTCCACTCCTGA--TGCTG-T-T-ATGGGC-AACCCTAAGGTGAAGGCTC-ATGGCAAGAAAGTGCTCGG-T----GCCTTT-AG-TGATGGCCTGGCTCA--C-CT-GGACAACCTCAAGGGCACCTTTGCCA-CA-CTGAGTGAG-CTGC-ACTGTGACAAGCTGCAC-GTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGT--GC-T-G-G-TCTGTGTGCTGGC-CCATC-ACTT-TGG-CAAAG-AATTCACCCCACCA-GTGCA--GGCTGCCTATC-AGAAAG-TGGTGGCTGGT-GTGGCTA-A-TGCCCTGGCC-CACAAGTATC-ACTAAGCT-CGCT-TTCTTGC-TGTC-CAATT-TCTATTA----AA-GGTTCCT--TTGTTCCCTAAGTCCA-A-CTACTAAACTGGG---G--GATAT-T-A---TGA-AGG--G-CCTT-GAGCATCTGGAT-TCTG-CCTAA-TAA-AAAA--CATTTAT-TTTCAT-TGCAA', seqB='-----TGCTTCTG--ACATA-GTTGTGTT-GACTCA-CAA-ACTCAGAAACAGACACCATGGTGCA-CCTGA---CTGATGCTGAGAAG-GCTG-CTGTTA-ATG-GCCTGTGGGG-AAAGGTGAAC-CCT-GATG-ATGTTGGTGG-CGAGGCCCTGGGCAGGCTGCTGGT-TGTCTACCCTTGGACCCAGAGG-TACTTTGA-T-AGCTTTGGGGA-CCTGT-C-CT-CTG-CCT-CTGCTATCATGGG-TAACCCTAAGGTGAAGGC-CCATGGCAAG-AAG-G-T--GATAAACGCC-TTCA-ATGATGGCCT-G---AAACACTTGGACAACCTCAAGGGCACCTTTG-C-TCATCTGAGTGA-ACT-CCACTGTGACAAGCTGCA-TGTGGATCCTGAGAACTTCAGGCTCCTGGGCAA--TATG-ATTGTGAT-TGTGT--TGG-GCCA-CCAC--CTGGGC-AAGGAATTCA-CCC-CC-TGTGCACAGGCTGCCT-TCCAG-AAGGTGGTGGCTGG-AGTGGC--CAGTGCCCTGG-CTCACAAGTA-CCACTAA---AC-CTCTT-TT-CCTG-CTC--TTGTCT-TT-GTGCAATGG-T-C-AATTGTTCCC-AAG---AGAGC-A-T---CT--GTCAGTTG-T-TGTCAAAATGACA--AAGACCTTTGA--A-----A-ATCTGTCCT-ACTAATAAAAGGCATTTA-CTTTCA-CTGC--', score=513.0, start=0, end=735)\n"
     ]
    }
   ],
   "source": [
    "# `xx` means no parameters, i.e. same character=1 else 0; no gap penalty\n",
    "global_alignment = pairwise2.align.globalxx(hbb_hum, hbb_rat)\n",
    "print(global_alignment[0]) # returns a list of alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment(seqA='ACATTTGCTTCTGACACA-AC-T-GTGTTC-ACT-AGCAAC-CTCA-AA-CAGACACCATGGTGCATC-TGACTCCTGA-G--GAGAAGT-CTGCC-GTTAC-TGC-CCTGTGGGGCAA-GGTGAACG--TGGATGAA-GTTGGTGGT-GAGGCCCTGGGCAGGCTGCTGGTG-GTCTACCCTTGGACCCAGAGGTT-CTTTGAGTC--CTTTGGGGATC-TGTCCACTCCTGA--TGCTG-T-T-ATGGGC-AACCCTAAGGTGAAGGCTC-ATGGCAAGAAAGTGCTCGG-T----GCCTTT-AG-TGATGGCCTGGCTCA--C-CT-GGACAACCTCAAGGGCACCTTTGCCA-CA-CTGAGTGAG-CTGC-ACTGTGACAAGCTGCAC-GTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGT--GC-T-G-G-TCTGTGTGCTGGC-CCATC-ACTT-TGG-CAAAG-AATTCACCCCACCA-GTGCA--GGCTGCCTATC-AGAAAG-TGGTGGCTGGT-GTGGCTA-A-TGCCCTGGCC-CACAAGTATC-ACTAAGCT-CGCT-TTCTTGC-TGTC-CAATT-TCTATTA----AA-GGTTCCT--TTGTTCCCTAAGTCCA-A-CTACTAAACTGGG---G--GATAT-T-A---TGA-AGG--G-CCTT-GAGCATCTGGAT-TCTG-CCTAA-TAA-AAAA--CATTTAT-TTTCAT-TGCAA', seqB='-----TGCTTCTG--ACATA-GTTGTGTT-GACTCA-CAA-ACTCAGAAACAGACACCATGGTGCA-CCTGA---CTGATGCTGAGAAG-GCTG-CTGTTA-ATG-GCCTGTGGGG-AAAGGTGAAC-CCT-GATG-ATGTTGGTGG-CGAGGCCCTGGGCAGGCTGCTGGT-TGTCTACCCTTGGACCCAGAGG-TACTTTGA-T-AGCTTTGGGGA-CCTGT-C-CT-CTG-CCT-CTGCTATCATGGG-TAACCCTAAGGTGAAGGC-CCATGGCAAG-AAG-G-T--GATAAACGCC-TTCA-ATGATGGCCT-G---AAACACTTGGACAACCTCAAGGGCACCTTTG-C-TCATCTGAGTGA-ACT-CCACTGTGACAAGCTGCA-TGTGGATCCTGAGAACTTCAGGCTCCTGGGCAA--TATG-ATTGTGAT-TGTGT--TGG-GCCA-CCAC--CTGGGC-AAGGAATTCA-CCC-CC-TGTGCACAGGCTGCCT-TCCAG-AAGGTGGTGGCTGG-AGTGGC--CAGTGCCCTGG-CTCACAAGTA-CCACTAA---AC-CTCTT-TT-CCTG-CTC--TTGTCT-TT-GTGCAATGG-T-C-AATTGTTCCC-AAG---AGAGC-A-T---CT--GTCAGTTG-T-TGTCAAAATGACA--AAGACCTTTGA--A-----A-ATCTGTCCT-ACTAATAAAAGGCATTTA-CTTTCA-CTGC--', score=513.0, start=0, end=735)\n",
      "Alignment(seqA='ACATTTGCTTCTGACACA-AC-T-GTGTTC-ACT-AGCAAC-CTCA-AA-CAGACACCATGGTGCATC-TGACTCCTGA-G--GAGAAGT-CTGCC-GTTAC-TGC-CCTGTGGGGCAA-GGTGAACG--TGGATGAA-GTTGGTGGT-GAGGCCCTGGGCAGGCTGCTGGTG-GTCTACCCTTGGACCCAGAGGTT-CTTTGAGTC--CTTTGGGGATC-TGTCCACTCCTGA--TGCTG-T-T-ATGGGC-AACCCTAAGGTGAAGGCTC-ATGGCAAGAAAGTGCTCGG-T----GCCTTT-AG-TGATGGCCTGGCTCA--C-CT-GGACAACCTCAAGGGCACCTTTGCCA-CA-CTGAGTGAG-CTGC-ACTGTGACAAGCTGCAC-GTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGT--GC-T-G-G-TCTGTGTGCTGGC-CCATC-ACTT-TGG-CAAAG-AATTCACCCCACCA-GTGCA--GGCTGCCTATC-AGAAAG-TGGTGGCTGGT-GTGGCTA-A-TGCCCTGGCC-CACAAGTATC-ACTAAGCT-CGCT-TTCTTGC-TGTC-CAATT-TCTATTA----AA-GGTTCCT--TTGTTCCCTAAGTCCA-A-CTACTAAACTGGG---G--GATAT-T-A---TGA-AGG--G-CCTT-GAGCATCTGGAT-TCTG-CCTAA-TAA-AAAA--CATTTAT-TTTCAT-TGCAA', seqB='----T-GCTTCTG--ACATA-GTTGTGTT-GACTCA-CAA-ACTCAGAAACAGACACCATGGTGCA-CCTGA---CTGATGCTGAGAAG-GCTG-CTGTTA-ATG-GCCTGTGGGG-AAAGGTGAAC-CCT-GATG-ATGTTGGTGG-CGAGGCCCTGGGCAGGCTGCTGGT-TGTCTACCCTTGGACCCAGAGG-TACTTTGA-T-AGCTTTGGGGA-CCTGT-C-CT-CTG-CCT-CTGCTATCATGGG-TAACCCTAAGGTGAAGGC-CCATGGCAAG-AAG-G-T--GATAAACGCC-TTCA-ATGATGGCCT-G---AAACACTTGGACAACCTCAAGGGCACCTTTG-C-TCATCTGAGTGA-ACT-CCACTGTGACAAGCTGCA-TGTGGATCCTGAGAACTTCAGGCTCCTGGGCAA--TATG-ATTGTGAT-TGTGT--TGG-GCCA-CCAC--CTGGGC-AAGGAATTCA-CCC-CC-TGTGCACAGGCTGCCT-TCCAG-AAGGTGGTGGCTGG-AGTGGC--CAGTGCCCTGG-CTCACAAGTA-CCACTAA---AC-CTCTT-TT-CCTG-CTC--TTGTCT-TT-GTGCAATGG-T-C-AATTGTTCCC-AAG---AGAGC-A-T---CT--GTCAGTTG-T-TGTCAAAATGACA--AAGACCTTTGA--A-----A-ATCTGTCCT-ACTAATAAAAGGCATTTA-CTTTCA-CTGC--', score=513.0, start=0, end=735)\n",
      "Alignment(seqA='ACATTTGCTTCTGACACA-AC-T-GTGTTC-ACT-AGCAAC-CTCA-AA-CAGACACCATGGTGCATC-TGACTCCTGA-G--GAGAAGT-CTGCC-GTTAC-TGC-CCTGTGGGGCAA-GGTGAACG--TGGATGAA-GTTGGTGGT-GAGGCCCTGGGCAGGCTGCTGGTG-GTCTACCCTTGGACCCAGAGGTT-CTTTGAGTC--CTTTGGGGATC-TGTCCACTCCTGA--TGCTG-T-T-ATGGGC-AACCCTAAGGTGAAGGCTC-ATGGCAAGAAAGTGCTCGG-T----GCCTTT-AG-TGATGGCCTGGCTCA--C-CT-GGACAACCTCAAGGGCACCTTTGCCA-CA-CTGAGTGAG-CTGC-ACTGTGACAAGCTGCAC-GTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGT--GC-T-G-G-TCTGTGTGCTGGC-CCATC-ACTT-TGG-CAAAG-AATTCACCCCACCA-GTGCA--GGCTGCCTATC-AGAAAG-TGGTGGCTGGT-GTGGCTA-A-TGCCCTGGCC-CACAAGTATC-ACTAAGCT-CGCT-TTCTTGC-TGTC-CAATT-TCTATTA----AA-GGTTCCT--TTGTTCCCTAAGTCCA-A-CTACTAAACTGGG---G--GATAT-T-A---TGA-AGG--G-CCTT-GAGCATCTGGAT-TCTG-CCTAA-TAA-AAAA--CATTTAT-TTTCAT-TGCAA', seqB='---T--GCTTCTG--ACATA-GTTGTGTT-GACTCA-CAA-ACTCAGAAACAGACACCATGGTGCA-CCTGA---CTGATGCTGAGAAG-GCTG-CTGTTA-ATG-GCCTGTGGGG-AAAGGTGAAC-CCT-GATG-ATGTTGGTGG-CGAGGCCCTGGGCAGGCTGCTGGT-TGTCTACCCTTGGACCCAGAGG-TACTTTGA-T-AGCTTTGGGGA-CCTGT-C-CT-CTG-CCT-CTGCTATCATGGG-TAACCCTAAGGTGAAGGC-CCATGGCAAG-AAG-G-T--GATAAACGCC-TTCA-ATGATGGCCT-G---AAACACTTGGACAACCTCAAGGGCACCTTTG-C-TCATCTGAGTGA-ACT-CCACTGTGACAAGCTGCA-TGTGGATCCTGAGAACTTCAGGCTCCTGGGCAA--TATG-ATTGTGAT-TGTGT--TGG-GCCA-CCAC--CTGGGC-AAGGAATTCA-CCC-CC-TGTGCACAGGCTGCCT-TCCAG-AAGGTGGTGGCTGG-AGTGGC--CAGTGCCCTGG-CTCACAAGTA-CCACTAA---AC-CTCTT-TT-CCTG-CTC--TTGTCT-TT-GTGCAATGG-T-C-AATTGTTCCC-AAG---AGAGC-A-T---CT--GTCAGTTG-T-TGTCAAAATGACA--AAGACCTTTGA--A-----A-ATCTGTCCT-ACTAATAAAAGGCATTTA-CTTTCA-CTGC--', score=513.0, start=0, end=735)\n"
     ]
    }
   ],
   "source": [
    "for a in global_alignment[:3]:\n",
    "    print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACATTTGCTTCTGACACA-AC-T-GTGTTC-ACT-AGCAAC-CTCA-AA-CAGACACCATGGTGCATC-TGACTCCTGA-G--GAGAAGT-CTGCC-GTTAC-TGC-CCTGTGGGGCAA-GGTGAACG--TGGATGAA-GTTGGTGGT-GAGGCCCTGGGCAGGCTGCTGGTG-GTCTACCCTTGGACCCAGAGGTT-CTTTGAGTC--CTTTGGGGATC-TGTCCACTCCTGA--TGCTG-T-T-ATGGGC-AACCCTAAGGTGAAGGCTC-ATGGCAAGAAAGTGCTCGG-T----GCCTTT-AG-TGATGGCCTGGCTCA--C-CT-GGACAACCTCAAGGGCACCTTTGCCA-CA-CTGAGTGAG-CTGC-ACTGTGACAAGCTGCAC-GTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGT--GC-T-G-G-TCTGTGTGCTGGC-CCATC-ACTT-TGG-CAAAG-AATTCACCCCACCA-GTGCA--GGCTGCCTATC-AGAAAG-TGGTGGCTGGT-GTGGCTA-A-TGCCCTGGCC-CACAAGTATC-ACTAAGCT-CGCT-TTCTTGC-TGTC-CAATT-TCTATTA----AA-GGTTCCT--TTGTTCCCTAAGTCCA-A-CTACTAAACTGGG---G--GATAT-T-A---TGA-AGG--G-CCTT-GAGCATCTGGAT-TCTG-CCTAA-TAA-AAAA--CATTTAT-TTTCAT-TGCAA\n",
      "     ||||||||  ||| |  | |||||  ||| | |||  |||| || |||||||||||||||| | |||   |||| |  ||||||  ||| | ||||  ||  ||||||||| || |||||||   | |||| | ||||||||  |||||||||||||||||||||||  ||||||||||||||||||||| | |||||| |   ||||||||| | ||| | || |||   | ||| | | |||||  ||||||||||||||||| | |||||||| ||| | |  | |    ||| || |  ||||||||| |   |  | || ||||||||||||||||||||||| |  || ||||||||  || | ||||||||||||||||  ||||||||||||||||||||||||||||||||  |  |  | | | | |||||  |||  ||| | ||   ||| | ||| |||||| ||| ||  |||||  |||||||| || || ||| ||||||||||  |||||   | |||||||| | |||||||| | |||||    | || || || | || | |  || ||| ||     || || | |   |||||||| |||   | | | | |   ||  |   |  | | | | |   ||| |    | |||| ||  |     |  |||| ||| | ||| ||||  ||||||  |||||  |||  \n",
      "-----TGCTTCTG--ACATA-GTTGTGTT-GACTCA-CAA-ACTCAGAAACAGACACCATGGTGCA-CCTGA---CTGATGCTGAGAAG-GCTG-CTGTTA-ATG-GCCTGTGGGG-AAAGGTGAAC-CCT-GATG-ATGTTGGTGG-CGAGGCCCTGGGCAGGCTGCTGGT-TGTCTACCCTTGGACCCAGAGG-TACTTTGA-T-AGCTTTGGGGA-CCTGT-C-CT-CTG-CCT-CTGCTATCATGGG-TAACCCTAAGGTGAAGGC-CCATGGCAAG-AAG-G-T--GATAAACGCC-TTCA-ATGATGGCCT-G---AAACACTTGGACAACCTCAAGGGCACCTTTG-C-TCATCTGAGTGA-ACT-CCACTGTGACAAGCTGCA-TGTGGATCCTGAGAACTTCAGGCTCCTGGGCAA--TATG-ATTGTGAT-TGTGT--TGG-GCCA-CCAC--CTGGGC-AAGGAATTCA-CCC-CC-TGTGCACAGGCTGCCT-TCCAG-AAGGTGGTGGCTGG-AGTGGC--CAGTGCCCTGG-CTCACAAGTA-CCACTAA---AC-CTCTT-TT-CCTG-CTC--TTGTCT-TT-GTGCAATGG-T-C-AATTGTTCCC-AAG---AGAGC-A-T---CT--GTCAGTTG-T-TGTCAAAATGACA--AAGACCTTTGA--A-----A-ATCTGTCCT-ACTAATAAAAGGCATTTA-CTTTCA-CTGC--\n",
      "  Score=513\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(format_alignment(*global_alignment[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Local alignment\n",
    "\n",
    "This method finds the most similar subsequences among the 2 sequences. What is the maximum similarity between a subsequence of X and a subsequence of Y?\n",
    "\n",
    "One of the algorithms that uses dynamic programming to obtain local alignments within two given sequences is the Smith-Waterman algorithm. Smith and Waterman published an application of dynamic programming to find the optimal local alignments in 1981. This algorithm is similar to Needleman-Wunsch algorithm, but there are slight differences with the scoring process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6 TGCTTCTGACACA-AC-T-GTGTTC-ACT-AGCAAC-CTCA-AA-CAGACACCATGGTGCATC-TGACTCCTGA-G--GAGAAGT-CTGCC-GTTAC-TGC-CCTGTGGGGCAA-GGTGAACG--TGGATGAA-GTTGGTGGT-GAGGCCCTGGGCAGGCTGCTGGTG-GTCTACCCTTGGACCCAGAGGTT-CTTTGAGTC--CTTTGGGGATC-TGTCCACTCCTGA--TGCTG-T-T-ATGGGC-AACCCTAAGGTGAAGGCTC-ATGGCAAGAAAGTGCTCGG-T----GCCTTT-AG-TGATGGCCTGGCTCA--C-CT-GGACAACCTCAAGGGCACCTTTGCCA-CA-CTGAGTGAG-CTGC-ACTGTGACAAGCTGCAC-GTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGT--GC-T-G-G-TCTGTGTGCTGGC-CCATC-ACTT-TGG-CAAAG-AATTCACCCCACCA-GTGCA--GGCTGCCTATC-AGAAAG-TGGTGGCTGGT-GTGGCTA-A-TGCCCTGGCC-CACAAGTATC-ACTAAGCT-CGCT-TTCTTGC-TGTC-CAATT-TCTATTA----AA-GGTTCCT--TTGTTCCCTAAGTCCA-A-CTACTAAACTGGG---G--GATAT-T-A---TGA-AGG--G-CCTT-GAGCATCTGGAT-TCTG-CCTAA-TAA-AAAA--CATTTAT-TTTCAT-TGC\n",
      "  ||||||||  ||| |  | |||||  ||| | |||  |||| || |||||||||||||||| | |||   |||| |  ||||||  ||| | ||||  ||  ||||||||| || |||||||   | |||| | ||||||||  |||||||||||||||||||||||  ||||||||||||||||||||| | |||||| |   ||||||||| | ||| | || |||   | ||| | | |||||  ||||||||||||||||| | |||||||| ||| | |  | |    ||| || |  ||||||||| |   |  | || ||||||||||||||||||||||| |  || ||||||||  || | ||||||||||||||||  ||||||||||||||||||||||||||||||||  |  |  | | | | |||||  |||  ||| | ||   ||| | ||| |||||| ||| ||  |||||  |||||||| || || ||| ||||||||||  |||||   | |||||||| | |||||||| | |||||    | || || || | || | |  || ||| ||     || || | |   |||||||| |||   | | | | |   ||  |   |  | | | | |   ||| |    | |||| ||  |     |  |||| ||| | ||| ||||  ||||||  |||||  |||\n",
      "1 TGCTTCTG--ACATA-GTTGTGTT-GACTCA-CAA-ACTCAGAAACAGACACCATGGTGCA-CCTGA---CTGATGCTGAGAAG-GCTG-CTGTTA-ATG-GCCTGTGGGG-AAAGGTGAAC-CCT-GATG-ATGTTGGTGG-CGAGGCCCTGGGCAGGCTGCTGGT-TGTCTACCCTTGGACCCAGAGG-TACTTTGA-T-AGCTTTGGGGA-CCTGT-C-CT-CTG-CCT-CTGCTATCATGGG-TAACCCTAAGGTGAAGGC-CCATGGCAAG-AAG-G-T--GATAAACGCC-TTCA-ATGATGGCCT-G---AAACACTTGGACAACCTCAAGGGCACCTTTG-C-TCATCTGAGTGA-ACT-CCACTGTGACAAGCTGCA-TGTGGATCCTGAGAACTTCAGGCTCCTGGGCAA--TATG-ATTGTGAT-TGTGT--TGG-GCCA-CCAC--CTGGGC-AAGGAATTCA-CCC-CC-TGTGCACAGGCTGCCT-TCCAG-AAGGTGGTGGCTGG-AGTGGC--CAGTGCCCTGG-CTCACAAGTA-CCACTAA---AC-CTCTT-TT-CCTG-CTC--TTGTCT-TT-GTGCAATGG-T-C-AATTGTTCCC-AAG---AGAGC-A-T---CT--GTCAGTTG-T-TGTCAAAATGACA--AAGACCTTTGA--A-----A-ATCTGTCCT-ACTAATAAAAGGCATTTA-CTTTCA-CTGC\n",
      "  Score=513\n",
      "\n"
     ]
    }
   ],
   "source": [
    "local_alignment = pairwise2.align.localxx(hbb_hum, hbb_rat)\n",
    "print(format_alignment(*local_alignment[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on alignment_function in module Bio.pairwise2 object:\n",
      "\n",
      "localms = class alignment_function(builtins.object)\n",
      " |  localms(name)\n",
      " |  \n",
      " |  localms(sequenceA, sequenceB, match, mismatch, open, extend) -> alignments\n",
      " |  \n",
      " |  The following parameters can also be used with optional\n",
      " |  keywords of the same name.\n",
      " |  \n",
      " |  \n",
      " |  sequenceA and sequenceB must be of the same type, either\n",
      " |  strings, lists or Biopython sequence objects.\n",
      " |  \n",
      " |  \n",
      " |  match is the score to given to identical characters.\n",
      " |  mismatch is the score given to non-identical ones.\n",
      " |  \n",
      " |  open and extend are the gap penalties when a gap is\n",
      " |  opened and extended.  They should be negative.\n",
      " |  \n",
      " |  alignments is a list of named tuples (seqA, seqB, score,\n",
      " |  begin, end). seqA and seqB are strings showing the alignment\n",
      " |  between the sequences.  score is the score of the alignment.\n",
      " |  begin and end are indexes of seqA and seqB that indicate\n",
      " |  where the alignment occurs.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __call__(self, *args, **keywds)\n",
      " |      Call the alignment instance already created.\n",
      " |  \n",
      " |  __init__(self, name)\n",
      " |      Check to make sure the name of the function is reasonable.\n",
      " |  \n",
      " |  decode(self, *args, **keywds)\n",
      " |      Decode the arguments for the _align function.\n",
      " |      \n",
      " |      keywds will get passed to it, so translate the arguments\n",
      " |      to this function into forms appropriate for _align.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  match2args = {'c': (['match_fn'], 'match_fn is a callback function tha...\n",
      " |  \n",
      " |  penalty2args = {'c': (['gap_A_fn', 'gap_B_fn'], 'gap_A_fn and gap_B_fn...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(pairwise2.align.localms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment(seqA='ACATTTGCTTCTGACACA-AC--TGTGTTC-ACT-AGCAAC-CTC--AAACAGACACCATGGTGCATC-TGACTCCTGA-G--GAGAAGT-CTGCC-GTTAC-TGC-CCTGTGGGGCAA-GGTGAA--CGTGGATGAA-GTTGGTGGT-GAGGCCCTGGGCAGGCTGCTGGTG-GTCTACCCTTGGACCCAGAGGTT-CTTTG--AGTCCTTTGGGGATC-TGTCCA-CT-CCTGATGCTG-T-T-ATGGGC-AACCCTAAGGTGAAGGCTC-ATGGCAAGAAAG-TGC-T---CGGTGCCTTT--AGTGATGGCCTGGCTC--AC-CT-GGACAACCTCAAGGGCACCTTTGCCA-CA-CTGAGTGAG-CTGC-ACTGTGACAAGCTGCAC-GTGGATCCTGAGAACTTCAGGCTCCTGGGCAAC---G-TGCTG-G-TCTGTGTGCTGG-CC-CATCACTT-TGG-CAAAG-AATTCACCCCACCA-GTG--CAGGCTGCCTATC-AGAAAG-TGGTGGCTGGT-GTGGCTA-A-TGCCCTGGCC-CACAAGTATC-ACTAAG--CTCGCTTT-CTTGCTG-TCCAATT-TCTATTA----AA-GGTTCCT--TTGTTCCCTAAGTCCA-A-CTACTAAACTG---G--G--GGAT---ATTATG---AAGGG-CCTT-G--AGCATCTGGATTCTGC--CTAATAAAAAA--CATTTAT-TTTCAT-TGCAA', seqB='-----TGCTTCTG--ACATA-GTTGTGTT-GACTCA-CAA-ACTCAGAAACAGACACCATGGTGCA-CCTGA---CTGATGCTGAGAAG-GCTG-CTGTTA-ATG-GCCTGTGGGG-AAAGGTGAACCC-T-GATG-ATGTTGGTGG-CGAGGCCCTGGGCAGGCTGCTGGT-TGTCTACCCTTGGACCCAGAGG-TACTTTGATAG--CTTTGGGGA-CCTGTCC-TCTGCCT----CTGCTATCATGGG-TAACCCTAAGGTGAAGGC-CCATGGCAAG-AAGGTG-ATAAAC---GCC-TTCAA-TGATGGCCTG----AAACACTTGGACAACCTCAAGGGCACCTTTGC--TCATCTGAGTGA-ACT-CCACTGTGACAAGCTGCA-TGTGGATCCTGAGAACTTCAGGCTCCTGGGCAA-TATGAT--TGTGAT-TGTGT--TGGGCCAC--CAC--CTGGGC-AAGGAATTCACCC--CC-TGTGCACAGGCTGCCT-TCCAG-AAGGTGGTGGCTGG-AGTGGC--CAGTGCCCTGG-CTCACAAGTA-CCACTAA-ACCT--CTTTTC---CTGCTC---TTGTCT-TT-GTGCAATGGT--C-AATTGTTCCC-AAG---AGAGC-A-T---CTGTCAGTTGTTG--TCAAA--ATGACAAA--GACCTTTGAAA--ATCTG---TC--CTACTAAT--AAAAGGCATTTA-CTTTCA-CTGC--', score=944.4999999999987, start=5, end=733)\n"
     ]
    }
   ],
   "source": [
    "# Matching characters are given 2 points, 1 point is deducted for each mismatching character\n",
    "# 0.5 points are deducted when opening a gap, and 0.1 points are deducted when extending it\n",
    "local_alignment = pairwise2.align.localms(hbb_hum, hbb_rat, 2, -1, -.5, -.1)\n",
    "print(local_alignment[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'6 TGCTTCTGACACAACTGTGTTCACTAGCAACCTC--AAACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGCAGGCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTCGGTGCCTTTAGTGATGGCCTGGCTCACCTGGACAACCTCAAGGGCACCTTTGCCACACTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGATCCTGAGAACTTCAGGCTCCTGGGCAACGTGCTGGTCTGTGTGCTGGCCCATCACTTTGGCAAAGAATTCACCCCACCAGTGCAGGCTGCCTATCAGAAAGTGGTGGCTGGTGTGGCTAATGCCCTGGCCCACAAGTATCACTAAGCTCGCTTTCTTGCTGTCCAATTTCTATTAAAGGTTCCTTTGTTCCCTAAGTCCAACTACTAAACTGGGGGATATTATGAAGGGCCTTGAGCATCTGGATTCTGCCTAATAAAAAACATTTATTTTCATTGC\\n  |||||||||||.|..||||||.|||..|||.|||  |||||||||||||||||||.||||||..||..||||||.||||.||||.||.|||||||||.|||||||||...|||||.||||||||.|||||||||||||||||||||||.||||||||||||||||||||||.||||||...|||||||||.||||||.||.|...||||.|.|||||.|||||||||||||||||.|||||||||||.|||.|....|||||.|.||||||||||...|||.|||||||||||||||||||||||||....||||||||.||.|||||||||||||||||.||||||||||||||||||||||||||||||||..||.|.||...||||.|||.|||.|||.|.|||||.|||||||||||....|..||||||||||..|||||.|||||||||||.|||||.|.|||||||||.||||||||.||||||.|....||||.||||.|....|||.|...|..|.....|.|...|...|...|......|.....|.|...|.||.|..|||..|.||||..|||||   ||...|||||||||..||||||.|||||.|||\\n1 TGCTTCTGACATAGTTGTGTTGACTCACAAACTCAGAAACAGACACCATGGTGCACCTGACTGATGCTGAGAAGGCTGCTGTTAATGGCCTGTGGGGAAAGGTGAACCCTGATGATGTTGGTGGCGAGGCCCTGGGCAGGCTGCTGGTTGTCTACCCTTGGACCCAGAGGTACTTTGATAGCTTTGGGGACCTGTCCTCTGCCTCTGCTATCATGGGTAACCCTAAGGTGAAGGCCCATGGCAAGAAGGTGATAAACGCCTTCAATGATGGCCTGAAACACTTGGACAACCTCAAGGGCACCTTTGCTCATCTGAGTGAACTCCACTGTGACAAGCTGCATGTGGATCCTGAGAACTTCAGGCTCCTGGGCAATATGATTGTGATTGTGTTGGGCCACCACCTGGGCAAGGAATTCACCCCCTGTGCACAGGCTGCCTTCCAGAAGGTGGTGGCTGGAGTGGCCAGTGCCCTGGCTCACAAGTACCACTAAACCTCTTTTCCTGCTCTTGTCTTTGTGCAATGGTCAATTGTTCCCAAGAGAGCATCTGTCAGTTGTTGTCAAAATGACAAAGACCTTTGAAAATCTG---TCCTACTAATAAAAGGCATTTACTTTCACTGC\\n  Score=451.7\\n'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "local_alignment = pairwise2.align.localxs(hbb_hum, hbb_rat, -10, -.1)\n",
    "display(format_alignment(*local_alignment[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('results_local_alignment.txt', 'w') as file:\n",
    "    file.write(format_alignment(*local_alignment[0]))"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "b51595f36012c8da2bf2284ebbd8d377d5c2d31fa7484164f2d2bcda7a09b462"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit (system)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
